{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import modules\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "import pickle\n",
    "\n",
    "# import scikit-learn modules\n",
    "from sklearn.metrics import RocCurveDisplay, PrecisionRecallDisplay, precision_recall_curve, roc_curve\n",
    "\n",
    "# import visualization modules\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save path\n",
    "get_data_path = lambda folders, fname: os.path.normpath(os.environ['DRIVE_PATH'] + '/' + '/'.join(folders) + '/' + fname)\n",
    "\n",
    "training_data_path = get_data_path(['SL PRED', 'NEW'], 'ito_pairs_annotated.csv')\n",
    "general_predictor_path = get_data_path(['SL PRED', 'input'], 'processed_DeKegel_TableS8.csv')\n",
    "\n",
    "# load the data\n",
    "file_RF_model_I_early = get_data_path(['SL PRED', 'NEW', 'cross_validation', 'model_I'], 'RF_model_early.pickle')\n",
    "file_RF_model_I_late = get_data_path(['SL PRED', 'NEW', 'cross_validation', 'model_I'], 'RF_model_late.pickle')\n",
    "\n",
    "file_RF_model_II_early = get_data_path(['SL PRED', 'NEW', 'cross_validation', 'model_II'], 'RF_model_early.pickle')\n",
    "file_RF_model_II_late = get_data_path(['SL PRED', 'NEW', 'cross_validation', 'model_II'], 'RF_model_late.pickle')\n",
    "\n",
    "file_RF_model_III_early = get_data_path(['SL PRED', 'NEW', 'cross_validation', 'model_III'], 'RF_model_early.pickle')\n",
    "file_RF_model_III_late = get_data_path(['SL PRED', 'NEW', 'cross_validation', 'model_III'], 'RF_model_late.pickle')\n",
    "\n",
    "file_RF_model_IV_early = get_data_path(['SL PRED', 'NEW', 'cross_validation', 'model_IV'], 'RF_model_early.pickle')\n",
    "file_RF_model_IV_late = get_data_path(['SL PRED', 'NEW', 'cross_validation', 'model_IV'], 'RF_model_late.pickle')\n",
    "\n",
    "\n",
    "# path for the plots\n",
    "figure_path_model1 = get_data_path(['SL PRED', 'NEW'], 'model1.png')\n",
    "figure_path_model2 = get_data_path(['SL PRED', 'NEW'], 'model2.png')\n",
    "figure_path_model3 = get_data_path(['SL PRED', 'NEW'], 'model3.png')\n",
    "figure_path_model4 = get_data_path(['SL PRED', 'NEW'], 'model4.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_data_path = get_data_path(['SL PRED', 'NEW'], 'data_SL_pred.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "training_df = pd.read_csv(training_data_path)\n",
    "test_df = pd.read_csv('/Users/narod/Documents/GitHub/context_specific_SL_prediction/1_feature_calculation/annotations/output_files/ito_pairs_annotated_final.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['genepair', 'A1', 'A2', 'A1_entrez', 'A2_entrez', 'DepMap_ID',\n",
       "       'cell_line', 'Gemini_FDR', 'raw_LFC', 'SL', 'org_A1', 'org_A2',\n",
       "       'A1_copy_number_data', 'A1_expression_data', 'A1_gene_effect_data',\n",
       "       'A1_zexpression_data', 'A1_zgene_effect_data', 'A2_copy_number_data',\n",
       "       'A2_expression_data', 'A2_gene_effect_data', 'A2_zexpression_data',\n",
       "       'A2_zgene_effect_data', 'A1_Deleterious', 'A1_mut', 'A2_Deleterious',\n",
       "       'A2_mut', 'zMaxExp_A1A2', 'zMinExp_A1A2', 'rMaxExp_A1A2',\n",
       "       'rMinExp_A1A2', 'max_cn', 'min_cn', 'zMaxESS_A1A2', 'zMinESS_A1A2',\n",
       "       'rMaxESS_A1A2', 'rMinESS_A1A2', 'Protein_Altering', 'Damaging',\n",
       "       'z_ranked_BioGRID_MW', 'ranked_Essentiality_weighted_PPI',\n",
       "       'ranked_BioGRID_MV', 'z_ranked_BioGRID_ALL', 'Expression_weighted_PPI',\n",
       "       'ranked_BioGRID_ALL', 'smallest_GO_CC_expression',\n",
       "       'smallest_GO_ranked_ess', 'smallest_gene_expression',\n",
       "       'smallest_GO_CC_ranked_ess', 'min_seq_id', 'max_seq_id', 'A1_rank',\n",
       "       'zA1_rank', 'A2_rank', 'zA2_rank', 'max_ranked_A1A2', 'min_ranked_A1A2',\n",
       "       'z_max_ranked_A1A2', 'z_min_ranked_A1A2', 'prediction_score', 'GEMINI',\n",
       "       'LFC', 'SL_new'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "training_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['genepair', 'A1', 'A2', 'A1_entrez', 'A2_entrez', 'DepMap_ID',\n",
       "       'cell_line', 'Gemini_FDR', 'raw_LFC', 'SL', 'org_A1', 'org_A2',\n",
       "       'A1_copy_number_data', 'A1_expression_data', 'A1_gene_effect_data',\n",
       "       'A1_zexpression_data', 'A1_zgene_effect_data', 'A2_copy_number_data',\n",
       "       'A2_expression_data', 'A2_gene_effect_data', 'A2_zexpression_data',\n",
       "       'A2_zgene_effect_data', 'A1_Deleterious', 'A1_mut', 'A2_Deleterious',\n",
       "       'A2_mut', 'zMaxExp_A1A2', 'zMinExp_A1A2', 'rMaxExp_A1A2',\n",
       "       'rMinExp_A1A2', 'max_cn', 'min_cn', 'zMaxESS_A1A2', 'zMinESS_A1A2',\n",
       "       'rMaxESS_A1A2', 'rMinESS_A1A2', 'Protein_Altering', 'Damaging',\n",
       "       'z_ranked_BioGRID_MW', 'ranked_Essentiality_weighted_PPI',\n",
       "       'ranked_BioGRID_MV', 'z_ranked_BioGRID_ALL', 'Expression_weighted_PPI',\n",
       "       'ranked_BioGRID_ALL', 'smallest_GO_CC_expression',\n",
       "       'smallest_GO_ranked_ess', 'smallest_gene_expression',\n",
       "       'smallest_GO_CC_ranked_ess', 'A1_rank', 'zA1_rank', 'A2_rank',\n",
       "       'zA2_rank', 'max_ranked_A1A2', 'min_ranked_A1A2', 'z_max_ranked_A1A2',\n",
       "       'z_min_ranked_A1A2', 'GEMINI', 'LFC', 'SL_new'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save path\n",
    "get_data_path = lambda folders, fname: os.path.normpath(os.environ['DRIVE_PATH'] + '/' + '/'.join(folders) + '/' + fname)\n",
    "\n",
    "training_data_path = get_data_path(['SL PRED', 'NEW'], 'ito_pairs_annotated.csv')\n",
    "general_predictor_path = get_data_path(['SL PRED', 'input'], 'processed_DeKegel_TableS8.csv')\n",
    "\n",
    "# load the data\n",
    "file_RF_model_I_early = get_data_path(['SL PRED', 'NEW', 'mean_imputation', 'cross_validation', 'model_I'], 'RF_model_early.pickle')\n",
    "file_RF_model_I_late = get_data_path(['SL PRED', 'NEW', 'mean_imputation', 'cross_validation', 'model_I'], 'RF_model_late.pickle')\n",
    "\n",
    "file_RF_model_II_early = get_data_path(['SL PRED', 'NEW', 'mean_imputation', 'cross_validation', 'model_II'], 'RF_model_early.pickle')\n",
    "file_RF_model_II_late = get_data_path(['SL PRED', 'NEW', 'mean_imputation', 'cross_validation', 'model_II'], 'RF_model_late.pickle')\n",
    "\n",
    "file_RF_model_III_early = get_data_path(['SL PRED', 'NEW', 'mean_imputation', 'cross_validation', 'model_III'], 'RF_model_early.pickle')\n",
    "file_RF_model_III_late = get_data_path(['SL PRED', 'NEW', 'mean_imputation', 'cross_validation', 'model_III'], 'RF_model_late.pickle')\n",
    "\n",
    "file_RF_model_IV_early = get_data_path(['SL PRED', 'NEW', 'mean_imputation', 'cross_validation', 'model_IV'], 'RF_model_early.pickle')\n",
    "file_RF_model_IV_late = get_data_path(['SL PRED', 'NEW', 'mean_imputation', 'cross_validation', 'model_IV'], 'RF_model_late.pickle')\n",
    "\n",
    "# path for the plots\n",
    "figure_path_model1 = get_data_path(['SL PRED', 'NEW', 'mean_imputation'], 'model1.png')\n",
    "figure_path_model2 = get_data_path(['SL PRED', 'NEW', 'mean_imputation'], 'model2.png')\n",
    "figure_path_model3 = get_data_path(['SL PRED', 'NEW', 'mean_imputation'], 'model3.png')\n",
    "figure_path_model4 = get_data_path(['SL PRED', 'NEW', 'mean_imputation'], 'model4.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # save path\n",
    "# get_data_path = lambda folders, fname: os.path.normpath(os.environ['DRIVE_PATH'] + '/' + '/'.join(folders) + '/' + fname)\n",
    "\n",
    "# training_data_path = get_data_path(['SL PRED', 'input','NEW'], 'ito_pairs_all.csv')\n",
    "# general_predictor_path = get_data_path(['SL PRED', 'input'], 'processed_DeKegel_TableS8.csv')\n",
    "\n",
    "# # load the data\n",
    "# file_RF_model_I_early = get_data_path(['SL PRED', 'output', 'cross_validation', 'model_I'], 'RF_model_early_repeated.pickle')\n",
    "# file_RF_model_I_late = get_data_path(['SL PRED', 'output', 'cross_validation', 'model_I'], 'RF_model_late_repeated.pickle')\n",
    "\n",
    "# file_RF_model_II_early = get_data_path(['SL PRED', 'output', 'cross_validation', 'model_II'], 'RF_model_early_repeated.pickle')\n",
    "# file_RF_model_II_late = get_data_path(['SL PRED', 'output', 'cross_validation', 'model_II'], 'RF_model_late_repeated.pickle')\n",
    "\n",
    "# file_RF_model_III_early = get_data_path(['SL PRED', 'output', 'cross_validation', 'model_III'], 'RF_model_early_repeated.pickle')\n",
    "# file_RF_model_III_late = get_data_path(['SL PRED', 'output', 'cross_validation', 'model_III'], 'RF_model_late_repeated.pickle')\n",
    "\n",
    "# #file_RF_model_IV_early = get_data_path(['SL PRED', 'output', 'cross_validation', 'model_IV'], 'RF_model_early_repeated.pickle')\n",
    "# #file_RF_model_IV_late = get_data_path(['SL PRED', 'output', 'cross_validation', 'model_IV'], 'RF_model_late_repeated.pickle')\n",
    "\n",
    "\n",
    "# # path for the plots\n",
    "# figure_path_model1 = get_data_path(['SL PRED', 'Fig 3', 'Fig3_new_1704'], 'model1_1804.png')\n",
    "# figure_path_model2 = get_data_path(['SL PRED', 'Fig 3', 'Fig3_new_1704'], 'model2_1804.png')\n",
    "# figure_path_model3 = get_data_path(['SL PRED', 'Fig 3', 'Fig3_new_1704'], 'model3_1804.png')\n",
    "# figure_path_model4 = get_data_path(['SL PRED', 'Fig 3', 'Fig3_new_1704'], 'model4_1804.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "training_df = pd.read_csv(training_data_path)\n",
    "general_classifier_training_df = pd.read_csv(general_predictor_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "late_features_df = general_classifier_training_df[['genepair', 'A1_entrez_new', 'A2_entrez_new', 'min_sequence_identity', 'closest', 'WGD', 'family_size', 'cds_length_ratio', 'shared_domains', 'has_pombe_ortholog',\n",
    "                                            'has_essential_pombe_ortholog', 'has_cerevisiae_ortholog', 'has_essential_cerevisiae_ortholog', 'conservation_score', 'mean_age', \n",
    "                                            'either_in_complex', 'mean_complex_essentiality', 'colocalisation', 'interact', 'n_total_ppi', 'fet_ppi_overlap','gtex_spearman_corr', \n",
    "                                            'gtex_min_mean_expr', 'gtex_max_mean_expr', 'shared_ppi_mean_essentiality']]\n",
    "late_features_df = late_features_df.rename(columns={'A1_entrez_new': 'A1_entrez', 'A2_entrez_new': 'A2_entrez'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genepair</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>A1_entrez</th>\n",
       "      <th>A2_entrez</th>\n",
       "      <th>DepMap_ID</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>Gemini_FDR</th>\n",
       "      <th>raw_LFC</th>\n",
       "      <th>SL</th>\n",
       "      <th>...</th>\n",
       "      <th>either_in_complex</th>\n",
       "      <th>mean_complex_essentiality</th>\n",
       "      <th>colocalisation</th>\n",
       "      <th>interact</th>\n",
       "      <th>n_total_ppi</th>\n",
       "      <th>fet_ppi_overlap</th>\n",
       "      <th>gtex_spearman_corr</th>\n",
       "      <th>gtex_min_mean_expr</th>\n",
       "      <th>gtex_max_mean_expr</th>\n",
       "      <th>shared_ppi_mean_essentiality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A3GALT2_ABO</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>ABO</td>\n",
       "      <td>127550</td>\n",
       "      <td>28</td>\n",
       "      <td>ACH-000022</td>\n",
       "      <td>PATU8988S_PANCREAS</td>\n",
       "      <td>0.998944</td>\n",
       "      <td>0.088856</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.114847</td>\n",
       "      <td>0.258739</td>\n",
       "      <td>11.702</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A3GALT2_ABO</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>ABO</td>\n",
       "      <td>127550</td>\n",
       "      <td>28</td>\n",
       "      <td>ACH-000307</td>\n",
       "      <td>PK1_PANCREAS</td>\n",
       "      <td>0.986587</td>\n",
       "      <td>0.201704</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.114847</td>\n",
       "      <td>0.258739</td>\n",
       "      <td>11.702</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A3GALT2_ABO</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>ABO</td>\n",
       "      <td>127550</td>\n",
       "      <td>28</td>\n",
       "      <td>ACH-000632</td>\n",
       "      <td>HS944T_SKIN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.069772</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.114847</td>\n",
       "      <td>0.258739</td>\n",
       "      <td>11.702</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 84 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      genepair       A1   A2  A1_entrez  A2_entrez   DepMap_ID  \\\n",
       "0  A3GALT2_ABO  A3GALT2  ABO     127550         28  ACH-000022   \n",
       "1  A3GALT2_ABO  A3GALT2  ABO     127550         28  ACH-000307   \n",
       "2  A3GALT2_ABO  A3GALT2  ABO     127550         28  ACH-000632   \n",
       "\n",
       "            cell_line  Gemini_FDR   raw_LFC     SL  ... either_in_complex  \\\n",
       "0  PATU8988S_PANCREAS    0.998944  0.088856  False  ...             False   \n",
       "1        PK1_PANCREAS    0.986587  0.201704  False  ...             False   \n",
       "2         HS944T_SKIN    1.000000  0.069772  False  ...             False   \n",
       "\n",
       "  mean_complex_essentiality  colocalisation  interact  n_total_ppi  \\\n",
       "0                       0.0             0.0     False          3.0   \n",
       "1                       0.0             0.0     False          3.0   \n",
       "2                       0.0             0.0     False          3.0   \n",
       "\n",
       "   fet_ppi_overlap  gtex_spearman_corr  gtex_min_mean_expr  \\\n",
       "0              0.0            0.114847            0.258739   \n",
       "1              0.0            0.114847            0.258739   \n",
       "2              0.0            0.114847            0.258739   \n",
       "\n",
       "   gtex_max_mean_expr  shared_ppi_mean_essentiality  \n",
       "0              11.702                           0.0  \n",
       "1              11.702                           0.0  \n",
       "2              11.702                           0.0  \n",
       "\n",
       "[3 rows x 84 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "integrated_training_df = pd.merge(\n",
    "    training_df, late_features_df, \n",
    "    left_on=['genepair', 'A1_entrez', 'A2_entrez'], \n",
    "    right_on=['genepair', 'A1_entrez', 'A2_entrez'], \n",
    "    how='left')\n",
    "\n",
    "bool_columns = ['closest', 'WGD', 'has_pombe_ortholog', 'has_cerevisiae_ortholog', 'has_essential_cerevisiae_ortholog', 'either_in_complex', 'interact']\n",
    "\n",
    "integrated_training_df[bool_columns] = integrated_training_df[bool_columns].astype(bool)\n",
    "\n",
    "integrated_training_df[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genepair</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>A1_entrez</th>\n",
       "      <th>A2_entrez</th>\n",
       "      <th>DepMap_ID</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>Gemini_FDR</th>\n",
       "      <th>raw_LFC</th>\n",
       "      <th>SL</th>\n",
       "      <th>...</th>\n",
       "      <th>either_in_complex</th>\n",
       "      <th>mean_complex_essentiality</th>\n",
       "      <th>colocalisation</th>\n",
       "      <th>interact</th>\n",
       "      <th>n_total_ppi</th>\n",
       "      <th>fet_ppi_overlap</th>\n",
       "      <th>gtex_spearman_corr</th>\n",
       "      <th>gtex_min_mean_expr</th>\n",
       "      <th>gtex_max_mean_expr</th>\n",
       "      <th>shared_ppi_mean_essentiality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A3GALT2_ABO</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>ABO</td>\n",
       "      <td>127550</td>\n",
       "      <td>28</td>\n",
       "      <td>ACH-000022</td>\n",
       "      <td>PATU8988S_PANCREAS</td>\n",
       "      <td>0.998944</td>\n",
       "      <td>0.088856</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.114847</td>\n",
       "      <td>0.258739</td>\n",
       "      <td>11.702</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A3GALT2_ABO</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>ABO</td>\n",
       "      <td>127550</td>\n",
       "      <td>28</td>\n",
       "      <td>ACH-000307</td>\n",
       "      <td>PK1_PANCREAS</td>\n",
       "      <td>0.986587</td>\n",
       "      <td>0.201704</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.114847</td>\n",
       "      <td>0.258739</td>\n",
       "      <td>11.702</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A3GALT2_ABO</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>ABO</td>\n",
       "      <td>127550</td>\n",
       "      <td>28</td>\n",
       "      <td>ACH-000632</td>\n",
       "      <td>HS944T_SKIN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.069772</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.114847</td>\n",
       "      <td>0.258739</td>\n",
       "      <td>11.702</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 81 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      genepair       A1   A2  A1_entrez  A2_entrez   DepMap_ID  \\\n",
       "0  A3GALT2_ABO  A3GALT2  ABO     127550         28  ACH-000022   \n",
       "1  A3GALT2_ABO  A3GALT2  ABO     127550         28  ACH-000307   \n",
       "2  A3GALT2_ABO  A3GALT2  ABO     127550         28  ACH-000632   \n",
       "\n",
       "            cell_line  Gemini_FDR   raw_LFC     SL  ... either_in_complex  \\\n",
       "0  PATU8988S_PANCREAS    0.998944  0.088856  False  ...             False   \n",
       "1        PK1_PANCREAS    0.986587  0.201704  False  ...             False   \n",
       "2         HS944T_SKIN    1.000000  0.069772  False  ...             False   \n",
       "\n",
       "  mean_complex_essentiality  colocalisation  interact  n_total_ppi  \\\n",
       "0                       0.0             0.0     False          3.0   \n",
       "1                       0.0             0.0     False          3.0   \n",
       "2                       0.0             0.0     False          3.0   \n",
       "\n",
       "   fet_ppi_overlap  gtex_spearman_corr  gtex_min_mean_expr  \\\n",
       "0              0.0            0.114847            0.258739   \n",
       "1              0.0            0.114847            0.258739   \n",
       "2              0.0            0.114847            0.258739   \n",
       "\n",
       "   gtex_max_mean_expr  shared_ppi_mean_essentiality  \n",
       "0              11.702                           0.0  \n",
       "1              11.702                           0.0  \n",
       "2              11.702                           0.0  \n",
       "\n",
       "[3 rows x 81 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "integrated_test_df = pd.merge(\n",
    "    test_df, late_features_df, \n",
    "    left_on=['genepair', 'A1_entrez', 'A2_entrez'], \n",
    "    right_on=['genepair', 'A1_entrez', 'A2_entrez'], \n",
    "    how='left')\n",
    "\n",
    "bool_columns = ['closest', 'WGD', 'has_pombe_ortholog', 'has_cerevisiae_ortholog', 'has_essential_cerevisiae_ortholog', 'either_in_complex', 'interact']\n",
    "\n",
    "integrated_test_df[bool_columns] = integrated_test_df[bool_columns].astype(bool)\n",
    "\n",
    "integrated_test_df[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "num of features: 16\n"
     ]
    }
   ],
   "source": [
    "feature_columns_1 = ['rMaxExp_A1A2', 'rMinExp_A1A2',\n",
    "                     'max_ranked_A1A2', 'min_ranked_A1A2',\n",
    "                     'max_cn', 'min_cn', 'Protein_Altering', 'Damaging', \n",
    "                     'min_sequence_identity',\n",
    "                     'prediction_score', \n",
    "                     'ranked_Essentiality_weighted_PPI', 'Expression_weighted_PPI',\n",
    "                     'smallest_GO_ranked_ess', 'smallest_GO_CC_ranked_ess',\n",
    "                     'smallest_gene_expression', 'smallest_GO_CC_expression'\n",
    "                     ]\n",
    "\n",
    "target_column = 'SL_new'\n",
    "\n",
    "print('num of features:', len(feature_columns_1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "num of features: 35\n"
     ]
    }
   ],
   "source": [
    "feature_columns_2 = feature_columns_1 + ['closest', 'WGD', 'family_size',\n",
    "                                         'cds_length_ratio', 'shared_domains', 'has_pombe_ortholog',\n",
    "                                         'has_essential_pombe_ortholog', 'has_cerevisiae_ortholog', 'has_essential_cerevisiae_ortholog', \n",
    "                                         'conservation_score', 'mean_age', 'either_in_complex', 'mean_complex_essentiality', 'colocalisation',\n",
    "                                         'interact', 'n_total_ppi', 'fet_ppi_overlap',\n",
    "                                         'gtex_spearman_corr', 'gtex_min_mean_expr', 'gtex_max_mean_expr']\n",
    "feature_columns_2.remove('prediction_score')\n",
    "print('num of features:', len(feature_columns_2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocess_dataset(df, old_df,\n",
    "                       required_genepairs_col='genepair',\n",
    "                       dropna_cols=None,\n",
    "                       fillna_zero_cols=None,\n",
    "                       fillna_large_cols=None,\n",
    "                       fillna_large_value=18000):\n",
    "    \"\"\"\n",
    "    Preprocess a training or testing dataset:\n",
    "    - Keep only rows with genepairs present in `old_df`\n",
    "    - Drop rows with NaN in specific columns\n",
    "    - Fill missing values with default values\n",
    "    \n",
    "    Parameters:\n",
    "        df (pd.DataFrame): Dataset to process\n",
    "        old_df (pd.DataFrame): Dataset with allowed genepairs\n",
    "        required_genepairs_col (str): Column name for genepairs to match\n",
    "        dropna_cols (list): Columns for which rows with NaNs should be dropped\n",
    "        fillna_zero_cols (list): Columns to fill NaNs with 0\n",
    "        fillna_large_cols (list): Columns to fill NaNs with large constant\n",
    "        fillna_large_value (int or float): The large value to fill (default: 18000)\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: Cleaned and processed DataFrame\n",
    "    \"\"\"\n",
    "\n",
    "    # Step 1: Filter rows to only those in old_df\n",
    "    df_filtered = df[df[required_genepairs_col].isin(old_df[required_genepairs_col])].copy()\n",
    "\n",
    "    # Step 2: Drop rows with any NA in required columns\n",
    "    if dropna_cols:\n",
    "        df_filtered = df_filtered.dropna(axis=0, how='any', subset=dropna_cols)\n",
    "\n",
    "    # Step 3: Fill NaNs with 0 or large number\n",
    "    if fillna_zero_cols:\n",
    "        df_filtered[fillna_zero_cols] = df_filtered[fillna_zero_cols].fillna(0)\n",
    "\n",
    "    if fillna_large_cols:\n",
    "        df_filtered[fillna_large_cols] = df_filtered[fillna_large_cols].fillna(fillna_large_value)\n",
    "\n",
    "    # Step 4: Reset index for clean result\n",
    "    return df_filtered.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "drop_na_values = ['rMaxExp_A1A2', 'rMinExp_A1A2', 'max_ranked_A1A2', 'min_ranked_A1A2']\n",
    "fillna_values = ['Expression_weighted_PPI', 'smallest_gene_expression', 'smallest_GO_CC_expression']\n",
    "fillna_values_v2 = ['ranked_Essentiality_weighted_PPI', 'smallest_GO_ranked_ess', 'smallest_GO_CC_ranked_ess']\n",
    "\n",
    "# Apply to training set\n",
    "integrated_training_df_clean = preprocess_dataset(\n",
    "    df=integrated_training_df,\n",
    "    old_df=general_classifier_training_df,\n",
    "    dropna_cols=drop_na_values,\n",
    "    fillna_zero_cols=fillna_values,\n",
    "    fillna_large_cols=fillna_values_v2\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocess_dataset(df, old_df,\n",
    "                       required_genepairs_col='genepair',\n",
    "                       dropna_cols=None,\n",
    "                       impute_mean_cols=None,\n",
    "                       groupby_col='DepMap_ID'):\n",
    "    \"\"\"\n",
    "    Preprocess a training or testing dataset:\n",
    "    - Keep only rows with genepairs present in `old_df`\n",
    "    - Drop rows with NaN in specific columns\n",
    "    - Fill missing values using mean imputation within groups\n",
    "    \n",
    "    Parameters:\n",
    "        df (pd.DataFrame): Dataset to process\n",
    "        old_df (pd.DataFrame): Dataset with allowed genepairs\n",
    "        required_genepairs_col (str): Column name for genepairs to match\n",
    "        dropna_cols (list): Columns for which rows with NaNs should be dropped\n",
    "        impute_mean_cols (list): Columns to fill NaNs using mean of group\n",
    "        groupby_col (str): Column to group by for mean imputation\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: Cleaned and processed DataFrame\n",
    "    \"\"\"\n",
    "\n",
    "    # Step 1: Filter rows to only those in old_df\n",
    "    df_filtered = df[df[required_genepairs_col].isin(old_df[required_genepairs_col])].copy()\n",
    "\n",
    "    # Step 2: Drop rows with any NA in required columns\n",
    "    if dropna_cols:\n",
    "        df_filtered = df_filtered.dropna(axis=0, how='any', subset=dropna_cols)\n",
    "\n",
    "    # Step 3: Mean imputation within groups\n",
    "    if impute_mean_cols:\n",
    "        df_filtered[impute_mean_cols] = (\n",
    "            df_filtered.groupby(groupby_col)[impute_mean_cols]\n",
    "            .transform(lambda x: x.fillna(x.mean()))\n",
    "        )\n",
    "\n",
    "    # Step 4: Reset index for clean result\n",
    "    return df_filtered.reset_index(drop=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "drop_na_values = ['rMaxExp_A1A2', 'rMinExp_A1A2', 'max_ranked_A1A2', 'min_ranked_A1A2']\n",
    "impute_features = ['ranked_Essentiality_weighted_PPI', 'Expression_weighted_PPI', 'smallest_GO_ranked_ess', 'smallest_GO_CC_ranked_ess', 'smallest_gene_expression', 'smallest_GO_CC_expression']\n",
    "\n",
    "# Apply to training set\n",
    "integrated_training_df_clean = preprocess_dataset(\n",
    "    df=integrated_training_df,\n",
    "    old_df=general_classifier_training_df,\n",
    "    dropna_cols=drop_na_values,\n",
    "    impute_mean_cols=impute_features,\n",
    "    groupby_col='DepMap_ID'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num SL: 958 / 41244\n",
      "Num non-SL: 40286 / 41244\n",
      "Number of unique gene pairs: 4170\n",
      "Number of unique cell lines: 10\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genepair</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>A1_entrez</th>\n",
       "      <th>A2_entrez</th>\n",
       "      <th>DepMap_ID</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>Gemini_FDR</th>\n",
       "      <th>raw_LFC</th>\n",
       "      <th>SL</th>\n",
       "      <th>...</th>\n",
       "      <th>either_in_complex</th>\n",
       "      <th>mean_complex_essentiality</th>\n",
       "      <th>colocalisation</th>\n",
       "      <th>interact</th>\n",
       "      <th>n_total_ppi</th>\n",
       "      <th>fet_ppi_overlap</th>\n",
       "      <th>gtex_spearman_corr</th>\n",
       "      <th>gtex_min_mean_expr</th>\n",
       "      <th>gtex_max_mean_expr</th>\n",
       "      <th>shared_ppi_mean_essentiality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A3GALT2_ABO</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>ABO</td>\n",
       "      <td>127550</td>\n",
       "      <td>28</td>\n",
       "      <td>ACH-000022</td>\n",
       "      <td>PATU8988S_PANCREAS</td>\n",
       "      <td>0.998944</td>\n",
       "      <td>0.088856</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.114847</td>\n",
       "      <td>0.258739</td>\n",
       "      <td>11.702</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A3GALT2_ABO</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>ABO</td>\n",
       "      <td>127550</td>\n",
       "      <td>28</td>\n",
       "      <td>ACH-000307</td>\n",
       "      <td>PK1_PANCREAS</td>\n",
       "      <td>0.986587</td>\n",
       "      <td>0.201704</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.114847</td>\n",
       "      <td>0.258739</td>\n",
       "      <td>11.702</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A3GALT2_ABO</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>ABO</td>\n",
       "      <td>127550</td>\n",
       "      <td>28</td>\n",
       "      <td>ACH-000632</td>\n",
       "      <td>HS944T_SKIN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.069772</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.114847</td>\n",
       "      <td>0.258739</td>\n",
       "      <td>11.702</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 84 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      genepair       A1   A2  A1_entrez  A2_entrez   DepMap_ID  \\\n",
       "0  A3GALT2_ABO  A3GALT2  ABO     127550         28  ACH-000022   \n",
       "1  A3GALT2_ABO  A3GALT2  ABO     127550         28  ACH-000307   \n",
       "2  A3GALT2_ABO  A3GALT2  ABO     127550         28  ACH-000632   \n",
       "\n",
       "            cell_line  Gemini_FDR   raw_LFC     SL  ... either_in_complex  \\\n",
       "0  PATU8988S_PANCREAS    0.998944  0.088856  False  ...             False   \n",
       "1        PK1_PANCREAS    0.986587  0.201704  False  ...             False   \n",
       "2         HS944T_SKIN    1.000000  0.069772  False  ...             False   \n",
       "\n",
       "  mean_complex_essentiality  colocalisation  interact  n_total_ppi  \\\n",
       "0                       0.0             0.0     False          3.0   \n",
       "1                       0.0             0.0     False          3.0   \n",
       "2                       0.0             0.0     False          3.0   \n",
       "\n",
       "   fet_ppi_overlap  gtex_spearman_corr  gtex_min_mean_expr  \\\n",
       "0              0.0            0.114847            0.258739   \n",
       "1              0.0            0.114847            0.258739   \n",
       "2              0.0            0.114847            0.258739   \n",
       "\n",
       "   gtex_max_mean_expr  shared_ppi_mean_essentiality  \n",
       "0              11.702                           0.0  \n",
       "1              11.702                           0.0  \n",
       "2              11.702                           0.0  \n",
       "\n",
       "[3 rows x 84 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('Num SL:', integrated_training_df_clean[integrated_training_df_clean[target_column] == True].shape[0], '/', integrated_training_df_clean.shape[0])\n",
    "print('Num non-SL:', integrated_training_df_clean[integrated_training_df_clean[target_column] == False].shape[0], '/', integrated_training_df_clean.shape[0])\n",
    "print(f'Number of unique gene pairs: {integrated_training_df_clean.genepair.nunique()}')\n",
    "print(f'Number of unique cell lines: {integrated_training_df_clean.cell_line.nunique()}')\n",
    "integrated_training_df_clean[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_df = preprocess_dataset(\n",
    "    df=integrated_test_df,\n",
    "    old_df=general_classifier_training_df,\n",
    "    dropna_cols=drop_na_values,\n",
    "    impute_mean_cols=impute_features,\n",
    "    groupby_col='DepMap_ID'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num SL: 958 / 41244\n",
      "Num non-SL: 40286 / 41244\n",
      "Number of unique gene pairs: 4170\n",
      "Number of unique cell lines: 10\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genepair</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>A1_entrez</th>\n",
       "      <th>A2_entrez</th>\n",
       "      <th>DepMap_ID</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>Gemini_FDR</th>\n",
       "      <th>raw_LFC</th>\n",
       "      <th>SL</th>\n",
       "      <th>...</th>\n",
       "      <th>either_in_complex</th>\n",
       "      <th>mean_complex_essentiality</th>\n",
       "      <th>colocalisation</th>\n",
       "      <th>interact</th>\n",
       "      <th>n_total_ppi</th>\n",
       "      <th>fet_ppi_overlap</th>\n",
       "      <th>gtex_spearman_corr</th>\n",
       "      <th>gtex_min_mean_expr</th>\n",
       "      <th>gtex_max_mean_expr</th>\n",
       "      <th>shared_ppi_mean_essentiality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A3GALT2_ABO</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>ABO</td>\n",
       "      <td>127550</td>\n",
       "      <td>28</td>\n",
       "      <td>ACH-000022</td>\n",
       "      <td>PATU8988S_PANCREAS</td>\n",
       "      <td>0.998944</td>\n",
       "      <td>0.088856</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.114847</td>\n",
       "      <td>0.258739</td>\n",
       "      <td>11.702</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A3GALT2_ABO</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>ABO</td>\n",
       "      <td>127550</td>\n",
       "      <td>28</td>\n",
       "      <td>ACH-000307</td>\n",
       "      <td>PK1_PANCREAS</td>\n",
       "      <td>0.986587</td>\n",
       "      <td>0.201704</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.114847</td>\n",
       "      <td>0.258739</td>\n",
       "      <td>11.702</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A3GALT2_ABO</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>ABO</td>\n",
       "      <td>127550</td>\n",
       "      <td>28</td>\n",
       "      <td>ACH-000632</td>\n",
       "      <td>HS944T_SKIN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.069772</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.114847</td>\n",
       "      <td>0.258739</td>\n",
       "      <td>11.702</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 81 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      genepair       A1   A2  A1_entrez  A2_entrez   DepMap_ID  \\\n",
       "0  A3GALT2_ABO  A3GALT2  ABO     127550         28  ACH-000022   \n",
       "1  A3GALT2_ABO  A3GALT2  ABO     127550         28  ACH-000307   \n",
       "2  A3GALT2_ABO  A3GALT2  ABO     127550         28  ACH-000632   \n",
       "\n",
       "            cell_line  Gemini_FDR   raw_LFC     SL  ... either_in_complex  \\\n",
       "0  PATU8988S_PANCREAS    0.998944  0.088856  False  ...             False   \n",
       "1        PK1_PANCREAS    0.986587  0.201704  False  ...             False   \n",
       "2         HS944T_SKIN    1.000000  0.069772  False  ...             False   \n",
       "\n",
       "  mean_complex_essentiality  colocalisation  interact  n_total_ppi  \\\n",
       "0                       0.0             0.0     False          3.0   \n",
       "1                       0.0             0.0     False          3.0   \n",
       "2                       0.0             0.0     False          3.0   \n",
       "\n",
       "   fet_ppi_overlap  gtex_spearman_corr  gtex_min_mean_expr  \\\n",
       "0              0.0            0.114847            0.258739   \n",
       "1              0.0            0.114847            0.258739   \n",
       "2              0.0            0.114847            0.258739   \n",
       "\n",
       "   gtex_max_mean_expr  shared_ppi_mean_essentiality  \n",
       "0              11.702                           0.0  \n",
       "1              11.702                           0.0  \n",
       "2              11.702                           0.0  \n",
       "\n",
       "[3 rows x 81 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('Num SL:', test_df[test_df[target_column] == True].shape[0], '/', test_df.shape[0])\n",
    "print('Num non-SL:', test_df[test_df[target_column] == False].shape[0], '/', test_df.shape[0])\n",
    "print(f'Number of unique gene pairs: {test_df.genepair.nunique()}')\n",
    "print(f'Number of unique cell lines: {test_df.cell_line.nunique()}')\n",
    "test_df[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "integrated_training_df_clean[feature_columns_2].isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# impute_features = ['ranked_Essentiality_weighted_PPI', 'Expression_weighted_PPI', 'smallest_GO_ranked_ess', 'smallest_GO_CC_ranked_ess', 'smallest_gene_expression', 'smallest_GO_CC_expression']\n",
    "\n",
    "# integrated_training_df_clean = integrated_training_df.copy()\n",
    "\n",
    "# integrated_training_df_clean[impute_features] = integrated_training_df_clean.groupby(\"DepMap_ID\")[impute_features].transform(lambda x: x.fillna(x.mean()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #remove NA values before training the model\n",
    "# integrated_training_df_clean = integrated_training_df_clean.dropna(axis=0, how='any', subset=feature_columns_2 + [target_column]).reset_index(drop=True) \n",
    "\n",
    "# #summary of the training dataset after removing NA values\n",
    "# print('Num SL:', integrated_training_df_clean[integrated_training_df_clean[target_column] == True].shape[0], '/', integrated_training_df_clean.shape[0])\n",
    "# print('Num non-SL:', integrated_training_df_clean[integrated_training_df_clean[target_column] == False].shape[0], '/', integrated_training_df_clean.shape[0])\n",
    "# print(f'Number of unique gene pairs: {integrated_training_df_clean.genepair.nunique()}')\n",
    "# print(f'Number of unique cell lines: {integrated_training_df_clean.cell_line.nunique()}')\n",
    "# integrated_training_df_clean[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define feature sets\n",
    "data_1 = integrated_training_df_clean[feature_columns_1]\n",
    "data_2 = integrated_training_df_clean[feature_columns_2]\n",
    "target = integrated_training_df_clean[target_column]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load RF classifier from file\n",
    "# with open(file_RF_model_I_early, 'rb') as file:\n",
    "#     model_I_early = pickle.load(file)\n",
    "\n",
    "# with open(file_RF_model_I_late, 'rb') as file:\n",
    "#     model_I_late = pickle.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load RF classifier from file\n",
    "with open(file_RF_model_I_early, 'rb') as file:\n",
    "    model_I_early = pickle.load(file)\n",
    "\n",
    "with open(file_RF_model_I_late, 'rb') as file:\n",
    "    model_I_late = pickle.load(file)\n",
    "\n",
    "with open(file_RF_model_II_early, 'rb') as file:\n",
    "    model_II_early = pickle.load(file)\n",
    "\n",
    "with open(file_RF_model_II_late, 'rb') as file:\n",
    "    model_II_late = pickle.load(file)\n",
    "\n",
    "with open(file_RF_model_III_early, 'rb') as file:\n",
    "    model_III_early = pickle.load(file)\n",
    "\n",
    "with open(file_RF_model_III_late, 'rb') as file:\n",
    "    model_III_late = pickle.load(file)\n",
    "\n",
    "with open(file_RF_model_IV_early, 'rb') as file:\n",
    "    model_IV_early = pickle.load(file)\n",
    "\n",
    "with open(file_RF_model_IV_late, 'rb') as file:\n",
    "    model_IV_late = pickle.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(file_RF_model_I_early_repeated, 'rb') as file:\n",
    "    model_I_early_repeated = pickle.load(file)\n",
    "\n",
    "with open(file_RF_model_I_late_repeated, 'rb') as file:\n",
    "    model_I_late_repeated = pickle.load(file)\n",
    "\n",
    "with open(file_RF_model_II_early_repeated, 'rb') as file:\n",
    "    model_II_early_repeated = pickle.load(file)\n",
    "\n",
    "with open(file_RF_model_II_late_repeated, 'rb') as file:\n",
    "    model_II_late_repeated = pickle.load(file)\n",
    "\n",
    "with open(file_RF_model_III_early_repeated, 'rb') as file:\n",
    "    model_III_early_repeated = pickle.load(file)\n",
    "\n",
    "with open(file_RF_model_III_late_repeated, 'rb') as file:\n",
    "    model_III_late_repeated = pickle.load(file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(file_RF_model_I_early_reduced, 'rb') as file:\n",
    "    model_I_early_reduced = pickle.load(file)\n",
    "\n",
    "with open(file_RF_model_I_late_reduced, 'rb') as file:\n",
    "    model_I_late_reduced = pickle.load(file)\n",
    "\n",
    "with open(file_RF_model_II_early_reduced, 'rb') as file:\n",
    "    model_II_early_reduced = pickle.load(file)\n",
    "\n",
    "with open(file_RF_model_II_late_reduced, 'rb') as file:\n",
    "    model_II_late_reduced = pickle.load(file)\n",
    "\n",
    "with open(file_RF_model_III_early_reduced, 'rb') as file:\n",
    "    model_III_early_reduced = pickle.load(file)\n",
    "\n",
    "with open(file_RF_model_III_late_reduced, 'rb') as file:\n",
    "    model_III_late_reduced = pickle.load(file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot ROC AUC & PR AUC Curves"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def plot_model_performance(df, data, target, model_early, model_late):#, file_path):\n",
    "#     sns.set_context('paper')\n",
    "#     f, ax = plt.subplots(1, 2, figsize=(9.5, 4.5))  # Adjusted for better visibility\n",
    "#     #f, ax = plt.subplots(1, 2, figsize=(2.6, 1.35))\n",
    "\n",
    "#     # Main classifier ROC and PR curves (early)\n",
    "#     roc_display = RocCurveDisplay(fpr=model_early['mean_fpr'], tpr=model_early['mean_tpr'], roc_auc=model_early['mean_auc'], estimator_name='Context-Specific Classifier (early)')\n",
    "#     roc_display.plot(ax=ax[0], alpha=0.8, color='#D55E00', linewidth=0.8)\n",
    "\n",
    "#     pr_display = PrecisionRecallDisplay(precision=model_early['precision'], recall=model_early['recall'], average_precision=model_early['mean_aps'], estimator_name='Context-Specific Classifier (early)')\n",
    "#     pr_display.plot(ax=ax[1], alpha=0.8, color='#D55E00', linewidth=0.8)\n",
    "\n",
    "#     # Main classifier ROC and PR curves (late)\n",
    "#     roc_display = RocCurveDisplay(fpr=model_late['mean_fpr'], tpr=model_late['mean_tpr'], roc_auc=model_late['mean_auc'], estimator_name='Context-Specific Classifier (late)')\n",
    "#     roc_display.plot(ax=ax[0], alpha=0.8, color='#000000', linewidth=0.8)\n",
    "\n",
    "#     pr_display = PrecisionRecallDisplay(precision=model_late['precision'], recall=model_late['recall'], average_precision=model_late['mean_aps'], estimator_name='Context-Specific Classifier (late)')\n",
    "#     pr_display.plot(ax=ax[1], alpha=0.8, color='#000000', linewidth=0.8)\n",
    "\n",
    "#     # Additional classifiers\n",
    "#     metrics = ['prediction_score', 'rMinExp_A1A2', 'min_ranked_A1A2', 'max_seq_id']\n",
    "#     colors = ['#CC79A7', '#56B4E9', '#009E73', '#E69F00']\n",
    "#     labels = ['General Classifier', 'Gene Expression', 'Gene Essentiality', 'Sequence Identity']\n",
    "#     auc_labels = ['pred_auc', 'gene_expr_auc', 'gene_ess_auc', 'seq_auc']\n",
    "#     ap_labels = ['pred_ap', 'gene_expr_ap', 'gene_ess_ap', 'seq_ap']\n",
    "\n",
    "#     for metric, color, label, auc_label, ap_label in zip(metrics, colors, labels, auc_labels, ap_labels):\n",
    "#         fpr, tpr, _ = roc_curve(target, data[metric] if metric != 'min_ranked_A1A2' else -data[metric])\n",
    "#         roc_display = RocCurveDisplay(fpr=fpr, tpr=tpr, roc_auc=model_early[auc_label], estimator_name=label)\n",
    "#         roc_display.plot(ax=ax[0], alpha=0.8, color=color, linewidth=0.8)\n",
    "\n",
    "#         precision, recall, _ = precision_recall_curve(target, data[metric] if metric != 'min_ranked_A1A2' else -data[metric])\n",
    "#         recall = recall[:-1]\n",
    "#         precision = precision[:-1]\n",
    "#         pr_display = PrecisionRecallDisplay(precision=precision, recall=recall, average_precision=model_early[ap_label], estimator_name=label)\n",
    "#         pr_display.plot(ax=ax[1], alpha=0.8, color=color, linewidth=0.8)\n",
    "\n",
    "#     ax[0].plot([0, 1], [0, 1], color=\"lightgrey\", linestyle=\"--\", label='Chance (0.50)')\n",
    "#     ax[0].set_xlim([-0.025, 1.025])\n",
    "#     ax[0].set_ylim([-0.025, 1.025])\n",
    "#     ax[0].spines.top.set(visible=False)\n",
    "#     ax[0].spines.right.set(visible=False)\n",
    "#     ax[0].tick_params(axis='both', which='major', labelsize=5)\n",
    "#     ax[0].set_xlabel('False Positive Rate', fontsize=12)\n",
    "#     ax[0].set_ylabel('True Positive Rate', fontsize=12)\n",
    "#     ax[0].legend(loc='lower right', fontsize=9, bbox_to_anchor=(1.02, 0))\n",
    "\n",
    "#     no_skill = round(sum(df['SL_new']) / len(df['SL_new']), 2)\n",
    "#     ax[1].plot([0, 1], [no_skill, no_skill], linestyle='--', color='lightgrey', label=f'Chance ({no_skill:.2f})')\n",
    "#     ax[1].set_xlim([-0.025, 1.025])\n",
    "#     ax[1].set_ylim([-0.025, 1.025])\n",
    "#     ax[1].spines.top.set(visible=False)\n",
    "#     ax[1].spines.right.set(visible=False)\n",
    "#     ax[1].set_xlabel('Recall', fontsize=12)\n",
    "#     ax[1].tick_params(axis='both', which='major', labelsize=5)\n",
    "#     ax[1].set_ylabel('Precision', fontsize=12)\n",
    "#     ax[1].legend(loc='upper right', fontsize=9)\n",
    "\n",
    "#     plt.tight_layout(h_pad=0.7)\n",
    "    #plt.savefig(f'{file_path}', bbox_inches='tight', dpi=350)\n",
    "    #plt.close(f)  # Optionally close the figure to save memory if generating many plots\n",
    "\n",
    "# Example usage:\n",
    "# plot_model_performance(data, target, cv_params, '/Users/narod/Desktop/model_performance.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot_model_performance(integrated_training_df_clean, data_1, target, model_I_early, model_I_late)#, figure_path_model1_repeated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the labels and their styles\n",
    "legend_labels = [\n",
    "    \"Context-Specific Classifier (early)\",\n",
    "    \"Context-Specific Classifier (late)\",\n",
    "    \"General Classifier\",\n",
    "    #\"Gene Expression\",\n",
    "    #\"Gene Essentiality\",\n",
    "    \"Sequence Identity\",\n",
    "    \"Chance\"\n",
    "]\n",
    "\n",
    "legend_styles = [\n",
    "    {\"color\": \"#009E73\", \"linestyle\": \"-\"},\n",
    "    {\"color\": \"#CC79A7\", \"linestyle\": \"-\"},\n",
    "    {\"color\": '#777777', \"linestyle\": \"-\"},\n",
    "   # {\"color\": '#56B4E9', \"linestyle\": \"-\"},\n",
    "   # {\"color\": '#009E73', \"linestyle\": \"-\"},\n",
    "    {\"color\": '#E69F00', \"linestyle\": \"-\"},\n",
    "    {\"color\": \"lightgrey\", \"linestyle\": \"--\"}\n",
    "]\n",
    "\n",
    "plt.figure(figsize=(9, 0.10))\n",
    "\n",
    "# Create a dummy plot\n",
    "for style in legend_styles:\n",
    "    plt.plot([], [], color=style[\"color\"], linestyle=style[\"linestyle\"], linewidth=4)\n",
    "\n",
    "# Add the legend\n",
    "plt.legend(\n",
    "    legend_labels,\n",
    "    fontsize=14.25,\n",
    "    loc=\"upper center\",\n",
    "    ncol=len(legend_labels),  # Arrange in a single row\n",
    "    frameon=False\n",
    ")\n",
    "\n",
    "# Remove axes for a clean legend-only figure\n",
    "plt.gca().set_axis_off()\n",
    "\n",
    "# Display or save the legend\n",
    "#plt.show()\n",
    "#plt.savefig('/Users/narod/Library/CloudStorage/GoogleDrive-narod.kebabci@ucdconnect.ie/My Drive/SL PRED/Fig 3/legend_new.png', bbox_inches='tight', dpi=500)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def plot_model_performance(df, data, target, model_early, model_late):#, file_path):\n",
    "#     sns.set_context('paper')\n",
    "#     f, ax = plt.subplots(1, 2, figsize=(9.5, 4.5))  # Adjusted for better visibility\n",
    "\n",
    "#     # Main classifier ROC and PR curves (early)\n",
    "#     roc_display = RocCurveDisplay(\n",
    "#         fpr=model_early['mean_fpr'],\n",
    "#         tpr=model_early['mean_tpr'],\n",
    "#         roc_auc=model_early['mean_auc']\n",
    "#     )\n",
    "#     roc_display.plot(ax=ax[0], color='#009E73', linewidth=1.75, label=f\"{model_early['mean_auc']:.2f}\")\n",
    "\n",
    "#     pr_display = PrecisionRecallDisplay(\n",
    "#         precision=model_early['precision'],\n",
    "#         recall=model_early['recall'],\n",
    "#         average_precision=model_early['mean_aps']\n",
    "#     )\n",
    "#     pr_display.plot(ax=ax[1], color='#009E73', linewidth=1.75, label=f\"{model_early['mean_aps']:.2f}\")\n",
    "\n",
    "#     # Main classifier ROC and PR curves (late)\n",
    "#     roc_display = RocCurveDisplay(\n",
    "#         fpr=model_late['mean_fpr'],\n",
    "#         tpr=model_late['mean_tpr'],\n",
    "#         roc_auc=model_late['mean_auc']\n",
    "#     )\n",
    "#     roc_display.plot(ax=ax[0], alpha=0.8, color='#CC79A7', linewidth=1.75, label=f\"{model_late['mean_auc']:.2f}\")\n",
    "\n",
    "#     pr_display = PrecisionRecallDisplay(\n",
    "#         precision=model_late['precision'],\n",
    "#         recall=model_late['recall'],\n",
    "#         average_precision=model_late['mean_aps']\n",
    "#     )\n",
    "#     pr_display.plot(ax=ax[1], alpha=0.8, color='#CC79A7', linewidth=1.75, label=f\"{model_late['mean_aps']:.2f}\")\n",
    "\n",
    "#     # Additional classifiers\n",
    "#     #metrics = ['prediction_score', 'rMinExp_A1A2', 'min_ranked_A1A2', 'max_seq_id']\n",
    "#     #colors = ['#CC79A7', '#56B4E9', '#009E73', '#E69F00']\n",
    "#     #auc_labels = ['pred_auc', 'gene_expr_auc', 'gene_ess_auc', 'seq_auc']\n",
    "#     #ap_labels = ['pred_ap', 'gene_expr_ap', 'gene_ess_ap', 'seq_ap']\n",
    "\n",
    "#     metrics = ['prediction_score', 'max_seq_id']\n",
    "#     colors = ['#777777', '#E69F00']\n",
    "#     auc_labels = ['pred_auc',  'seq_auc']\n",
    "#     ap_labels = ['pred_ap', 'seq_ap']\n",
    "    \n",
    "#     for metric, color, auc_label, ap_label in zip(metrics, colors, auc_labels, ap_labels):\n",
    "#         fpr, tpr, _ = roc_curve(target, data[metric] if metric != 'min_ranked_A1A2' else -data[metric])\n",
    "#         roc_auc = model_early[auc_label]\n",
    "#         roc_display = RocCurveDisplay(fpr=fpr, tpr=tpr, roc_auc=roc_auc)\n",
    "#         roc_display.plot(ax=ax[0], color=color, linewidth=1.75, label=f\"{roc_auc:.3f}\")\n",
    "\n",
    "#         precision, recall, _ = precision_recall_curve(target, data[metric] if metric != 'min_ranked_A1A2' else -data[metric])\n",
    "#         pr_auc = model_early[ap_label]\n",
    "#         pr_display = PrecisionRecallDisplay(precision=precision, recall=recall, average_precision=pr_auc)\n",
    "#         pr_display.plot(ax=ax[1], color=color, linewidth=1.75, label=f\"{pr_auc:.3f}\")\n",
    "\n",
    "#     # Chance line for ROC curve\n",
    "#     ax[0].plot([0, 1], [0, 1], color=\"lightgrey\", linestyle=\"--\", label=\"0.50\")\n",
    "#     ax[0].set_xlim([-0.025, 1.025])\n",
    "#     ax[0].set_ylim([-0.025, 1.025])\n",
    "#     ax[0].spines.top.set(visible=False)\n",
    "#     ax[0].spines.right.set(visible=False)\n",
    "#     ax[0].tick_params(axis='both', which='major', labelsize=11)\n",
    "#     ax[0].set_xlabel('False Positive Rate', fontsize=12)\n",
    "#     ax[0].set_ylabel('True Positive Rate', fontsize=12)\n",
    "#     ax[0].legend(loc='lower right', fontsize=12, bbox_to_anchor=(1.02, 0))\n",
    "\n",
    "#     # Chance line for PR curve\n",
    "#     no_skill = round(sum(df['SL_new']) / len(df['SL_new']), 3)\n",
    "#     ax[1].plot([0, 1], [no_skill, no_skill], linestyle='--', color='lightgrey', label=f\"{no_skill:.2f}\")\n",
    "#     ax[1].set_xlim([-0.025, 1.025])\n",
    "#     ax[1].set_ylim([-0.025, 1.025])\n",
    "#     ax[1].spines.top.set(visible=False)\n",
    "#     ax[1].spines.right.set(visible=False)\n",
    "#     ax[1].set_xlabel('Recall', fontsize=12)\n",
    "#     ax[1].tick_params(axis='both', which='major', labelsize=11)\n",
    "#     ax[1].set_ylabel('Precision', fontsize=12)\n",
    "#     ax[1].legend(loc='upper right', fontsize=12)\n",
    "\n",
    "#     # Save plot\n",
    "#     plt.tight_layout(h_pad=0.7)\n",
    "#     #plt.savefig(file_path, bbox_inches='tight', dpi=350)\n",
    "#     #plt.close(f)  # Optionally close the figure to save memory if generating many plots\n",
    "\n",
    "# # Example usage:\n",
    "# # plot_model_performance(df, data, target, model_early, model_late, '/Users/narod/Desktop/model_performance.png')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_model_performance(df, data, target, model_early, model_late, file_path):\n",
    "    sns.set_context('paper')\n",
    "    f, ax = plt.subplots(1, 2, figsize=(9.5, 4.5))  # ROC and PR side by side\n",
    "\n",
    "    # --- Early model ---\n",
    "    roc_display = RocCurveDisplay(\n",
    "        fpr=model_early['mean_fpr'],\n",
    "        tpr=model_early['mean_tpr'],\n",
    "        roc_auc=model_early['mean_auc']\n",
    "    )\n",
    "    roc_display.plot(ax=ax[0], color='#009E73', linewidth=1.75, label=f\"{model_early['mean_auc']:.2f}\")\n",
    "\n",
    "    pr_display = PrecisionRecallDisplay(\n",
    "        precision=model_early['precision'],\n",
    "        recall=model_early['recall'],\n",
    "        average_precision=model_early['mean_aps']\n",
    "    )\n",
    "    pr_display.plot(ax=ax[1], color='#009E73', linewidth=1.75, label=f\"{model_early['mean_aps']:.2f}\")\n",
    "\n",
    "    # --- Late model ---\n",
    "    roc_display = RocCurveDisplay(\n",
    "        fpr=model_late['mean_fpr'],\n",
    "        tpr=model_late['mean_tpr'],\n",
    "        roc_auc=model_late['mean_auc']\n",
    "    )\n",
    "    roc_display.plot(ax=ax[0], alpha=0.8, color='#CC79A7', linewidth=1.75, label=f\"{model_late['mean_auc']:.2f}\")\n",
    "\n",
    "    pr_display = PrecisionRecallDisplay(\n",
    "        precision=model_late['precision'],\n",
    "        recall=model_late['recall'],\n",
    "        average_precision=model_late['mean_aps']\n",
    "    )\n",
    "    pr_display.plot(ax=ax[1], alpha=0.8, color='#CC79A7', linewidth=1.75, label=f\"{model_late['mean_aps']:.2f}\")\n",
    "\n",
    "    # --- Additional classifier-like metrics ---\n",
    "    metrics = ['prediction_score', 'min_sequence_identity']\n",
    "    colors = ['#777777', '#E69F00']\n",
    "    auc_labels = ['pred_auc', 'seq_auc']\n",
    "    ap_labels = ['pred_ap', 'seq_ap']\n",
    "\n",
    "    for metric, color, auc_label, ap_label in zip(metrics, colors, auc_labels, ap_labels):\n",
    "        score = data[metric] if metric != 'min_ranked_A1A2' else -data[metric]\n",
    "        mask = ~score.isna()\n",
    "        y_true = target[mask]\n",
    "        y_score = score[mask]\n",
    "\n",
    "        if len(np.unique(y_true)) < 2:\n",
    "            print(f\"Skipping {metric}: not enough label variation after NaN removal.\")\n",
    "            continue\n",
    "\n",
    "        # ROC Curve\n",
    "        fpr, tpr, _ = roc_curve(y_true, y_score)\n",
    "        roc_auc = model_early[auc_label]\n",
    "        roc_display = RocCurveDisplay(fpr=fpr, tpr=tpr, roc_auc=roc_auc)\n",
    "        roc_display.plot(ax=ax[0], color=color, linewidth=1.75, label=f\"{roc_auc:.3f}\")\n",
    "\n",
    "        # PR Curve\n",
    "        precision, recall, _ = precision_recall_curve(y_true, y_score)\n",
    "        pr_auc = model_early[ap_label]\n",
    "        pr_display = PrecisionRecallDisplay(precision=precision, recall=recall, average_precision=pr_auc)\n",
    "        pr_display.plot(ax=ax[1], color=color, linewidth=1.75, label=f\"{pr_auc:.3f}\")\n",
    "\n",
    "    # --- ROC chance line ---\n",
    "    ax[0].plot([0, 1], [0, 1], color=\"lightgrey\", linestyle=\"--\", label=\"0.50\")\n",
    "    ax[0].set_xlim([-0.025, 1.025])\n",
    "    ax[0].set_ylim([-0.025, 1.025])\n",
    "    ax[0].spines.top.set(visible=False)\n",
    "    ax[0].spines.right.set(visible=False)\n",
    "    ax[0].tick_params(axis='both', which='major', labelsize=11)\n",
    "    ax[0].set_xlabel('False Positive Rate', fontsize=12)\n",
    "    ax[0].set_ylabel('True Positive Rate', fontsize=12)\n",
    "    ax[0].legend(loc='lower right', fontsize=12, bbox_to_anchor=(1.02, 0))\n",
    "\n",
    "    # --- PR baseline ---\n",
    "    no_skill = round(sum(df['SL_new']) / len(df['SL_new']), 3)\n",
    "    ax[1].plot([0, 1], [no_skill, no_skill], linestyle='--', color='lightgrey', label=f\"{no_skill:.2f}\")\n",
    "    ax[1].set_xlim([-0.025, 1.025])\n",
    "    ax[1].set_ylim([-0.025, 1.025])\n",
    "    ax[1].spines.top.set(visible=False)\n",
    "    ax[1].spines.right.set(visible=False)\n",
    "    ax[1].set_xlabel('Recall', fontsize=12)\n",
    "    ax[1].set_ylabel('Precision', fontsize=12)\n",
    "    ax[1].tick_params(axis='both', which='major', labelsize=11)\n",
    "    ax[1].legend(loc='upper right', fontsize=12)\n",
    "\n",
    "    plt.tight_layout(h_pad=0.7)\n",
    "    plt.savefig(file_path, bbox_inches='tight', dpi=400)\n",
    "    plt.close(f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_model_performance(integrated_training_df_clean, data_1, target, model_I_early, model_I_late, figure_path_model1)\n",
    "plot_model_performance(integrated_training_df_clean, data_1, target, model_II_early, model_II_late, figure_path_model2)\n",
    "plot_model_performance(integrated_training_df_clean, data_1, target, model_III_early, model_III_late, figure_path_model3)\n",
    "plot_model_performance(integrated_training_df_clean, data_1, target, model_IV_early, model_IV_late, figure_path_model4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_model_performance(integrated_training_df_clean, data_1, target, model_I_early_repeated, model_I_late_repeated, figure_path_model1_repeated)\n",
    "plot_model_performance(integrated_training_df_clean, data_1, target, model_II_early_repeated, model_II_late_repeated, figure_path_model2_repeated)\n",
    "plot_model_performance(integrated_training_df_clean, data_1, target, model_III_early_repeated, model_III_late_repeated, figure_path_model3_repeated)\n",
    "plot_model_performance(integrated_training_df_clean, data_1, target, model_IV_early, model_IV_late, figure_path_model4_repeated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_model_performance(integrated_training_df_clean, data_1, target, model_I_early_reduced, model_I_late_reduced, figure_path_model1_reduced)\n",
    "plot_model_performance(integrated_training_df_clean, data_1, target, model_II_early_reduced, model_II_late_reduced, figure_path_model2_reduced)\n",
    "plot_model_performance(integrated_training_df_clean, data_1, target, model_III_early_reduced, model_III_late_reduced, figure_path_model3_reduced)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Draw ROC curve from each fold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def visualize_roc_auc_folds(results):\n",
    "    tprs = results['tprs']  # List of TPRs for each fold\n",
    "    fprs = results['fprs']  # List of FPRs for each fold\n",
    "    mean_fpr = results['mean_fpr']\n",
    "    aucs = results['aucs']\n",
    "    \n",
    "    plt.figure(figsize=(7, 5))\n",
    "    \n",
    "    # Plot ROC curve for each fold\n",
    "    for i in range(len(tprs)):\n",
    "        plt.plot(fprs[i], tprs[i], lw=1, label=f'ROC fold {i+1} (AUC = {aucs[i]:.2f})')\n",
    "\n",
    "    # Plot mean ROC\n",
    "    mean_tpr = results['mean_tpr']\n",
    "    mean_auc = results['mean_auc']\n",
    "    std_auc = results['std_auc']\n",
    "    \n",
    "    plt.plot(mean_fpr, mean_tpr, color='b', linestyle='--',\n",
    "             label=f'Mean ROC (AUC = {mean_auc:.2f} ± {std_auc:.2f})', lw=2)\n",
    "    \n",
    "    # Plot random guessing line\n",
    "    plt.plot([0, 1], [0, 1], linestyle='--', color='gray', label='Random guessing (AUC = 0.5)')\n",
    "    \n",
    "    # Set plot labels and title\n",
    "    plt.xlim([-0.05, 1.05])\n",
    "    plt.ylim([-0.05, 1.05])\n",
    "    plt.xlabel('False Positive Rate')\n",
    "    plt.ylabel('True Positive Rate')\n",
    "    plt.title('ROC Curve - Cross Validation')\n",
    "    #plt.legend(loc='lower right')\n",
    "    plt.tight_layout()\n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "visualize_roc_auc_folds(model_I_early)\n",
    "#visualize_roc_auc_folds(model_I_early_repeated)\n",
    "#visualize_roc_auc_folds(model_I_early_reduced)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Draw bar plots to visualize the class balance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def class_balance(df, splits, k, title):\n",
    "    percentages_train = []\n",
    "    percentages_test = []\n",
    "    train_sizes = []\n",
    "    test_sizes = []\n",
    "    # Loop through each split and calculate the percentage of True values\n",
    "    for i in range(k):\n",
    "    # Get the subset of the DataFrame based on the split indices\n",
    "        subset1 = df.iloc[splits[i][0]]\n",
    "        subset2 = df.iloc[splits[i][1]]\n",
    "        # Calculate the percentage of True values in the subset\n",
    "        percentage_true_train = (subset1['SL_08'].sum() / len(subset1['SL_08'])) * 100\n",
    "        percentage_true_test = (subset2['SL_08'].sum() / len(subset2['SL_08'])) * 100\n",
    "        # Append the percentage to the list\n",
    "        percentages_train.append(percentage_true_train)\n",
    "        percentages_test.append(percentage_true_test)\n",
    "        train_sizes.append(len(subset1))\n",
    "        test_sizes.append(len(subset2))\n",
    "    width = 0.35\n",
    "    x = np.arange(k)\n",
    "    #define colors\n",
    "    train_color = '#0072B2'\n",
    "    test_color = '#E69F00'\n",
    "    # Create the figure and subplots\n",
    "    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 5))\n",
    "    # Plot 1: Grouped bar chart for percentages\n",
    "    ax1.bar(x - width/2, percentages_train, width, label='Train', color=train_color)\n",
    "    ax1.bar(x + width/2, percentages_test, width, label='Test', color=test_color)\n",
    "    ax1.set_xlabel('Split Number')\n",
    "    ax1.set_ylabel('Percentage of True Values')\n",
    "    ax1.set_title('Percentage of True Values in Train and Test Splits:'+ title)\n",
    "    ax1.set_xticks(x)\n",
    "    ax1.set_xticklabels([f'Split {i}' for i in range(k)])\n",
    "    ax1.bar_label(ax1.containers[0], label_type='edge', fmt='%.2f')\n",
    "    ax1.bar_label(ax1.containers[1], label_type='edge', fmt='%.2f')\n",
    "    ax1.legend(loc='lower center', bbox_to_anchor=(0.5, -0.2), ncol=2)\n",
    "    # Plot 2: Grouped bar chart for number of rows\n",
    "    ax2.bar(x - width/2, train_sizes, width, label='Train Size', color=train_color)\n",
    "    ax2.bar(x + width/2, test_sizes, width, label='Test Size', color=test_color)\n",
    "    ax2.set_xlabel('Split Number')\n",
    "    ax2.set_ylabel('Number of Rows')\n",
    "    ax2.set_title('Number of Rows in Train and Test Subsets:' + title)\n",
    "    ax2.set_xticks(x)\n",
    "    ax2.set_xticklabels([f'Split {i}' for i in range(k)])\n",
    "    ax2.bar_label(ax2.containers[0], label_type='edge')\n",
    "    ax2.bar_label(ax2.containers[1], label_type='edge')\n",
    "    #ax2.legend()\n",
    "    ax2.legend(loc='lower center', bbox_to_anchor=(0.5, -0.2), ncol=2)\n",
    "    # Adjust layout\n",
    "    plt.tight_layout()\n",
    "    # Show the plots\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/Users/narod/Library/CloudStorage/GoogleDrive-narod.kebabci@ucdconnect.ie/My Drive/SL PRED/output/cross_validation/model_I/splits.pkl', 'rb') as f:\n",
    "    splits_I = pickle.load(f)\n",
    "\n",
    "class_balance(integrated_training_df_clean, splits_I, 5, \"Model I\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/Users/narod/Library/CloudStorage/GoogleDrive-narod.kebabci@ucdconnect.ie/My Drive/SL PRED/output/cross_validation/model_I/repeated_splits_I.pkl', 'rb') as f:\n",
    "    repeated_splits_I = pickle.load(f)\n",
    "\n",
    "class_balance(integrated_training_df_clean, repeated_splits_I, 10, \"Model I (repeated)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/Users/narod/Library/CloudStorage/GoogleDrive-narod.kebabci@ucdconnect.ie/My Drive/SL PRED/output/cross_validation/model_I/reduced_splits_I.pkl', 'rb') as f:\n",
    "    reduced_splits_I = pickle.load(f)\n",
    "\n",
    "class_balance(integrated_training_df_clean, reduced_splits_I, 10, \"Model I (reduced)\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Two sample t test \n",
    "\n",
    "ROC AUC from the two different models are identical == null hypothesis\n",
    "\n",
    "if pval < 0.05 reject the null hypothesis, if pval > 0.05 fail to reject the null hypothesis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "roc_auc_values = {\n",
    "    'Model 1': model_I_early_repeated['aucs'],\n",
    "    'Model 2': model_II_early_repeated['aucs'],\n",
    "    'Model 3': model_III_early_repeated['aucs'],\n",
    "    'Model 4': model_IV_early['aucs']\n",
    "}\n",
    "\n",
    "roc_auc_df = pd.DataFrame(roc_auc_values).melt(var_name='Model', value_name='ROC AUC')\n",
    "roc_auc_df['label'] = roc_auc_df['Model'].apply(lambda x: 'First' if x == 'Model 1' else 'Others')\n",
    "\n",
    "# Performing t-tests between Model 1 and other models\n",
    "t_tests_results = {\n",
    "    'Comparison': [],\n",
    "    't_statistic': [],\n",
    "    'p_value': []\n",
    "}\n",
    "\n",
    "for model in ['Model 2', 'Model 3', 'Model 4']:\n",
    "    t_stat, p_value = stats.ttest_ind(roc_auc_values['Model 1'], roc_auc_values[model])\n",
    "    t_tests_results['Comparison'].append(f'Model 1 vs {model}')\n",
    "    t_tests_results['t_statistic'].append(t_stat)\n",
    "    t_tests_results['p_value'].append(p_value)\n",
    "\n",
    "t_tests_df = pd.DataFrame(t_tests_results)\n",
    "\n",
    "# Adding significance column to the t_tests_df based on p-value threshold (e.g., p < 0.05)\n",
    "t_tests_df['Significance'] = t_tests_df['p_value'].apply(lambda p: 'not identical' if p < 0.05 else 'identical')\n",
    "\n",
    "display(t_tests_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize results dictionary\n",
    "t_tests_results_v2 = {\n",
    "    'Comparison': [],\n",
    "    't_statistic': [],\n",
    "    'p_value': []\n",
    "}\n",
    "\n",
    "# Define pairs of models for pairwise comparisons\n",
    "model_pairs = [('Model 2', 'Model 1'), ('Model 3', 'Model 2'), ('Model 2', 'Model 4'), ('Model 4', 'Model 3')]\n",
    "\n",
    "# Perform t-tests for each pair\n",
    "for model_a, model_b in model_pairs:\n",
    "    t_stat, p_value = stats.ttest_ind(roc_auc_values[model_a], roc_auc_values[model_b])\n",
    "    t_tests_results_v2['Comparison'].append(f'{model_a} vs {model_b}')\n",
    "    t_tests_results_v2['t_statistic'].append(t_stat)\n",
    "    t_tests_results_v2['p_value'].append(p_value)\n",
    "\n",
    "# Create DataFrame to store results\n",
    "t_tests_df_v2 = pd.DataFrame(t_tests_results_v2)\n",
    "\n",
    "# Add significance column based on p-value threshold (e.g., p < 0.05)\n",
    "t_tests_df_v2['Significance'] = t_tests_df_v2['p_value'].apply(lambda p: 'not identical' if p < 0.05 else 'identical')\n",
    "\n",
    "# Display the results\n",
    "display(t_tests_df_v2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the function for drawing significance lines on a separate axis\n",
    "def draw_signif_line(ax, x0, x1, y, pval, hd=10):\n",
    "    h = y / hd\n",
    "    if pval < 0.0001: \n",
    "        text = '$p$=%.1e' % pval\n",
    "    else: \n",
    "        text = '$p$=%.4f' % pval\n",
    "    #ax.plot([x0, x0, x1, x1], [y + h * .5, y + 2 * h, y + 2 * h, y + h * .5], lw=1, c='#666')\n",
    "    ax.plot([x0, x0, x1, x1], [y + h, y + 2 * h, y + 2 * h, y + h], lw=1, c='#666')\n",
    "    ax.text((x0 + x1) * .5, y + 2.5 * h, text, ha='center', va='bottom', color='#666', fontsize=9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up the figure with two subplots\n",
    "fig, ax = plt.subplots(2, 1, figsize=(4, 4), gridspec_kw={'height_ratios': [1, 4]}, sharex=True)\n",
    "\n",
    "# Draw significance lines on the top subplot\n",
    "draw_signif_line(ax[0], 0, 1, 0.055, t_tests_df.loc[0, 'p_value'])  # Between model 1 and model 2\n",
    "draw_signif_line(ax[0], 0, 2, 0.2, t_tests_df.loc[1, 'p_value'])  # Between model 1 and model 3\n",
    "draw_signif_line(ax[0], 0, 3, 0.32, t_tests_df.loc[2, 'p_value'])  # Between model 1 and model 4\n",
    "\n",
    "# Customize the top subplot for significance lines\n",
    "ax[0].axis('off')\n",
    "ax[0].set_ylim([0, 0.4])  # Adjust based on space needed for significance lines\n",
    "\n",
    "# Main bar plot in the bottom subplot\n",
    "color_dict = {'First': '#D55E00', 'Others': '#009E73'}\n",
    "#sns.barplot(data=roc_auc_df, x='Model', y='ROC AUC', hue='label', errorbar=\"sd\", dodge=False, ax=ax[1], palette=color_dict)\n",
    "sns.boxplot(x='Model', y='ROC AUC', hue='label', data=roc_auc_df, ax=ax[1], linewidth=0.8, saturation=0.8, \n",
    "            showfliers=False, showmeans=False, palette=color_dict,\n",
    "            medianprops={'color':'black', 'linewidth':0.8},\n",
    "            boxprops={'edgecolor':'black', 'alpha':0.75},)\n",
    "sns.swarmplot(data=roc_auc_df, x='Model', y='ROC AUC', color='black', size=2, dodge=True, ax=ax[1])#, alpha=0.7)\n",
    "\n",
    "# Customize the main plot appearance\n",
    "ax[1].spines.top.set(visible=False)\n",
    "ax[1].spines.right.set(visible=False)\n",
    "ax[1].tick_params(axis='both', which='major', labelsize=11)\n",
    "ax[1].set_xlabel('')\n",
    "ax[1].set_ylabel('ROC AUC', fontsize=11)\n",
    "ax[1].set_ylim([0.5, 1])\n",
    "ax[1].legend().remove()\n",
    "\n",
    "plt.tight_layout(h_pad=0)  # Reduce vertical space between subplots\n",
    "\n",
    "boxplot_path = get_data_path(['SL PRED', 'Fig 3', 't_test'], 'boxplot_roc_v1_newparam.png')\n",
    "plt.savefig(boxplot_path, bbox_inches='tight', dpi=350)\n",
    "\n",
    "#plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up the figure with two subplots\n",
    "fig, ax = plt.subplots(2, 1, figsize=(3.85, 4), gridspec_kw={'height_ratios': [1.5, 3.5]}, sharex=True)\n",
    "\n",
    "# Draw significance lines on the top subplot\n",
    "draw_signif_line(ax[0], 0, 1, 0.085, t_tests_df_v2.loc[0, 'p_value'])  # Between model 2 and model 1\n",
    "draw_signif_line(ax[0], 1, 2, 0.15, t_tests_df_v2.loc[1, 'p_value'])  # Between model 3 and model 2\n",
    "#draw_signif_line(ax[0], 1, 3, 0.22, t_tests_df_v2.loc[2, 'p_value'])  # Between model 4 and model 2\n",
    "draw_signif_line(ax[0], 2, 3, 0.32, t_tests_df_v2.loc[3, 'p_value'])  # Between model 4 and model 3\n",
    "\n",
    "# Customize the top subplot for significance lines\n",
    "ax[0].axis('off')\n",
    "ax[0].set_ylim([0, 0.4])  # Adjust based on space needed for significance lines\n",
    "\n",
    "# Main bar plot in the bottom subplot\n",
    "#color_dict = {'First': '#D55E00', 'Others': '#009E73'}\n",
    "#sns.barplot(data=roc_auc_df, x='Model', y='ROC AUC', hue='label', errorbar=\"sd\", dodge=False, ax=ax[1], palette=color_dict)\n",
    "sns.boxplot(x='Model', y='ROC AUC', data=roc_auc_df, ax=ax[1], linewidth=0.8, saturation=0.8, \n",
    "            showfliers=False, showmeans=False, color='#009E73',\n",
    "            medianprops={'color':'black', 'linewidth':0.8},\n",
    "            boxprops={'edgecolor':'black', 'alpha':0.75},)\n",
    "sns.swarmplot(data=roc_auc_df, x='Model', y='ROC AUC', color='black', size=2, dodge=True, ax=ax[1])#, alpha=0.7)\n",
    "\n",
    "# Customize the main plot appearance\n",
    "ax[1].spines.top.set(visible=False)\n",
    "ax[1].spines.right.set(visible=False)\n",
    "ax[1].tick_params(axis='both', which='major', labelsize=11)\n",
    "ax[1].set_xlabel('')\n",
    "ax[1].set_ylabel('ROC AUC', fontsize=11)\n",
    "ax[1].set_ylim([0.5, 1])\n",
    "#ax[1].legend().remove()\n",
    "\n",
    "plt.tight_layout(h_pad=0)  # Reduce vertical space between subplots\n",
    "\n",
    "boxplot_path = get_data_path(['SL PRED', 'Fig 3', 't_test'], 'boxplot_roc_v2_newparam.png')\n",
    "plt.savefig(boxplot_path, bbox_inches='tight', dpi=350)\n",
    "\n",
    "#plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pr_auc_values = {\n",
    "    'Model 1': model_I_early_repeated['aps'],\n",
    "    'Model 2': model_II_early_repeated['aps'],\n",
    "    'Model 3': model_III_early_repeated['aps'],\n",
    "    'Model 4': model_IV_early['aps']\n",
    "}\n",
    "\n",
    "pr_auc_df = pd.DataFrame(pr_auc_values).melt(var_name='Model', value_name='PR AUC')\n",
    "pr_auc_df['label'] = pr_auc_df['Model'].apply(lambda x: 'First' if x == 'Model 1' else 'Others')\n",
    "\n",
    "# Performing t-tests between Model 1 and other models\n",
    "t_tests_results_pr = {\n",
    "    'Comparison': [],\n",
    "    't_statistic': [],\n",
    "    'p_value': []\n",
    "}\n",
    "\n",
    "for model in ['Model 2', 'Model 3', 'Model 4']:\n",
    "    t_stat_pr, p_value_pr = stats.ttest_ind(pr_auc_values['Model 1'], pr_auc_values[model])\n",
    "    t_tests_results_pr['Comparison'].append(f'Model 1 vs {model}')\n",
    "    t_tests_results_pr['t_statistic'].append(t_stat_pr)\n",
    "    t_tests_results_pr['p_value'].append(p_value_pr)\n",
    "\n",
    "t_tests_df_pr = pd.DataFrame(t_tests_results_pr)\n",
    "\n",
    "# Adding significance column to the t_tests_df based on p-value threshold (e.g., p < 0.05)\n",
    "t_tests_df_pr['Significance'] = t_tests_df_pr['p_value'].apply(lambda p: 'not identical' if p < 0.05 else 'identical')\n",
    "display(t_tests_df_pr)\n",
    "\n",
    "# Performing t-tests between Model 1 and other models\n",
    "t_tests_results_pr_v2 = {\n",
    "    'Comparison': [],\n",
    "    't_statistic': [],\n",
    "    'p_value': []\n",
    "}\n",
    "\n",
    "# Define pairs of models for pairwise comparisons\n",
    "model_pairs = [('Model 1', 'Model 2'), ('Model 2', 'Model 3'), ('Model 2', 'Model 4'), ('Model 3', 'Model 4')]\n",
    "\n",
    "# Perform t-tests for each pair\n",
    "for model_a, model_b in model_pairs:\n",
    "    t_stat_pr, p_value_pr = stats.ttest_ind(pr_auc_values[model_a], pr_auc_values[model_b])\n",
    "    t_tests_results_pr_v2['Comparison'].append(f'{model_a} vs {model_b}')\n",
    "    t_tests_results_pr_v2['t_statistic'].append(t_stat_pr)\n",
    "    t_tests_results_pr_v2['p_value'].append(p_value_pr)\n",
    "\n",
    "# Create DataFrame to store results\n",
    "t_tests_df_pr_v2 = pd.DataFrame(t_tests_results_pr_v2)\n",
    "\n",
    "# Add significance column based on p-value threshold (e.g., p < 0.05)\n",
    "t_tests_df_pr_v2['Significance'] = t_tests_df_pr_v2['p_value'].apply(lambda p: 'not identical' if p < 0.05 else 'identical')\n",
    "\n",
    "# Display the results\n",
    "display(t_tests_df_pr_v2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def draw_signif_line(ax, x0, x1, y, pval, hd=20):\n",
    "    h = y / hd\n",
    "    if pval < 0.0001: \n",
    "        text = '$p$=%.1e' % pval\n",
    "    else: \n",
    "        text = '$p$=%.4f' % pval\n",
    "    ax.plot([x0, x0, x1, x1], [y + h, y + 2 * h, y + 2 * h, y + h], lw=1, c='#666')\n",
    "    ax.text((x0 + x1) * .5, y + 2.2 * h, text, ha='center', va='bottom', color='#666', fontsize=9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_context('paper')\n",
    "fig, ax = plt.subplots(figsize=(4, 4))\n",
    "color_dict = {'First': '#D55E00', 'Others': '#009E73'}\n",
    "#sns.barplot(data=pr_auc_df, x='Model', y='PR AUC', hue='label', errorbar=\"sd\", dodge=False, ax=ax, palette=color_dict)\n",
    "sns.boxplot(x='Model', y='PR AUC', hue='label', data=pr_auc_df, ax=ax, linewidth=0.8, saturation=0.8, \n",
    "            showfliers=False, showmeans=False, palette=color_dict,\n",
    "            medianprops={'color':'black', 'linewidth':0.8},\n",
    "            boxprops={'edgecolor':'black', 'alpha':0.75},)\n",
    "sns.swarmplot(data=pr_auc_df, x='Model', y='PR AUC', color='black', size=2, dodge=True, ax=ax)#, alpha=0.7)\n",
    "\n",
    "# Add significance lines\n",
    "draw_signif_line(ax, 0, 1, 0.41, t_tests_df_pr.loc[0, 'p_value'])\n",
    "draw_signif_line(ax, 0, 2, 0.47, t_tests_df_pr.loc[1, 'p_value'])  # Between model 1 and model 3\n",
    "draw_signif_line(ax, 0, 3, 0.54, t_tests_df_pr.loc[2, 'p_value'])  # Between model 1 and model 4\n",
    "\n",
    "ax.spines.top.set(visible=False)\n",
    "ax.spines.right.set(visible=False)\n",
    "ax.tick_params(axis='both', which='major', labelsize=11)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('PR AUC', fontsize=11)\n",
    "ax.set_ylim([0, 0.6])\n",
    "ax.legend().remove()\n",
    "#ax[0].legend(loc='lower right', fontsize=9, bbox_to_anchor=(1.02, 0))\n",
    "\n",
    "boxplot_path_v2 = get_data_path(['SL PRED', 'Fig 3', 't_test'], 'boxplot_pr_v1_newparam.png')\n",
    "plt.savefig(boxplot_path_v2, bbox_inches='tight', dpi=350)\n",
    "\n",
    "#plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_context('paper')\n",
    "fig, ax = plt.subplots(figsize=(4, 4))\n",
    "#color_dict = {'First': '#D55E00', 'Others': '#009E73'}\n",
    "#sns.barplot(data=pr_auc_df, x='Model', y='PR AUC', hue='label', errorbar=\"sd\", dodge=False, ax=ax, palette=color_dict)\n",
    "sns.boxplot(x='Model', y='PR AUC', data=pr_auc_df, ax=ax, linewidth=0.8, saturation=0.8, \n",
    "            showfliers=False, showmeans=False, color='#009E73',\n",
    "            medianprops={'color':'black', 'linewidth':0.8},\n",
    "            boxprops={'edgecolor':'black', 'alpha':0.75},)\n",
    "sns.swarmplot(data=pr_auc_df, x='Model', y='PR AUC', color='black', size=2, dodge=True, ax=ax)#, alpha=0.7)\n",
    "\n",
    "# Add significance lines\n",
    "draw_signif_line(ax, 0, 1, 0.41, t_tests_df_pr_v2.loc[0, 'p_value'])\n",
    "draw_signif_line(ax, 1, 2, 0.47, t_tests_df_pr_v2.loc[1, 'p_value'])  # Between model 1 and model 3\n",
    "#draw_signif_line(ax, 1, 3, 0.44, t_tests_df_pr_v2.loc[2, 'p_value'])  # Between model 1 and model 4\n",
    "draw_signif_line(ax, 2, 3, 0.54, t_tests_df_pr_v2.loc[3, 'p_value'])  # Between model 1 and model 4\n",
    "\n",
    "ax.spines.top.set(visible=False)\n",
    "ax.spines.right.set(visible=False)\n",
    "ax.tick_params(axis='both', which='major', labelsize=11)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('PR AUC', fontsize=11)\n",
    "ax.set_ylim([0, 0.6])\n",
    "#ax.legend().remove()\n",
    "#ax[0].legend(loc='lower right', fontsize=9, bbox_to_anchor=(1.02, 0))\n",
    "\n",
    "boxplot_path_v2 = get_data_path(['SL PRED', 'Fig 3', 't_test'], 'boxplot_pr_v2_newparam.png')\n",
    "plt.savefig(boxplot_path_v2, bbox_inches='tight', dpi=350)\n",
    "\n",
    "#plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "predictor",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
