{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import modules\n",
    "import os\n",
    "import csv\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pyarrow as pa\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set the base directory for the project\n",
    "cwd = os.getcwd()\n",
    "BASE_DIR = os.path.abspath(os.path.join(cwd, \"..\", \"..\"))\n",
    "\n",
    "# build paths inside the repo\n",
    "get_data_path = lambda folders, fname: os.path.normpath(\n",
    "    os.path.join(BASE_DIR, *folders, fname)\n",
    ")\n",
    "\n",
    "file_path_gene_expression = get_data_path(['output', 'processed_DepMap22Q4'], 'expression_data.csv')\n",
    "file_path_gene_zexpression = get_data_path(['output', 'processed_DepMap22Q4'], 'zexpression_data.csv')\n",
    "\n",
    "dekegel_table8_path = get_data_path(['input', 'other'], 'processed_DeKegel_TableS8.csv')\n",
    "\n",
    "file_path_genenames = get_data_path(['input', 'other'], 'genenames.txt')\n",
    "\n",
    "file_path_ensembl = get_data_path(['input', 'other'], 'mart_export.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HGNC ID</th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>NCBI Gene ID</th>\n",
       "      <th>NCBI Gene ID(supplied by NCBI)</th>\n",
       "      <th>Ensembl gene ID</th>\n",
       "      <th>Ensembl ID(supplied by Ensembl)</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>ensembl_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HGNC:37133</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>503538</td>\n",
       "      <td>503538</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>503538</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>29974</td>\n",
       "      <td>29974</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>29974</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      HGNC ID Approved symbol NCBI Gene ID NCBI Gene ID(supplied by NCBI)  \\\n",
       "0      HGNC:5            A1BG            1                              1   \n",
       "1  HGNC:37133        A1BG-AS1       503538                         503538   \n",
       "2  HGNC:24086            A1CF        29974                          29974   \n",
       "\n",
       "   Ensembl gene ID Ensembl ID(supplied by Ensembl)  entrez_id       ensembl_id  \n",
       "0  ENSG00000121410                 ENSG00000121410          1  ENSG00000121410  \n",
       "1  ENSG00000268895                 ENSG00000268895     503538  ENSG00000268895  \n",
       "2  ENSG00000148584                 ENSG00000148584      29974  ENSG00000148584  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read HGNC\n",
    "id_map_raw = pd.read_table(file_path_genenames, dtype = \"str\")\n",
    "\n",
    "id_map = id_map_raw[['HGNC ID', 'Approved symbol', 'NCBI Gene ID', 'NCBI Gene ID(supplied by NCBI)', 'Ensembl gene ID', 'Ensembl ID(supplied by Ensembl)']]\n",
    "\n",
    "id_map = id_map.copy()\n",
    "id_map['entrez_id'] = id_map['NCBI Gene ID(supplied by NCBI)'].combine_first(id_map['NCBI Gene ID'])\n",
    "id_map['ensembl_id'] = id_map['Ensembl ID(supplied by Ensembl)'].combine_first(id_map['Ensembl gene ID'])\n",
    "\n",
    "id_map_na = id_map.dropna(axis=0, how='any', subset=['entrez_id', 'ensembl_id']).reset_index(drop=True)\n",
    "id_map_na['entrez_id'] = id_map_na['entrez_id'].astype(int)\n",
    "id_map_na[:3]\n",
    "#id_map.hgnc_id = id_map.hgnc_id.apply(lambda x: x.split(':')[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Expression Profiles\n",
    "\n",
    "#### Differential expression analysis\n",
    "- Null hypothesis: the activity of the gene stays the same in two different conditions\n",
    "- Alternative hypothesis: the activity of the gene is different in two different conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>7105</th>\n",
       "      <th>64102</th>\n",
       "      <th>8813</th>\n",
       "      <th>57147</th>\n",
       "      <th>55732</th>\n",
       "      <th>2268</th>\n",
       "      <th>3075</th>\n",
       "      <th>2519</th>\n",
       "      <th>2729</th>\n",
       "      <th>4800</th>\n",
       "      <th>...</th>\n",
       "      <th>8916</th>\n",
       "      <th>29970</th>\n",
       "      <th>255027</th>\n",
       "      <th>883</th>\n",
       "      <th>221468</th>\n",
       "      <th>55486</th>\n",
       "      <th>552900</th>\n",
       "      <th>220074</th>\n",
       "      <th>440519</th>\n",
       "      <th>284391</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACH-001113</th>\n",
       "      <td>4.331992</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.364660</td>\n",
       "      <td>2.792855</td>\n",
       "      <td>4.471187</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>1.226509</td>\n",
       "      <td>3.044394</td>\n",
       "      <td>6.500005</td>\n",
       "      <td>4.739848</td>\n",
       "      <td>...</td>\n",
       "      <td>1.367371</td>\n",
       "      <td>1.201634</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.659925</td>\n",
       "      <td>0.286881</td>\n",
       "      <td>6.397974</td>\n",
       "      <td>4.149747</td>\n",
       "      <td>2.831877</td>\n",
       "      <td>2.153805</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-001289</th>\n",
       "      <td>4.567424</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>7.106641</td>\n",
       "      <td>2.543496</td>\n",
       "      <td>3.504620</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.189034</td>\n",
       "      <td>3.813525</td>\n",
       "      <td>4.221877</td>\n",
       "      <td>3.481557</td>\n",
       "      <td>...</td>\n",
       "      <td>3.053111</td>\n",
       "      <td>6.179710</td>\n",
       "      <td>0.748461</td>\n",
       "      <td>3.019702</td>\n",
       "      <td>1.807355</td>\n",
       "      <td>5.778734</td>\n",
       "      <td>3.134221</td>\n",
       "      <td>3.689299</td>\n",
       "      <td>1.669027</td>\n",
       "      <td>1.485427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-001339</th>\n",
       "      <td>3.150560</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.379118</td>\n",
       "      <td>2.333424</td>\n",
       "      <td>4.228049</td>\n",
       "      <td>0.056584</td>\n",
       "      <td>1.310340</td>\n",
       "      <td>6.687201</td>\n",
       "      <td>3.682573</td>\n",
       "      <td>3.273516</td>\n",
       "      <td>...</td>\n",
       "      <td>2.596935</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.815575</td>\n",
       "      <td>0.622930</td>\n",
       "      <td>6.204180</td>\n",
       "      <td>3.363171</td>\n",
       "      <td>1.895303</td>\n",
       "      <td>0.150560</td>\n",
       "      <td>1.269033</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 19154 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              7105      64102     8813      57147     55732     2268    \\\n",
       "ACH-001113  4.331992  0.000000  7.364660  2.792855  4.471187  0.028569   \n",
       "ACH-001289  4.567424  0.584963  7.106641  2.543496  3.504620  0.000000   \n",
       "ACH-001339  3.150560  0.000000  7.379118  2.333424  4.228049  0.056584   \n",
       "\n",
       "              3075      2519      2729      4800    ...    8916      29970   \\\n",
       "ACH-001113  1.226509  3.044394  6.500005  4.739848  ...  1.367371  1.201634   \n",
       "ACH-001289  0.189034  3.813525  4.221877  3.481557  ...  3.053111  6.179710   \n",
       "ACH-001339  1.310340  6.687201  3.682573  3.273516  ...  2.596935  1.000000   \n",
       "\n",
       "              255027    883       221468    55486     552900    220074  \\\n",
       "ACH-001113  0.000000  1.659925  0.286881  6.397974  4.149747  2.831877   \n",
       "ACH-001289  0.748461  3.019702  1.807355  5.778734  3.134221  3.689299   \n",
       "ACH-001339  0.000000  1.815575  0.622930  6.204180  3.363171  1.895303   \n",
       "\n",
       "              440519    284391  \n",
       "ACH-001113  2.153805  0.000000  \n",
       "ACH-001289  1.669027  1.485427  \n",
       "ACH-001339  0.150560  1.269033  \n",
       "\n",
       "[3 rows x 19154 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_expression = pd.read_csv(file_path_gene_expression, \n",
    "                            index_col=0, low_memory=False)\n",
    "gene_expression = gene_expression.drop(['cell_name'], axis=1)\n",
    "gene_expression.columns = gene_expression.columns.astype(int)\n",
    "gene_expression[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>7105</th>\n",
       "      <th>64102</th>\n",
       "      <th>8813</th>\n",
       "      <th>57147</th>\n",
       "      <th>55732</th>\n",
       "      <th>2268</th>\n",
       "      <th>3075</th>\n",
       "      <th>2519</th>\n",
       "      <th>2729</th>\n",
       "      <th>4800</th>\n",
       "      <th>...</th>\n",
       "      <th>8916</th>\n",
       "      <th>29970</th>\n",
       "      <th>255027</th>\n",
       "      <th>883</th>\n",
       "      <th>221468</th>\n",
       "      <th>55486</th>\n",
       "      <th>552900</th>\n",
       "      <th>220074</th>\n",
       "      <th>440519</th>\n",
       "      <th>284391</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACH-001113</th>\n",
       "      <td>0.588734</td>\n",
       "      <td>-0.202156</td>\n",
       "      <td>1.341424</td>\n",
       "      <td>0.782413</td>\n",
       "      <td>1.015674</td>\n",
       "      <td>-0.333763</td>\n",
       "      <td>-0.419278</td>\n",
       "      <td>-1.154559</td>\n",
       "      <td>1.615601</td>\n",
       "      <td>0.920474</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.225467</td>\n",
       "      <td>-0.143460</td>\n",
       "      <td>-0.692331</td>\n",
       "      <td>-1.157672</td>\n",
       "      <td>-0.770213</td>\n",
       "      <td>1.149857</td>\n",
       "      <td>1.662524</td>\n",
       "      <td>-0.122769</td>\n",
       "      <td>0.979970</td>\n",
       "      <td>-1.078924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-001289</th>\n",
       "      <td>0.731947</td>\n",
       "      <td>1.493293</td>\n",
       "      <td>0.942510</td>\n",
       "      <td>0.325524</td>\n",
       "      <td>-0.214189</td>\n",
       "      <td>-0.356637</td>\n",
       "      <td>-0.882359</td>\n",
       "      <td>-0.730890</td>\n",
       "      <td>-0.360896</td>\n",
       "      <td>-0.750642</td>\n",
       "      <td>...</td>\n",
       "      <td>0.631138</td>\n",
       "      <td>3.312879</td>\n",
       "      <td>0.182974</td>\n",
       "      <td>0.446167</td>\n",
       "      <td>1.850917</td>\n",
       "      <td>0.027580</td>\n",
       "      <td>0.242377</td>\n",
       "      <td>1.015032</td>\n",
       "      <td>0.405184</td>\n",
       "      <td>0.672085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-001339</th>\n",
       "      <td>-0.129934</td>\n",
       "      <td>-0.202156</td>\n",
       "      <td>1.363777</td>\n",
       "      <td>-0.059381</td>\n",
       "      <td>0.706304</td>\n",
       "      <td>-0.311333</td>\n",
       "      <td>-0.381860</td>\n",
       "      <td>0.852051</td>\n",
       "      <td>-0.828794</td>\n",
       "      <td>-1.026939</td>\n",
       "      <td>...</td>\n",
       "      <td>0.128724</td>\n",
       "      <td>-0.283457</td>\n",
       "      <td>-0.692331</td>\n",
       "      <td>-0.974084</td>\n",
       "      <td>-0.190901</td>\n",
       "      <td>0.798635</td>\n",
       "      <td>0.562549</td>\n",
       "      <td>-1.365604</td>\n",
       "      <td>-1.395212</td>\n",
       "      <td>0.417002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 19154 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              7105      64102     8813      57147     55732     2268    \\\n",
       "ACH-001113  0.588734 -0.202156  1.341424  0.782413  1.015674 -0.333763   \n",
       "ACH-001289  0.731947  1.493293  0.942510  0.325524 -0.214189 -0.356637   \n",
       "ACH-001339 -0.129934 -0.202156  1.363777 -0.059381  0.706304 -0.311333   \n",
       "\n",
       "              3075      2519      2729      4800    ...    8916      29970   \\\n",
       "ACH-001113 -0.419278 -1.154559  1.615601  0.920474  ... -1.225467 -0.143460   \n",
       "ACH-001289 -0.882359 -0.730890 -0.360896 -0.750642  ...  0.631138  3.312879   \n",
       "ACH-001339 -0.381860  0.852051 -0.828794 -1.026939  ...  0.128724 -0.283457   \n",
       "\n",
       "              255027    883       221468    55486     552900    220074  \\\n",
       "ACH-001113 -0.692331 -1.157672 -0.770213  1.149857  1.662524 -0.122769   \n",
       "ACH-001289  0.182974  0.446167  1.850917  0.027580  0.242377  1.015032   \n",
       "ACH-001339 -0.692331 -0.974084 -0.190901  0.798635  0.562549 -1.365604   \n",
       "\n",
       "              440519    284391  \n",
       "ACH-001113  0.979970 -1.078924  \n",
       "ACH-001289  0.405184  0.672085  \n",
       "ACH-001339 -1.395212  0.417002  \n",
       "\n",
       "[3 rows x 19154 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zexpression_data = pd.read_csv(file_path_gene_zexpression,\n",
    "                               index_col=0, low_memory=False)\n",
    "zexpression_data = zexpression_data.drop(['cell_name'], axis=1)\n",
    "zexpression_data.columns = zexpression_data.columns.astype(int)\n",
    "zexpression_data[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <th>ENSG00000001036</th>\n",
       "      <th>ENSG00000001084</th>\n",
       "      <th>ENSG00000001167</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000138641</th>\n",
       "      <th>ENSG00000151967</th>\n",
       "      <th>ENSG00000156968</th>\n",
       "      <th>ENSG00000171097</th>\n",
       "      <th>ENSG00000172738</th>\n",
       "      <th>ENSG00000175193</th>\n",
       "      <th>ENSG00000183336</th>\n",
       "      <th>ENSG00000284922</th>\n",
       "      <th>ENSG00000196081</th>\n",
       "      <th>ENSG00000223547</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACH-001113</th>\n",
       "      <td>4.331992</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.364660</td>\n",
       "      <td>2.792855</td>\n",
       "      <td>4.471187</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>1.226509</td>\n",
       "      <td>3.044394</td>\n",
       "      <td>6.500005</td>\n",
       "      <td>4.739848</td>\n",
       "      <td>...</td>\n",
       "      <td>1.367371</td>\n",
       "      <td>1.201634</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.659925</td>\n",
       "      <td>0.286881</td>\n",
       "      <td>6.397974</td>\n",
       "      <td>4.149747</td>\n",
       "      <td>2.831877</td>\n",
       "      <td>2.153805</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-001289</th>\n",
       "      <td>4.567424</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>7.106641</td>\n",
       "      <td>2.543496</td>\n",
       "      <td>3.504620</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.189034</td>\n",
       "      <td>3.813525</td>\n",
       "      <td>4.221877</td>\n",
       "      <td>3.481557</td>\n",
       "      <td>...</td>\n",
       "      <td>3.053111</td>\n",
       "      <td>6.179710</td>\n",
       "      <td>0.748461</td>\n",
       "      <td>3.019702</td>\n",
       "      <td>1.807355</td>\n",
       "      <td>5.778734</td>\n",
       "      <td>3.134221</td>\n",
       "      <td>3.689299</td>\n",
       "      <td>1.669027</td>\n",
       "      <td>1.485427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-001339</th>\n",
       "      <td>3.150560</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.379118</td>\n",
       "      <td>2.333424</td>\n",
       "      <td>4.228049</td>\n",
       "      <td>0.056584</td>\n",
       "      <td>1.310340</td>\n",
       "      <td>6.687201</td>\n",
       "      <td>3.682573</td>\n",
       "      <td>3.273516</td>\n",
       "      <td>...</td>\n",
       "      <td>2.596935</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.815575</td>\n",
       "      <td>0.622930</td>\n",
       "      <td>6.204180</td>\n",
       "      <td>3.363171</td>\n",
       "      <td>1.895303</td>\n",
       "      <td>0.150560</td>\n",
       "      <td>1.269033</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 19154 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ENSG00000000003  ENSG00000000005  ENSG00000000419  \\\n",
       "ACH-001113         4.331992         0.000000         7.364660   \n",
       "ACH-001289         4.567424         0.584963         7.106641   \n",
       "ACH-001339         3.150560         0.000000         7.379118   \n",
       "\n",
       "            ENSG00000000457  ENSG00000000460  ENSG00000000938  \\\n",
       "ACH-001113         2.792855         4.471187         0.028569   \n",
       "ACH-001289         2.543496         3.504620         0.000000   \n",
       "ACH-001339         2.333424         4.228049         0.056584   \n",
       "\n",
       "            ENSG00000000971  ENSG00000001036  ENSG00000001084  \\\n",
       "ACH-001113         1.226509         3.044394         6.500005   \n",
       "ACH-001289         0.189034         3.813525         4.221877   \n",
       "ACH-001339         1.310340         6.687201         3.682573   \n",
       "\n",
       "            ENSG00000001167  ...  ENSG00000138641  ENSG00000151967  \\\n",
       "ACH-001113         4.739848  ...         1.367371         1.201634   \n",
       "ACH-001289         3.481557  ...         3.053111         6.179710   \n",
       "ACH-001339         3.273516  ...         2.596935         1.000000   \n",
       "\n",
       "            ENSG00000156968  ENSG00000171097  ENSG00000172738  \\\n",
       "ACH-001113         0.000000         1.659925         0.286881   \n",
       "ACH-001289         0.748461         3.019702         1.807355   \n",
       "ACH-001339         0.000000         1.815575         0.622930   \n",
       "\n",
       "            ENSG00000175193  ENSG00000183336  ENSG00000284922  \\\n",
       "ACH-001113         6.397974         4.149747         2.831877   \n",
       "ACH-001289         5.778734         3.134221         3.689299   \n",
       "ACH-001339         6.204180         3.363171         1.895303   \n",
       "\n",
       "            ENSG00000196081  ENSG00000223547  \n",
       "ACH-001113         2.153805         0.000000  \n",
       "ACH-001289         1.669027         1.485427  \n",
       "ACH-001339         0.150560         1.269033  \n",
       "\n",
       "[3 rows x 19154 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <th>ENSG00000001036</th>\n",
       "      <th>ENSG00000001084</th>\n",
       "      <th>ENSG00000001167</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000138641</th>\n",
       "      <th>ENSG00000151967</th>\n",
       "      <th>ENSG00000156968</th>\n",
       "      <th>ENSG00000171097</th>\n",
       "      <th>ENSG00000172738</th>\n",
       "      <th>ENSG00000175193</th>\n",
       "      <th>ENSG00000183336</th>\n",
       "      <th>ENSG00000284922</th>\n",
       "      <th>ENSG00000196081</th>\n",
       "      <th>ENSG00000223547</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACH-001113</th>\n",
       "      <td>0.588734</td>\n",
       "      <td>-0.202156</td>\n",
       "      <td>1.341424</td>\n",
       "      <td>0.782413</td>\n",
       "      <td>1.015674</td>\n",
       "      <td>-0.333763</td>\n",
       "      <td>-0.419278</td>\n",
       "      <td>-1.154559</td>\n",
       "      <td>1.615601</td>\n",
       "      <td>0.920474</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.225467</td>\n",
       "      <td>-0.143460</td>\n",
       "      <td>-0.692331</td>\n",
       "      <td>-1.157672</td>\n",
       "      <td>-0.770213</td>\n",
       "      <td>1.149857</td>\n",
       "      <td>1.662524</td>\n",
       "      <td>-0.122769</td>\n",
       "      <td>0.979970</td>\n",
       "      <td>-1.078924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-001289</th>\n",
       "      <td>0.731947</td>\n",
       "      <td>1.493293</td>\n",
       "      <td>0.942510</td>\n",
       "      <td>0.325524</td>\n",
       "      <td>-0.214189</td>\n",
       "      <td>-0.356637</td>\n",
       "      <td>-0.882359</td>\n",
       "      <td>-0.730890</td>\n",
       "      <td>-0.360896</td>\n",
       "      <td>-0.750642</td>\n",
       "      <td>...</td>\n",
       "      <td>0.631138</td>\n",
       "      <td>3.312879</td>\n",
       "      <td>0.182974</td>\n",
       "      <td>0.446167</td>\n",
       "      <td>1.850917</td>\n",
       "      <td>0.027580</td>\n",
       "      <td>0.242377</td>\n",
       "      <td>1.015032</td>\n",
       "      <td>0.405184</td>\n",
       "      <td>0.672085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-001339</th>\n",
       "      <td>-0.129934</td>\n",
       "      <td>-0.202156</td>\n",
       "      <td>1.363777</td>\n",
       "      <td>-0.059381</td>\n",
       "      <td>0.706304</td>\n",
       "      <td>-0.311333</td>\n",
       "      <td>-0.381860</td>\n",
       "      <td>0.852051</td>\n",
       "      <td>-0.828794</td>\n",
       "      <td>-1.026939</td>\n",
       "      <td>...</td>\n",
       "      <td>0.128724</td>\n",
       "      <td>-0.283457</td>\n",
       "      <td>-0.692331</td>\n",
       "      <td>-0.974084</td>\n",
       "      <td>-0.190901</td>\n",
       "      <td>0.798635</td>\n",
       "      <td>0.562549</td>\n",
       "      <td>-1.365604</td>\n",
       "      <td>-1.395212</td>\n",
       "      <td>0.417002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 19154 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ENSG00000000003  ENSG00000000005  ENSG00000000419  \\\n",
       "ACH-001113         0.588734        -0.202156         1.341424   \n",
       "ACH-001289         0.731947         1.493293         0.942510   \n",
       "ACH-001339        -0.129934        -0.202156         1.363777   \n",
       "\n",
       "            ENSG00000000457  ENSG00000000460  ENSG00000000938  \\\n",
       "ACH-001113         0.782413         1.015674        -0.333763   \n",
       "ACH-001289         0.325524        -0.214189        -0.356637   \n",
       "ACH-001339        -0.059381         0.706304        -0.311333   \n",
       "\n",
       "            ENSG00000000971  ENSG00000001036  ENSG00000001084  \\\n",
       "ACH-001113        -0.419278        -1.154559         1.615601   \n",
       "ACH-001289        -0.882359        -0.730890        -0.360896   \n",
       "ACH-001339        -0.381860         0.852051        -0.828794   \n",
       "\n",
       "            ENSG00000001167  ...  ENSG00000138641  ENSG00000151967  \\\n",
       "ACH-001113         0.920474  ...        -1.225467        -0.143460   \n",
       "ACH-001289        -0.750642  ...         0.631138         3.312879   \n",
       "ACH-001339        -1.026939  ...         0.128724        -0.283457   \n",
       "\n",
       "            ENSG00000156968  ENSG00000171097  ENSG00000172738  \\\n",
       "ACH-001113        -0.692331        -1.157672        -0.770213   \n",
       "ACH-001289         0.182974         0.446167         1.850917   \n",
       "ACH-001339        -0.692331        -0.974084        -0.190901   \n",
       "\n",
       "            ENSG00000175193  ENSG00000183336  ENSG00000284922  \\\n",
       "ACH-001113         1.149857         1.662524        -0.122769   \n",
       "ACH-001289         0.027580         0.242377         1.015032   \n",
       "ACH-001339         0.798635         0.562549        -1.365604   \n",
       "\n",
       "            ENSG00000196081  ENSG00000223547  \n",
       "ACH-001113         0.979970        -1.078924  \n",
       "ACH-001289         0.405184         0.672085  \n",
       "ACH-001339        -1.395212         0.417002  \n",
       "\n",
       "[3 rows x 19154 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "entrez_id_to_ensembl = dict(zip(id_map_na.entrez_id, id_map_na.ensembl_id))\n",
    "\n",
    "gene_expression = gene_expression.rename(columns=entrez_id_to_ensembl) #replace the entrez_id with ensembl_id\n",
    "display(gene_expression[:3])\n",
    "\n",
    "zexpression_data = zexpression_data.rename(columns=entrez_id_to_ensembl) #replace the entrez_id with ensembl_id\n",
    "display(zexpression_data[:3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# map lineage to gene expression file\n",
    "#sample_info = pd.read_csv(\"/Users/narod/Documents/GitHub/SL_prediction/ccle_files/sample_info.csv\", sep=',', usecols=['DepMap_ID', 'lineage'], low_memory=False)  \n",
    "#lineage_dict = dict(zip(sample_info.DepMap_ID, sample_info.lineage))\n",
    "#gene_expression.insert(0, 'lineage', gene_expression.index.map(lineage_dict))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Paralog Pairs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "paralog_pairs_dict = {}\n",
    "list_of_paralogs = set()\n",
    "#paralog_pairs = []\n",
    "with open(dekegel_table8_path, \"r\") as f :\n",
    "    reader = csv.DictReader(f)\n",
    "    for r in reader :\n",
    "        sorted_gene_pair = r['genepair']\n",
    "        a1 = r['A1_ensembl_new']\n",
    "        a2 = r['A2_ensembl_new']\n",
    "        if a1 in gene_expression.columns and a2 in gene_expression.columns:\n",
    "            #paralog_pairs_dict[sorted_gene_pair] = (a1, a2)\n",
    "            paralog_pairs_dict[(a1, a2)] = sorted_gene_pair\n",
    "            list_of_paralogs.update((a1, a2))\n",
    "            #paralog_pairs.append(set([a1, a2]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#paralog_pairs = []\n",
    "#for key in paralog_pairs_dict.keys():\n",
    "#    if key[0] != key[1]:\n",
    "#        paralog_pairs.append([key[0], key[1]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "paralog_pairs = pd.Series(list(paralog_pairs_dict.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('ENSG00000080503', 'ENSG00000127616')"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paralog_pairs[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#paralog_pairs_dict[tuple(paralog_pairs[0])]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Protein-coding paralog pairs w/ their corresponding GO term annotations\n",
    "\n",
    "Ensembl Filters:\n",
    "\n",
    "- Paralogous Human Genes: Only\n",
    "- Gene type: protein_coding\n",
    "\n",
    "Attributes:\n",
    "\n",
    "- Gene Stable ID\n",
    "- GO Term Accession\n",
    "- GO Term Name\n",
    "- GO Domain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>ensembl_transcript_id</th>\n",
       "      <th>GO_accession</th>\n",
       "      <th>GO_name</th>\n",
       "      <th>GO_domain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0008137</td>\n",
       "      <td>NADH dehydrogenase (ubiquinone) activity</td>\n",
       "      <td>molecular_function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0006120</td>\n",
       "      <td>mitochondrial electron transport, NADH to ubiq...</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0016020</td>\n",
       "      <td>membrane</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0016491</td>\n",
       "      <td>oxidoreductase activity</td>\n",
       "      <td>molecular_function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0005739</td>\n",
       "      <td>mitochondrion</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ensembl_id ensembl_transcript_id GO_accession  \\\n",
       "0  ENSG00000198763       ENST00000361453   GO:0008137   \n",
       "1  ENSG00000198763       ENST00000361453   GO:0006120   \n",
       "2  ENSG00000198763       ENST00000361453   GO:0016020   \n",
       "3  ENSG00000198763       ENST00000361453   GO:0016491   \n",
       "4  ENSG00000198763       ENST00000361453   GO:0005739   \n",
       "\n",
       "                                             GO_name           GO_domain  \n",
       "0           NADH dehydrogenase (ubiquinone) activity  molecular_function  \n",
       "1  mitochondrial electron transport, NADH to ubiq...  biological_process  \n",
       "2                                           membrane  cellular_component  \n",
       "3                            oxidoreductase activity  molecular_function  \n",
       "4                                      mitochondrion  cellular_component  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read the ensemble file\n",
    "raw_data = pd.read_table(file_path_ensembl, sep=',', dtype='str')\n",
    "raw_data = raw_data.rename(columns={'Gene stable ID':'ensembl_id', 'Transcript stable ID':'ensembl_transcript_id', \n",
    "                                    'GO term accession': 'GO_accession', 'GO term name':'GO_name',\n",
    "                                    'GO domain':'GO_domain'})\n",
    "raw_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# of genes w/ their corresponding GO term annotations (BP): 16207 \n",
      " # of unique GO term accession (BP): 11911\n",
      "# of genes w/ their corresponding GO term annotations (CC): 16832 \n",
      " # of unique GO term accession (CC): 1662\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>ensembl_transcript_id</th>\n",
       "      <th>GO_accession</th>\n",
       "      <th>GO_name</th>\n",
       "      <th>GO_domain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0006120</td>\n",
       "      <td>mitochondrial electron transport, NADH to ubiq...</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0032981</td>\n",
       "      <td>mitochondrial respiratory chain complex I asse...</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0006120</td>\n",
       "      <td>mitochondrial electron transport, NADH to ubiq...</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0006120</td>\n",
       "      <td>mitochondrial electron transport, NADH to ubiq...</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0006120</td>\n",
       "      <td>mitochondrial electron transport, NADH to ubiq...</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ensembl_id ensembl_transcript_id GO_accession  \\\n",
       "0  ENSG00000198763       ENST00000361453   GO:0006120   \n",
       "1  ENSG00000198763       ENST00000361453   GO:0032981   \n",
       "2  ENSG00000198763       ENST00000361453   GO:0006120   \n",
       "3  ENSG00000198763       ENST00000361453   GO:0006120   \n",
       "4  ENSG00000198763       ENST00000361453   GO:0006120   \n",
       "\n",
       "                                             GO_name           GO_domain  \n",
       "0  mitochondrial electron transport, NADH to ubiq...  biological_process  \n",
       "1  mitochondrial respiratory chain complex I asse...  biological_process  \n",
       "2  mitochondrial electron transport, NADH to ubiq...  biological_process  \n",
       "3  mitochondrial electron transport, NADH to ubiq...  biological_process  \n",
       "4  mitochondrial electron transport, NADH to ubiq...  biological_process  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>ensembl_transcript_id</th>\n",
       "      <th>GO_accession</th>\n",
       "      <th>GO_name</th>\n",
       "      <th>GO_domain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0016020</td>\n",
       "      <td>membrane</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0005739</td>\n",
       "      <td>mitochondrion</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0005743</td>\n",
       "      <td>mitochondrial inner membrane</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0070469</td>\n",
       "      <td>respirasome</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0005747</td>\n",
       "      <td>mitochondrial respiratory chain complex I</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ensembl_id ensembl_transcript_id GO_accession  \\\n",
       "0  ENSG00000198763       ENST00000361453   GO:0016020   \n",
       "1  ENSG00000198763       ENST00000361453   GO:0005739   \n",
       "2  ENSG00000198763       ENST00000361453   GO:0005743   \n",
       "3  ENSG00000198763       ENST00000361453   GO:0070469   \n",
       "4  ENSG00000198763       ENST00000361453   GO:0005747   \n",
       "\n",
       "                                     GO_name           GO_domain  \n",
       "0                                   membrane  cellular_component  \n",
       "1                              mitochondrion  cellular_component  \n",
       "2               mitochondrial inner membrane  cellular_component  \n",
       "3                                respirasome  cellular_component  \n",
       "4  mitochondrial respiratory chain complex I  cellular_component  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "protein_coding_genes = raw_data.loc[raw_data['ensembl_id'].isin(id_map_na.ensembl_id)] # filter out the protein coding genes only\n",
    "filtered_protein_coding_genes = protein_coding_genes.loc[~protein_coding_genes['GO_accession'].isna()].reset_index(drop=True) #remove the ones without GO_accession \n",
    "filtered_BP_genes = filtered_protein_coding_genes.loc[filtered_protein_coding_genes['GO_domain'] == 'biological_process',].reset_index(drop=True) #filter the biological process domain\n",
    "filtered_CC_genes = filtered_protein_coding_genes.loc[filtered_protein_coding_genes['GO_domain'] == 'cellular_component',].reset_index(drop=True) #filter the cellular component domain\n",
    "\n",
    "print(f'# of genes w/ their corresponding GO term annotations (BP): {filtered_BP_genes.ensembl_id.nunique()} \\n',\n",
    "      f'# of unique GO term accession (BP): {filtered_BP_genes.GO_accession.nunique()}')\n",
    "\n",
    "print(f'# of genes w/ their corresponding GO term annotations (CC): {filtered_CC_genes.ensembl_id.nunique()} \\n',\n",
    "      f'# of unique GO term accession (CC): {filtered_CC_genes.GO_accession.nunique()}')\n",
    "\n",
    "display(filtered_BP_genes.head())\n",
    "display(filtered_CC_genes.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# of paralogs w/ their corresponding GO term annotations (BP): 12012 \n",
      " # of unique GO term accession (BP): 11314\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>ensembl_transcript_id</th>\n",
       "      <th>GO_accession</th>\n",
       "      <th>GO_name</th>\n",
       "      <th>GO_domain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>GO:0006511</td>\n",
       "      <td>ubiquitin-dependent protein catabolic process</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>GO:0016579</td>\n",
       "      <td>protein deubiquitination</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>GO:0006508</td>\n",
       "      <td>proteolysis</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>GO:0016579</td>\n",
       "      <td>protein deubiquitination</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>GO:0016477</td>\n",
       "      <td>cell migration</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ensembl_id ensembl_transcript_id GO_accession  \\\n",
       "0  ENSG00000114374       ENST00000651177   GO:0006511   \n",
       "1  ENSG00000114374       ENST00000651177   GO:0016579   \n",
       "2  ENSG00000114374       ENST00000651177   GO:0006508   \n",
       "3  ENSG00000114374       ENST00000651177   GO:0016579   \n",
       "4  ENSG00000114374       ENST00000651177   GO:0016477   \n",
       "\n",
       "                                         GO_name           GO_domain  \n",
       "0  ubiquitin-dependent protein catabolic process  biological_process  \n",
       "1                       protein deubiquitination  biological_process  \n",
       "2                                    proteolysis  biological_process  \n",
       "3                       protein deubiquitination  biological_process  \n",
       "4                                 cell migration  biological_process  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# of paralogs w/ their corresponding GO term annotations (CC): 12493 \n",
      " # of unique GO term accession (CC): 1526\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>ensembl_transcript_id</th>\n",
       "      <th>GO_accession</th>\n",
       "      <th>GO_name</th>\n",
       "      <th>GO_domain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>GO:0005737</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>GO:0005737</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>GO:0005634</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>GO:0005829</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>ENST00000338981</td>\n",
       "      <td>GO:0005737</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ensembl_id ensembl_transcript_id GO_accession    GO_name  \\\n",
       "0  ENSG00000114374       ENST00000651177   GO:0005737  cytoplasm   \n",
       "1  ENSG00000114374       ENST00000651177   GO:0005737  cytoplasm   \n",
       "2  ENSG00000114374       ENST00000651177   GO:0005634    nucleus   \n",
       "3  ENSG00000114374       ENST00000651177   GO:0005829    cytosol   \n",
       "4  ENSG00000114374       ENST00000338981   GO:0005737  cytoplasm   \n",
       "\n",
       "            GO_domain  \n",
       "0  cellular_component  \n",
       "1  cellular_component  \n",
       "2  cellular_component  \n",
       "3  cellular_component  \n",
       "4  cellular_component  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# filter out the paralogs only\n",
    "filtered_BP_paralogs = filtered_BP_genes.loc[filtered_BP_genes['ensembl_id'].isin(list_of_paralogs)].reset_index(drop=True) \n",
    "\n",
    "print(f'# of paralogs w/ their corresponding GO term annotations (BP): {filtered_BP_paralogs.ensembl_id.nunique()} \\n',\n",
    "      f'# of unique GO term accession (BP): {filtered_BP_paralogs.GO_accession.nunique()}')\n",
    "\n",
    "display(filtered_BP_paralogs.head())\n",
    "\n",
    "filtered_CC_paralogs = filtered_CC_genes.loc[filtered_CC_genes['ensembl_id'].isin(list_of_paralogs)].reset_index(drop=True)\n",
    "\n",
    "print(f'# of paralogs w/ their corresponding GO term annotations (CC): {filtered_CC_paralogs.ensembl_id.nunique()} \\n',\n",
    "      f'# of unique GO term accession (CC): {filtered_CC_paralogs.GO_accession.nunique()}')\n",
    "\n",
    "display(filtered_CC_paralogs.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO_accession</th>\n",
       "      <th>GO_name</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000002</td>\n",
       "      <td>mitochondrial genome maintenance</td>\n",
       "      <td>{ENSG00000115204, ENSG00000114120, ENSG0000011...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000003</td>\n",
       "      <td>reproduction</td>\n",
       "      <td>{ENSG00000189409}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000012</td>\n",
       "      <td>single strand break repair</td>\n",
       "      <td>{ENSG00000225830, ENSG00000096717, ENSG0000013...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0000017</td>\n",
       "      <td>alpha-glucoside transport</td>\n",
       "      <td>{ENSG00000100170, ENSG00000140675}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0000018</td>\n",
       "      <td>regulation of DNA recombination</td>\n",
       "      <td>{ENSG00000114030, ENSG00000182481}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  GO_accession                           GO_name  \\\n",
       "0   GO:0000002  mitochondrial genome maintenance   \n",
       "1   GO:0000003                      reproduction   \n",
       "2   GO:0000012        single strand break repair   \n",
       "3   GO:0000017         alpha-glucoside transport   \n",
       "4   GO:0000018   regulation of DNA recombination   \n",
       "\n",
       "                                               genes  \n",
       "0  {ENSG00000115204, ENSG00000114120, ENSG0000011...  \n",
       "1                                  {ENSG00000189409}  \n",
       "2  {ENSG00000225830, ENSG00000096717, ENSG0000013...  \n",
       "3                 {ENSG00000100170, ENSG00000140675}  \n",
       "4                 {ENSG00000114030, ENSG00000182481}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO_accession</th>\n",
       "      <th>GO_name</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000015</td>\n",
       "      <td>phosphopyruvate hydratase complex</td>\n",
       "      <td>{ENSG00000108515, ENSG00000111674, ENSG0000007...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000118</td>\n",
       "      <td>histone deacetylase complex</td>\n",
       "      <td>{ENSG00000162227, ENSG00000163517, ENSG0000016...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000123</td>\n",
       "      <td>histone acetyltransferase complex</td>\n",
       "      <td>{ENSG00000156983, ENSG00000043143, ENSG0000017...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0000124</td>\n",
       "      <td>SAGA complex</td>\n",
       "      <td>{ENSG00000162227, ENSG00000108773, ENSG0000027...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0000137</td>\n",
       "      <td>Golgi cis cisterna</td>\n",
       "      <td>{ENSG00000116209, ENSG00000175265, ENSG0000016...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  GO_accession                            GO_name  \\\n",
       "0   GO:0000015  phosphopyruvate hydratase complex   \n",
       "1   GO:0000118        histone deacetylase complex   \n",
       "2   GO:0000123  histone acetyltransferase complex   \n",
       "3   GO:0000124                       SAGA complex   \n",
       "4   GO:0000137                 Golgi cis cisterna   \n",
       "\n",
       "                                               genes  \n",
       "0  {ENSG00000108515, ENSG00000111674, ENSG0000007...  \n",
       "1  {ENSG00000162227, ENSG00000163517, ENSG0000016...  \n",
       "2  {ENSG00000156983, ENSG00000043143, ENSG0000017...  \n",
       "3  {ENSG00000162227, ENSG00000108773, ENSG0000027...  \n",
       "4  {ENSG00000116209, ENSG00000175265, ENSG0000016...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# group by GO_accession and GO_name\n",
    "sum_bp_paralogs = filtered_BP_paralogs.groupby(['GO_accession', 'GO_name'])['ensembl_id'].apply(set).reset_index(name=\"genes\")\n",
    "display(sum_bp_paralogs.head())\n",
    "\n",
    "sum_cc_paralogs = filtered_CC_paralogs.groupby(['GO_accession', 'GO_name'])['ensembl_id'].apply(set).reset_index(name=\"genes\")\n",
    "display(sum_cc_paralogs.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO_accession</th>\n",
       "      <th>GO_name</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000002</td>\n",
       "      <td>mitochondrial genome maintenance</td>\n",
       "      <td>{ENSG00000115204, ENSG00000114120, ENSG0000011...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000038</td>\n",
       "      <td>very long-chain fatty acid metabolic process</td>\n",
       "      <td>{ENSG00000113396, ENSG00000184227, ENSG0000011...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000045</td>\n",
       "      <td>autophagosome assembly</td>\n",
       "      <td>{ENSG00000138069, ENSG00000174437, ENSG0000013...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0000050</td>\n",
       "      <td>urea cycle</td>\n",
       "      <td>{ENSG00000118520, ENSG00000021826, ENSG0000024...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0000070</td>\n",
       "      <td>mitotic sister chromatid segregation</td>\n",
       "      <td>{ENSG00000072501, ENSG00000186185, ENSG0000012...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  GO_accession                                       GO_name  \\\n",
       "0   GO:0000002              mitochondrial genome maintenance   \n",
       "1   GO:0000038  very long-chain fatty acid metabolic process   \n",
       "2   GO:0000045                        autophagosome assembly   \n",
       "3   GO:0000050                                    urea cycle   \n",
       "4   GO:0000070          mitotic sister chromatid segregation   \n",
       "\n",
       "                                               genes  \n",
       "0  {ENSG00000115204, ENSG00000114120, ENSG0000011...  \n",
       "1  {ENSG00000113396, ENSG00000184227, ENSG0000011...  \n",
       "2  {ENSG00000138069, ENSG00000174437, ENSG0000013...  \n",
       "3  {ENSG00000118520, ENSG00000021826, ENSG0000024...  \n",
       "4  {ENSG00000072501, ENSG00000186185, ENSG0000012...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# filter biological processes - include processes with 5 to 100 expresses genes\n",
    "\n",
    "sum_bp_paralogs['sm_biological_process'] = sum_bp_paralogs['genes'].apply(lambda x: 100 > len(x) > 5)\n",
    "sum_bp_paralogs_df = sum_bp_paralogs[sum_bp_paralogs['sm_biological_process'] == True].reset_index(drop=True).drop(['sm_biological_process'], axis=1)\n",
    "display(sum_bp_paralogs_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO_accession</th>\n",
       "      <th>GO_name</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000118</td>\n",
       "      <td>histone deacetylase complex</td>\n",
       "      <td>{ENSG00000162227, ENSG00000163517, ENSG0000016...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000123</td>\n",
       "      <td>histone acetyltransferase complex</td>\n",
       "      <td>{ENSG00000156983, ENSG00000043143, ENSG0000017...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000124</td>\n",
       "      <td>SAGA complex</td>\n",
       "      <td>{ENSG00000162227, ENSG00000108773, ENSG0000027...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0000137</td>\n",
       "      <td>Golgi cis cisterna</td>\n",
       "      <td>{ENSG00000116209, ENSG00000175265, ENSG0000016...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0000138</td>\n",
       "      <td>Golgi trans cisterna</td>\n",
       "      <td>{ENSG00000116209, ENSG00000130733, ENSG0000008...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  GO_accession                            GO_name  \\\n",
       "0   GO:0000118        histone deacetylase complex   \n",
       "1   GO:0000123  histone acetyltransferase complex   \n",
       "2   GO:0000124                       SAGA complex   \n",
       "3   GO:0000137                 Golgi cis cisterna   \n",
       "4   GO:0000138               Golgi trans cisterna   \n",
       "\n",
       "                                               genes  \n",
       "0  {ENSG00000162227, ENSG00000163517, ENSG0000016...  \n",
       "1  {ENSG00000156983, ENSG00000043143, ENSG0000017...  \n",
       "2  {ENSG00000162227, ENSG00000108773, ENSG0000027...  \n",
       "3  {ENSG00000116209, ENSG00000175265, ENSG0000016...  \n",
       "4  {ENSG00000116209, ENSG00000130733, ENSG0000008...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# filter cellular component - include processes with 5 to 100 expresses genes\n",
    "\n",
    "sum_cc_paralogs['sm_cellular_component'] = sum_cc_paralogs['genes'].apply(lambda x: 100 > len(x) > 5)\n",
    "sum_cc_paralogs_df = sum_cc_paralogs[sum_cc_paralogs['sm_cellular_component'] == True].reset_index(drop=True).drop(['sm_cellular_component'], axis=1)\n",
    "display(sum_cc_paralogs_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for i in range(sum_cc_paralogs_df.shape[0]):\n",
    "#     if set(('ENSG00000138758', 'ENSG00000125354')).issubset(sum_cc_paralogs_df.genes[i]):\n",
    "#         print(sum_cc_paralogs_df.GO_accession[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def precompute_go_effects(go_df, gene_expression):\n",
    "    go_to_mean_effect = {}\n",
    "    valid_go_df = []\n",
    "\n",
    "    for _, row in go_df.iterrows():\n",
    "        go_id = row['GO_accession']\n",
    "        gene_set = list(row['genes'])\n",
    "        genes_in_effect = [g for g in gene_set if g in gene_expression.columns]\n",
    "        if len(genes_in_effect) >= 2:\n",
    "            go_to_mean_effect[go_id] = gene_expression[genes_in_effect].mean(axis=1)\n",
    "            valid_go_df.append(row)\n",
    "\n",
    "    return go_to_mean_effect, pd.DataFrame(valid_go_df)\n",
    "\n",
    "def build_gene_to_go_map(go_df):\n",
    "    gene_to_go = defaultdict(set)\n",
    "    for _, row in go_df.iterrows():\n",
    "        go_id = row['GO_accession']\n",
    "        for gene in row['genes']:\n",
    "            gene_to_go[gene].add(go_id)\n",
    "    return gene_to_go\n",
    "\n",
    "def process_gene_pair_optimized(pair, go_df, go_to_mean_effect, gene_to_go):\n",
    "    a1, a2 = pair\n",
    "    shared_go_ids = gene_to_go.get(a1, set()).intersection(gene_to_go.get(a2, set()))\n",
    "\n",
    "    if not shared_go_ids:\n",
    "        return pd.DataFrame(), pd.DataFrame()\n",
    "\n",
    "    go_processes_for_pair = []\n",
    "    go_name_list = []\n",
    "    len_of_go = []\n",
    "    pair_expression_df = pd.DataFrame()\n",
    "\n",
    "    for go_id in shared_go_ids:\n",
    "        row = go_df[go_df['GO_accession'] == go_id].iloc[0]\n",
    "        gene_set = row['genes']\n",
    "        go_processes_for_pair.append(go_id)\n",
    "        go_name_list.append(row['GO_name'])\n",
    "        len_of_go.append(len(gene_set))\n",
    "        pair_expression_df[go_id] = go_to_mean_effect[go_id]\n",
    "\n",
    "    pair_df = pd.DataFrame({\n",
    "        'paralog_pairs': [pair] * len(go_processes_for_pair),\n",
    "        'GO_accession': go_processes_for_pair,\n",
    "        'GO_name': go_name_list,\n",
    "        'len_of_go': len_of_go\n",
    "    }).sort_values(by='len_of_go', ascending=False).reset_index(drop=True)\n",
    "\n",
    "    return pair_df, pair_expression_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def summarize_GO_expression_scores(pair_df, pair_expression_df, pair, paralog_pairs_dict=None):\n",
    "    if pair_df.empty or pair_expression_df.empty:\n",
    "        return None\n",
    "\n",
    "    result = {}\n",
    "\n",
    "    # --- smallest_GO_accession (per cell line) ---\n",
    "    min_len = pair_df['len_of_go'].min()\n",
    "    smallest_df = pair_df[pair_df['len_of_go'] == min_len]\n",
    "    smallest_go_ids = smallest_df['GO_accession'].tolist()\n",
    "    smallest_subset = pair_expression_df[smallest_go_ids]\n",
    "\n",
    "    # For each cell line, get the GO with the highest expression among the smallest GOs\n",
    "    result['smallest_GO_accession'] = smallest_subset.idxmax(axis=1)\n",
    "    result['smallest_gene_expression'] = smallest_subset.max(axis=1)\n",
    "\n",
    "    # --- min/max GO by expression (per cell line) ---\n",
    "    result['min_GO_accession'] = pair_expression_df.idxmin(axis=1)\n",
    "    result['min_gene_expression'] = pair_expression_df.min(axis=1)\n",
    "\n",
    "    result['max_GO_accession'] = pair_expression_df.idxmax(axis=1)\n",
    "    result['max_gene_expression'] = pair_expression_df.max(axis=1)\n",
    "\n",
    "    # --- paralog pair label ---\n",
    "    if paralog_pairs_dict is not None:\n",
    "        result['paralog_pair'] = paralog_pairs_dict.get(pair, '|'.join(pair))\n",
    "    else:\n",
    "        result['paralog_pair'] = '|'.join(pair)\n",
    "\n",
    "    # --- return as DataFrame ---\n",
    "    return pd.DataFrame({\n",
    "        'cell_line': pair_expression_df.index,\n",
    "        'paralog_pair': result['paralog_pair'],\n",
    "        'smallest_GO_accession': result['smallest_GO_accession'].values,\n",
    "        'smallest_gene_expression': result['smallest_gene_expression'].values,\n",
    "        'min_GO_accession': result['min_GO_accession'].values,\n",
    "        'min_gene_expression': result['min_gene_expression'].values,\n",
    "        'max_GO_accession': result['max_GO_accession'].values,\n",
    "        'max_gene_expression': result['max_gene_expression'].values,\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 1: Precompute mean gene expression per GO term\n",
    "go_to_mean_expression, filtered_go_df_expr = precompute_go_effects(\n",
    "    sum_bp_paralogs_df, gene_expression\n",
    ")\n",
    "\n",
    "# Step 2: Build gene → GO term mapping\n",
    "gene_to_go_expr = build_gene_to_go_map(filtered_go_df_expr)\n",
    "\n",
    "# Step 4: Run loop over paralog pairs\n",
    "expression_summaries = []\n",
    "\n",
    "for pair in paralog_pairs:\n",
    "    pair_df, pair_expression_df = process_gene_pair_optimized(\n",
    "        pair, filtered_go_df_expr, go_to_mean_expression, gene_to_go_expr\n",
    "    )\n",
    "    summary_df = summarize_GO_expression_scores(pair_df, pair_expression_df, pair, paralog_pairs_dict)\n",
    "\n",
    "    if summary_df is not None:\n",
    "        expression_summaries.append(summary_df)\n",
    "\n",
    "# Step 5: Final combined result\n",
    "final_expression_df = pd.concat(expression_summaries).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_line</th>\n",
       "      <th>paralog_pair</th>\n",
       "      <th>smallest_GO_accession</th>\n",
       "      <th>smallest_gene_expression</th>\n",
       "      <th>min_GO_accession</th>\n",
       "      <th>min_gene_expression</th>\n",
       "      <th>max_GO_accession</th>\n",
       "      <th>max_gene_expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-001113</td>\n",
       "      <td>SMARCA2_SMARCA4</td>\n",
       "      <td>GO:0070316</td>\n",
       "      <td>4.449597</td>\n",
       "      <td>GO:0045596</td>\n",
       "      <td>2.391188</td>\n",
       "      <td>GO:2000819</td>\n",
       "      <td>4.702816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-001289</td>\n",
       "      <td>SMARCA2_SMARCA4</td>\n",
       "      <td>GO:0070316</td>\n",
       "      <td>4.992393</td>\n",
       "      <td>GO:0045596</td>\n",
       "      <td>2.311941</td>\n",
       "      <td>GO:2000819</td>\n",
       "      <td>5.106746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-001339</td>\n",
       "      <td>SMARCA2_SMARCA4</td>\n",
       "      <td>GO:0070316</td>\n",
       "      <td>4.407439</td>\n",
       "      <td>GO:0045596</td>\n",
       "      <td>2.210668</td>\n",
       "      <td>GO:2000819</td>\n",
       "      <td>4.510038</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cell_line     paralog_pair smallest_GO_accession  \\\n",
       "0  ACH-001113  SMARCA2_SMARCA4            GO:0070316   \n",
       "1  ACH-001289  SMARCA2_SMARCA4            GO:0070316   \n",
       "2  ACH-001339  SMARCA2_SMARCA4            GO:0070316   \n",
       "\n",
       "   smallest_gene_expression min_GO_accession  min_gene_expression  \\\n",
       "0                  4.449597       GO:0045596             2.391188   \n",
       "1                  4.992393       GO:0045596             2.311941   \n",
       "2                  4.407439       GO:0045596             2.210668   \n",
       "\n",
       "  max_GO_accession  max_gene_expression  \n",
       "0       GO:2000819             4.702816  \n",
       "1       GO:2000819             5.106746  \n",
       "2       GO:2000819             4.510038  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_expression_df[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15793, 36237)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_expression_df['paralog_pair'].nunique(), paralog_pairs.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>smallest_GO_accession</th>\n",
       "      <th>smallest_gene_expression</th>\n",
       "      <th>min_GO_accession</th>\n",
       "      <th>min_gene_expression</th>\n",
       "      <th>max_GO_accession</th>\n",
       "      <th>max_gene_expression</th>\n",
       "      <th>GO:0007172</th>\n",
       "      <th>GO:0007173</th>\n",
       "      <th>GO:0007229</th>\n",
       "      <th>GO:0018108</th>\n",
       "      <th>GO:0030155</th>\n",
       "      <th>GO:0045860</th>\n",
       "      <th>GO:0048010</th>\n",
       "      <th>GO:0051128</th>\n",
       "      <th>GO:0051239</th>\n",
       "      <th>GO:0071310</th>\n",
       "      <th>GO:1902531</th>\n",
       "      <th>GO:2000060</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACH-000022</th>\n",
       "      <td>GO:0007172</td>\n",
       "      <td>4.378402</td>\n",
       "      <td>GO:0018108</td>\n",
       "      <td>2.120749</td>\n",
       "      <td>GO:0007172</td>\n",
       "      <td>4.378402</td>\n",
       "      <td>4.378402</td>\n",
       "      <td>3.406025</td>\n",
       "      <td>2.990841</td>\n",
       "      <td>2.120749</td>\n",
       "      <td>2.714527</td>\n",
       "      <td>2.675518</td>\n",
       "      <td>2.553962</td>\n",
       "      <td>3.321127</td>\n",
       "      <td>2.837618</td>\n",
       "      <td>2.525463</td>\n",
       "      <td>4.360907</td>\n",
       "      <td>2.968965</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           smallest_GO_accession  smallest_gene_expression min_GO_accession  \\\n",
       "ACH-000022            GO:0007172                  4.378402       GO:0018108   \n",
       "\n",
       "            min_gene_expression max_GO_accession  max_gene_expression  \\\n",
       "ACH-000022             2.120749       GO:0007172             4.378402   \n",
       "\n",
       "            GO:0007172  GO:0007173  GO:0007229  GO:0018108  GO:0030155  \\\n",
       "ACH-000022    4.378402    3.406025    2.990841    2.120749    2.714527   \n",
       "\n",
       "            GO:0045860  GO:0048010  GO:0051128  GO:0051239  GO:0071310  \\\n",
       "ACH-000022    2.675518    2.553962    3.321127    2.837618    2.525463   \n",
       "\n",
       "            GO:1902531  GO:2000060  \n",
       "ACH-000022    4.360907    2.968965  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_line</th>\n",
       "      <th>paralog_pair</th>\n",
       "      <th>smallest_GO_accession</th>\n",
       "      <th>smallest_gene_expression</th>\n",
       "      <th>min_GO_accession</th>\n",
       "      <th>min_gene_expression</th>\n",
       "      <th>max_GO_accession</th>\n",
       "      <th>max_gene_expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>220029</th>\n",
       "      <td>ACH-000022</td>\n",
       "      <td>PRKACA_PRKACB</td>\n",
       "      <td>GO:0034380</td>\n",
       "      <td>2.107183</td>\n",
       "      <td>GO:0070613</td>\n",
       "      <td>1.742455</td>\n",
       "      <td>GO:0001843</td>\n",
       "      <td>3.426666</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         cell_line   paralog_pair smallest_GO_accession  \\\n",
       "220029  ACH-000022  PRKACA_PRKACB            GO:0034380   \n",
       "\n",
       "        smallest_gene_expression min_GO_accession  min_gene_expression  \\\n",
       "220029                  2.107183       GO:0070613             1.742455   \n",
       "\n",
       "       max_GO_accession  max_gene_expression  \n",
       "220029       GO:0001843             3.426666  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "###### Test ######\n",
    "\n",
    "#PRKACA_PRKACB\n",
    "#pair = {'ENSG00000072062', 'ENSG00000142875'}\n",
    "\n",
    "#PTK2_PTK2B\n",
    "pair = {'ENSG00000169398', 'ENSG00000120899'}\n",
    "\n",
    "filtered_df = sum_bp_paralogs_df[sum_bp_paralogs_df['genes'].apply(lambda x: pair.issubset(x))]\n",
    "filtered_df.insert(filtered_df.shape[1], 'len_of_genes', filtered_df['genes'].apply(len))\n",
    "\n",
    "go_effect_subset = {go: go_to_mean_expression[go] for go in filtered_df['GO_accession']}\n",
    "go_effect_df = pd.DataFrame(go_effect_subset)\n",
    "\n",
    "min_len = filtered_df['len_of_genes'].min()\n",
    "smallest_df = filtered_df[filtered_df['len_of_genes'] == min_len]\n",
    "smallest_go_ids = smallest_df['GO_accession'].tolist()\n",
    "smallest_subset = go_effect_df[smallest_go_ids]\n",
    "\n",
    "#smallest GOs: GO:0007172 and GO:1902531\n",
    "\n",
    "result_df = pd.DataFrame(index=go_effect_df.index)\n",
    "result_df['smallest_GO_accession'] = smallest_subset.idxmax(axis=1)\n",
    "result_df['smallest_gene_expression'] = smallest_subset.max(axis=1)\n",
    "result_df['min_GO_accession'] = go_effect_df.idxmin(axis=1)\n",
    "result_df['min_gene_expression'] = go_effect_df.min(axis=1)\n",
    "result_df['max_GO_accession'] = go_effect_df.idxmax(axis=1)\n",
    "result_df['max_gene_expression'] = go_effect_df.max(axis=1)\n",
    "\n",
    "# include all GO expression values for comparison\n",
    "result_df = pd.concat([result_df, go_effect_df], axis=1)\n",
    "display(result_df.loc[result_df.index == 'ACH-000022'])\n",
    "\n",
    "# --- paralog pair ---\n",
    "display(final_expression_df.loc[(final_expression_df.paralog_pair == 'PRKACA_PRKACB') & (final_expression_df.cell_line == 'ACH-000022'), :])\n",
    "\n",
    "###### Test ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_to_mean_expr_norm, filtered_go_df_expr_norm = precompute_go_effects(\n",
    "    sum_bp_paralogs_df, zexpression_data\n",
    ")\n",
    "\n",
    "gene_to_go_expr_norm = build_gene_to_go_map(filtered_go_df_expr_norm)\n",
    "\n",
    "# Step 3: Process paralog pairs\n",
    "normalized_expr_summaries = []\n",
    "\n",
    "for pair in paralog_pairs:\n",
    "    pair_df, pair_expr_df = process_gene_pair_optimized(\n",
    "        pair, filtered_go_df_expr_norm, go_to_mean_expr_norm, gene_to_go_expr_norm\n",
    "    )\n",
    "    summary_df = summarize_GO_expression_scores(\n",
    "        pair_df, pair_expr_df, pair, paralog_pairs_dict\n",
    "    )\n",
    "\n",
    "    if summary_df is not None:\n",
    "        normalized_expr_summaries.append(summary_df)\n",
    "\n",
    "# Step 4: Final DataFrame\n",
    "final_expr_norm_df = pd.concat(normalized_expr_summaries).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15793, 36237)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_expr_norm_df['paralog_pair'].nunique(), paralog_pairs.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_expression_df = final_expression_df.rename(columns={\n",
    "    'smallest_GO_accession': 'smallest_BP_GO_accession',\n",
    "    'smallest_gene_expression': 'smallest_BP_GO_gene_expression',\n",
    "    'min_GO_accession': 'min_BP_GO_accession', \n",
    "    'min_gene_expression': 'min_BP_GO_gene_expression',\n",
    "    'max_GO_accession': 'max_BP_GO_accession', \n",
    "    'max_gene_expression': 'max_BP_GO_gene_expression',\n",
    "})\n",
    "\n",
    "final_expr_norm_df = final_expr_norm_df.rename(columns={\n",
    "    'smallest_GO_accession': 'zsmallest_BP_GO_accession',\n",
    "    'smallest_gene_expression': 'zsmallest_BP_GO_gene_expression',\n",
    "    'min_GO_accession': 'zmin_BP_GO_accession', \n",
    "    'min_gene_expression': 'zmin_BP_GO_gene_expression',\n",
    "    'max_GO_accession': 'zmax_BP_GO_accession', \n",
    "    'max_gene_expression': 'zmax_BP_GO_gene_expression',\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "BP_GO_merged_df = pd.merge(\n",
    "    final_expression_df, final_expr_norm_df,\n",
    "    on=['paralog_pair', 'cell_line'],\n",
    "    how='outer',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_line</th>\n",
       "      <th>paralog_pair</th>\n",
       "      <th>smallest_BP_GO_accession</th>\n",
       "      <th>smallest_BP_GO_gene_expression</th>\n",
       "      <th>min_BP_GO_accession</th>\n",
       "      <th>min_BP_GO_gene_expression</th>\n",
       "      <th>max_BP_GO_accession</th>\n",
       "      <th>max_BP_GO_gene_expression</th>\n",
       "      <th>zsmallest_BP_GO_accession</th>\n",
       "      <th>zsmallest_BP_GO_gene_expression</th>\n",
       "      <th>zmin_BP_GO_accession</th>\n",
       "      <th>zmin_BP_GO_gene_expression</th>\n",
       "      <th>zmax_BP_GO_accession</th>\n",
       "      <th>zmax_BP_GO_gene_expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>A1CF_RBM47</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>0.977073</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>0.977073</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>0.977073</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>-0.375680</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>-0.375680</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>-0.375680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000002</td>\n",
       "      <td>A1CF_RBM47</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>0.293647</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>0.293647</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>0.293647</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>-0.788860</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>-0.788860</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>-0.788860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000003</td>\n",
       "      <td>A1CF_RBM47</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>2.008663</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>2.008663</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>2.008663</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>0.650103</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>0.650103</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>0.650103</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cell_line paralog_pair smallest_BP_GO_accession  \\\n",
       "0  ACH-000001   A1CF_RBM47               GO:0016554   \n",
       "1  ACH-000002   A1CF_RBM47               GO:0016554   \n",
       "2  ACH-000003   A1CF_RBM47               GO:0016554   \n",
       "\n",
       "   smallest_BP_GO_gene_expression min_BP_GO_accession  \\\n",
       "0                        0.977073          GO:0016554   \n",
       "1                        0.293647          GO:0016554   \n",
       "2                        2.008663          GO:0016554   \n",
       "\n",
       "   min_BP_GO_gene_expression max_BP_GO_accession  max_BP_GO_gene_expression  \\\n",
       "0                   0.977073          GO:0016554                   0.977073   \n",
       "1                   0.293647          GO:0016554                   0.293647   \n",
       "2                   2.008663          GO:0016554                   2.008663   \n",
       "\n",
       "  zsmallest_BP_GO_accession  zsmallest_BP_GO_gene_expression  \\\n",
       "0                GO:0016554                        -0.375680   \n",
       "1                GO:0016554                        -0.788860   \n",
       "2                GO:0016554                         0.650103   \n",
       "\n",
       "  zmin_BP_GO_accession  zmin_BP_GO_gene_expression zmax_BP_GO_accession  \\\n",
       "0           GO:0016554                   -0.375680           GO:0016554   \n",
       "1           GO:0016554                   -0.788860           GO:0016554   \n",
       "2           GO:0016554                    0.650103           GO:0016554   \n",
       "\n",
       "   zmax_BP_GO_gene_expression  \n",
       "0                   -0.375680  \n",
       "1                   -0.788860  \n",
       "2                    0.650103  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BP_GO_merged_df[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 1: Precompute mean z-expression per GO term (CC ontology)\n",
    "go_to_mean_expr_cc, filtered_go_df_expr_cc = precompute_go_effects(\n",
    "    sum_cc_paralogs_df, gene_expression\n",
    ")\n",
    "\n",
    "# Step 2: Build gene → GO term map\n",
    "gene_to_go_expr_cc = build_gene_to_go_map(filtered_go_df_expr_cc)\n",
    "\n",
    "# Step 3: Process paralog pairs and summarize GO scores\n",
    "normalized_expr_summaries_cc = []\n",
    "\n",
    "for pair in paralog_pairs:\n",
    "    pair_df, pair_expr_df = process_gene_pair_optimized(\n",
    "        pair,\n",
    "        filtered_go_df_expr_cc,\n",
    "        go_to_mean_expr_cc,\n",
    "        gene_to_go_expr_cc\n",
    "    )\n",
    "\n",
    "    summary_df = summarize_GO_expression_scores(\n",
    "        pair_df,\n",
    "        pair_expr_df,\n",
    "        pair,\n",
    "        paralog_pairs_dict\n",
    "    )\n",
    "\n",
    "    if summary_df is not None:\n",
    "        normalized_expr_summaries_cc.append(summary_df)\n",
    "\n",
    "# Step 4: Combine results\n",
    "final_expr_cc_df = pd.concat(normalized_expr_summaries_cc).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 1: Precompute mean z-expression per GO term (CC ontology)\n",
    "go_to_mean_expr_norm_cc, filtered_go_df_expr_norm_cc = precompute_go_effects(\n",
    "    sum_cc_paralogs_df, zexpression_data\n",
    ")\n",
    "\n",
    "# Step 2: Build gene → GO term map\n",
    "gene_to_go_expr_norm_cc = build_gene_to_go_map(filtered_go_df_expr_norm_cc)\n",
    "\n",
    "# Step 3: Process paralog pairs and summarize GO scores\n",
    "normalized_expr_summaries_cc = []\n",
    "\n",
    "for pair in paralog_pairs:\n",
    "    pair_df, pair_expr_df = process_gene_pair_optimized(\n",
    "        pair,\n",
    "        filtered_go_df_expr_norm_cc,\n",
    "        go_to_mean_expr_norm_cc,\n",
    "        gene_to_go_expr_norm_cc\n",
    "    )\n",
    "\n",
    "    summary_df = summarize_GO_expression_scores(\n",
    "        pair_df,\n",
    "        pair_expr_df,\n",
    "        pair,\n",
    "        paralog_pairs_dict\n",
    "    )\n",
    "\n",
    "    if summary_df is not None:\n",
    "        normalized_expr_summaries_cc.append(summary_df)\n",
    "\n",
    "# Step 4: Combine results\n",
    "final_expr_norm_cc_df = pd.concat(normalized_expr_summaries_cc).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_expr_cc_df = final_expr_cc_df.rename(columns={\n",
    "    'smallest_GO_accession': 'smallest_CC_GO_accession',\n",
    "    'smallest_gene_expression': 'smallest_CC_GO_gene_expression',\n",
    "    'min_GO_accession': 'min_CC_GO_accession', \n",
    "    'min_gene_expression': 'min_CC_GO_gene_expression',\n",
    "    'max_GO_accession': 'max_CC_GO_accession', \n",
    "    'max_gene_expression': 'max_CC_GO_gene_expression',\n",
    "})\n",
    "\n",
    "final_expr_norm_cc_df = final_expr_norm_cc_df.rename(columns={\n",
    "    'smallest_GO_accession': 'zsmallest_CC_GO_accession',\n",
    "    'smallest_gene_expression': 'zsmallest_CC_GO_gene_expression',\n",
    "    'min_GO_accession': 'zmin_CC_GO_accession', \n",
    "    'min_gene_expression': 'zmin_CC_GO_gene_expression',\n",
    "    'max_GO_accession': 'zmax_CC_GO_accession', \n",
    "    'max_gene_expression': 'zmax_CC_GO_gene_expression',\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "CC_GO_merged_df = pd.merge(\n",
    "    final_expr_cc_df, final_expr_norm_cc_df,\n",
    "    on=['paralog_pair', 'cell_line'],\n",
    "    how='outer',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_line</th>\n",
       "      <th>paralog_pair</th>\n",
       "      <th>smallest_CC_GO_accession</th>\n",
       "      <th>smallest_CC_GO_gene_expression</th>\n",
       "      <th>min_CC_GO_accession</th>\n",
       "      <th>min_CC_GO_gene_expression</th>\n",
       "      <th>max_CC_GO_accession</th>\n",
       "      <th>max_CC_GO_gene_expression</th>\n",
       "      <th>zsmallest_CC_GO_accession</th>\n",
       "      <th>zsmallest_CC_GO_gene_expression</th>\n",
       "      <th>zmin_CC_GO_accession</th>\n",
       "      <th>zmin_CC_GO_gene_expression</th>\n",
       "      <th>zmax_CC_GO_accession</th>\n",
       "      <th>zmax_CC_GO_gene_expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>A2M_C3</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>2.248426</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>2.248426</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>2.248426</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>0.115564</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>0.115564</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>0.115564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000002</td>\n",
       "      <td>A2M_C3</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>1.409691</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>1.409691</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>1.409691</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>-0.363556</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>-0.363556</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>-0.363556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000003</td>\n",
       "      <td>A2M_C3</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>2.826250</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>2.826250</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>2.826250</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>0.790761</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>0.790761</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>0.790761</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cell_line paralog_pair smallest_CC_GO_accession  \\\n",
       "0  ACH-000001       A2M_C3               GO:0072562   \n",
       "1  ACH-000002       A2M_C3               GO:0072562   \n",
       "2  ACH-000003       A2M_C3               GO:0072562   \n",
       "\n",
       "   smallest_CC_GO_gene_expression min_CC_GO_accession  \\\n",
       "0                        2.248426          GO:0072562   \n",
       "1                        1.409691          GO:0072562   \n",
       "2                        2.826250          GO:0072562   \n",
       "\n",
       "   min_CC_GO_gene_expression max_CC_GO_accession  max_CC_GO_gene_expression  \\\n",
       "0                   2.248426          GO:0072562                   2.248426   \n",
       "1                   1.409691          GO:0072562                   1.409691   \n",
       "2                   2.826250          GO:0072562                   2.826250   \n",
       "\n",
       "  zsmallest_CC_GO_accession  zsmallest_CC_GO_gene_expression  \\\n",
       "0                GO:0072562                         0.115564   \n",
       "1                GO:0072562                        -0.363556   \n",
       "2                GO:0072562                         0.790761   \n",
       "\n",
       "  zmin_CC_GO_accession  zmin_CC_GO_gene_expression zmax_CC_GO_accession  \\\n",
       "0           GO:0072562                    0.115564           GO:0072562   \n",
       "1           GO:0072562                   -0.363556           GO:0072562   \n",
       "2           GO:0072562                    0.790761           GO:0072562   \n",
       "\n",
       "   zmax_CC_GO_gene_expression  \n",
       "0                    0.115564  \n",
       "1                   -0.363556  \n",
       "2                    0.790761  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CC_GO_merged_df[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_line</th>\n",
       "      <th>paralog_pair</th>\n",
       "      <th>smallest_CC_GO_accession</th>\n",
       "      <th>smallest_CC_GO_gene_expression</th>\n",
       "      <th>min_CC_GO_accession</th>\n",
       "      <th>min_CC_GO_gene_expression</th>\n",
       "      <th>max_CC_GO_accession</th>\n",
       "      <th>max_CC_GO_gene_expression</th>\n",
       "      <th>zsmallest_CC_GO_accession</th>\n",
       "      <th>zsmallest_CC_GO_gene_expression</th>\n",
       "      <th>zmin_CC_GO_accession</th>\n",
       "      <th>zmin_CC_GO_gene_expression</th>\n",
       "      <th>zmax_CC_GO_accession</th>\n",
       "      <th>zmax_CC_GO_gene_expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7437076</th>\n",
       "      <td>ACH-000022</td>\n",
       "      <td>PRKACA_PRKACB</td>\n",
       "      <td>GO:0005952</td>\n",
       "      <td>2.395621</td>\n",
       "      <td>GO:0097546</td>\n",
       "      <td>2.282365</td>\n",
       "      <td>GO:0005952</td>\n",
       "      <td>2.395621</td>\n",
       "      <td>GO:0005952</td>\n",
       "      <td>-0.383464</td>\n",
       "      <td>GO:0005952</td>\n",
       "      <td>-0.383464</td>\n",
       "      <td>GO:0097546</td>\n",
       "      <td>-0.190227</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          cell_line   paralog_pair smallest_CC_GO_accession  \\\n",
       "7437076  ACH-000022  PRKACA_PRKACB               GO:0005952   \n",
       "\n",
       "         smallest_CC_GO_gene_expression min_CC_GO_accession  \\\n",
       "7437076                        2.395621          GO:0097546   \n",
       "\n",
       "         min_CC_GO_gene_expression max_CC_GO_accession  \\\n",
       "7437076                   2.282365          GO:0005952   \n",
       "\n",
       "         max_CC_GO_gene_expression zsmallest_CC_GO_accession  \\\n",
       "7437076                   2.395621                GO:0005952   \n",
       "\n",
       "         zsmallest_CC_GO_gene_expression zmin_CC_GO_accession  \\\n",
       "7437076                        -0.383464           GO:0005952   \n",
       "\n",
       "         zmin_CC_GO_gene_expression zmax_CC_GO_accession  \\\n",
       "7437076                   -0.383464           GO:0097546   \n",
       "\n",
       "         zmax_CC_GO_gene_expression  \n",
       "7437076                   -0.190227  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CC_GO_merged_df.loc[(CC_GO_merged_df['paralog_pair'] == 'PRKACA_PRKACB') & (CC_GO_merged_df['cell_line'] == 'ACH-000022'), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the files\n",
    "output_path = get_data_path(['input', 'GO'], '')\n",
    "\n",
    "BP_GO_merged_df.to_parquet(os.path.join(output_path, 'go_BP_expression.parquet'), engine='pyarrow', index=True)\n",
    "CC_GO_merged_df.to_parquet(os.path.join(output_path, 'go_CC_expression.parquet'), engine='pyarrow', index=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "predictor",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
