{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import modules\n",
    "import os\n",
    "import re\n",
    "import csv\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pyarrow as pa\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set the base directory for the project\n",
    "cwd = os.getcwd()\n",
    "BASE_DIR = os.path.abspath(os.path.join(cwd, \"..\", \"..\"))\n",
    "\n",
    "# build paths inside the repo\n",
    "get_data_path = lambda folders, fname: os.path.normpath(\n",
    "    os.path.join(BASE_DIR, *folders, fname)\n",
    ")\n",
    "\n",
    "file_path_gene_essentiality = get_data_path(['data', 'output', 'ranked_essentiality'], 'ranked_essentiality.csv')\n",
    "file_path_gene_zessentiality = get_data_path(['data', 'output', 'ranked_essentiality'], 'ranked_zessentiality.csv')\n",
    "\n",
    "dekegel_table8_path = get_data_path(['data','input', 'other'], 'processed_DeKegel_TableS8.csv')\n",
    "\n",
    "file_path_genenames = get_data_path(['data', 'input', 'other'], 'approved_and_previous_symbols.csv')\n",
    "\n",
    "file_path_ensembl = get_data_path(['data', 'input', 'other'], 'mart_export.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read the gene names mapping file\n",
    "id_map = pd.read_csv(file_path_genenames)\n",
    "\n",
    "# create dictionaries to map gene symbols to Entrez IDs\n",
    "id_map_cleaned = id_map.dropna(axis=0, how='any', subset=['entrez_id', 'ensembl_id']).reset_index(drop=True)\n",
    "entrez_id_to_ensembl = dict(zip(id_map_cleaned.entrez_id, id_map_cleaned.ensembl_id))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Effect Score\n",
    "\n",
    "#### Differential activity score\n",
    "- Null hypothesis: the activity of the gene stays the same in two different conditions\n",
    "- Alternative hypothesis: the activity of the gene is different in two different conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_essentiality = pd.read_csv(file_path_gene_essentiality, sep=',', index_col=0, low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_z_essentiality = pd.read_csv(file_path_gene_zessentiality, sep=',', index_col=0, low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000121410</th>\n",
       "      <th>ENSG00000175899</th>\n",
       "      <th>ENSG00000171428</th>\n",
       "      <th>ENSG00000156006</th>\n",
       "      <th>ENSG00000196136</th>\n",
       "      <th>ENSG00000114771</th>\n",
       "      <th>ENSG00000127837</th>\n",
       "      <th>ENSG00000129673</th>\n",
       "      <th>ENSG00000090861</th>\n",
       "      <th>ENSG00000183044</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000236311</th>\n",
       "      <th>ENSG00000218819</th>\n",
       "      <th>ENSG00000185055</th>\n",
       "      <th>ENSG00000184224</th>\n",
       "      <th>100505705</th>\n",
       "      <th>ENSG00000204403</th>\n",
       "      <th>ENSG00000250733</th>\n",
       "      <th>ENSG00000214447</th>\n",
       "      <th>ENSG00000283787</th>\n",
       "      <th>ENSG00000162621</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACH-000004</th>\n",
       "      <td>11172.0</td>\n",
       "      <td>5346.0</td>\n",
       "      <td>15919.0</td>\n",
       "      <td>14531.0</td>\n",
       "      <td>7827.0</td>\n",
       "      <td>15880.0</td>\n",
       "      <td>681.0</td>\n",
       "      <td>3342.0</td>\n",
       "      <td>238.0</td>\n",
       "      <td>11965.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000005</th>\n",
       "      <td>3608.0</td>\n",
       "      <td>14189.0</td>\n",
       "      <td>17197.0</td>\n",
       "      <td>17118.0</td>\n",
       "      <td>4887.0</td>\n",
       "      <td>11519.0</td>\n",
       "      <td>785.0</td>\n",
       "      <td>9371.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>3221.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000007</th>\n",
       "      <td>9393.0</td>\n",
       "      <td>12274.0</td>\n",
       "      <td>15246.0</td>\n",
       "      <td>16239.0</td>\n",
       "      <td>16454.0</td>\n",
       "      <td>16579.0</td>\n",
       "      <td>645.0</td>\n",
       "      <td>12509.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>16790.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 19158 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ENSG00000121410  ENSG00000175899  ENSG00000171428  \\\n",
       "ACH-000004          11172.0           5346.0          15919.0   \n",
       "ACH-000005           3608.0          14189.0          17197.0   \n",
       "ACH-000007           9393.0          12274.0          15246.0   \n",
       "\n",
       "            ENSG00000156006  ENSG00000196136  ENSG00000114771  \\\n",
       "ACH-000004          14531.0           7827.0          15880.0   \n",
       "ACH-000005          17118.0           4887.0          11519.0   \n",
       "ACH-000007          16239.0          16454.0          16579.0   \n",
       "\n",
       "            ENSG00000127837  ENSG00000129673  ENSG00000090861  \\\n",
       "ACH-000004            681.0           3342.0            238.0   \n",
       "ACH-000005            785.0           9371.0             98.0   \n",
       "ACH-000007            645.0          12509.0             91.0   \n",
       "\n",
       "            ENSG00000183044  ...  ENSG00000236311  ENSG00000218819  \\\n",
       "ACH-000004          11965.0  ...              NaN              NaN   \n",
       "ACH-000005           3221.0  ...              NaN              NaN   \n",
       "ACH-000007          16790.0  ...              NaN              NaN   \n",
       "\n",
       "            ENSG00000185055  ENSG00000184224  100505705  ENSG00000204403  \\\n",
       "ACH-000004              NaN              NaN        NaN              NaN   \n",
       "ACH-000005              NaN              NaN        NaN              NaN   \n",
       "ACH-000007              NaN              NaN        NaN              NaN   \n",
       "\n",
       "            ENSG00000250733  ENSG00000214447  ENSG00000283787  ENSG00000162621  \n",
       "ACH-000004              NaN              NaN              NaN              NaN  \n",
       "ACH-000005              NaN              NaN              NaN              NaN  \n",
       "ACH-000007              NaN              NaN              NaN              NaN  \n",
       "\n",
       "[3 rows x 19158 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000121410</th>\n",
       "      <th>ENSG00000175899</th>\n",
       "      <th>ENSG00000171428</th>\n",
       "      <th>ENSG00000156006</th>\n",
       "      <th>ENSG00000196136</th>\n",
       "      <th>ENSG00000114771</th>\n",
       "      <th>ENSG00000127837</th>\n",
       "      <th>ENSG00000129673</th>\n",
       "      <th>ENSG00000090861</th>\n",
       "      <th>ENSG00000183044</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000236311</th>\n",
       "      <th>ENSG00000218819</th>\n",
       "      <th>ENSG00000185055</th>\n",
       "      <th>ENSG00000184224</th>\n",
       "      <th>100505705</th>\n",
       "      <th>ENSG00000204403</th>\n",
       "      <th>ENSG00000250733</th>\n",
       "      <th>ENSG00000214447</th>\n",
       "      <th>ENSG00000283787</th>\n",
       "      <th>ENSG00000162621</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACH-000004</th>\n",
       "      <td>0.591940</td>\n",
       "      <td>-1.861455</td>\n",
       "      <td>0.767406</td>\n",
       "      <td>0.765687</td>\n",
       "      <td>-1.188311</td>\n",
       "      <td>0.721604</td>\n",
       "      <td>-0.474503</td>\n",
       "      <td>-1.111145</td>\n",
       "      <td>0.112406</td>\n",
       "      <td>0.519840</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000005</th>\n",
       "      <td>-1.651996</td>\n",
       "      <td>0.848431</td>\n",
       "      <td>1.195610</td>\n",
       "      <td>1.503581</td>\n",
       "      <td>-2.063585</td>\n",
       "      <td>-0.574681</td>\n",
       "      <td>-0.316621</td>\n",
       "      <td>0.907069</td>\n",
       "      <td>-0.778173</td>\n",
       "      <td>-1.745288</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000007</th>\n",
       "      <td>0.064182</td>\n",
       "      <td>0.261590</td>\n",
       "      <td>0.541912</td>\n",
       "      <td>1.252863</td>\n",
       "      <td>1.380053</td>\n",
       "      <td>0.929378</td>\n",
       "      <td>-0.529154</td>\n",
       "      <td>1.957518</td>\n",
       "      <td>-0.822702</td>\n",
       "      <td>1.769753</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 19158 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ENSG00000121410  ENSG00000175899  ENSG00000171428  \\\n",
       "ACH-000004         0.591940        -1.861455         0.767406   \n",
       "ACH-000005        -1.651996         0.848431         1.195610   \n",
       "ACH-000007         0.064182         0.261590         0.541912   \n",
       "\n",
       "            ENSG00000156006  ENSG00000196136  ENSG00000114771  \\\n",
       "ACH-000004         0.765687        -1.188311         0.721604   \n",
       "ACH-000005         1.503581        -2.063585        -0.574681   \n",
       "ACH-000007         1.252863         1.380053         0.929378   \n",
       "\n",
       "            ENSG00000127837  ENSG00000129673  ENSG00000090861  \\\n",
       "ACH-000004        -0.474503        -1.111145         0.112406   \n",
       "ACH-000005        -0.316621         0.907069        -0.778173   \n",
       "ACH-000007        -0.529154         1.957518        -0.822702   \n",
       "\n",
       "            ENSG00000183044  ...  ENSG00000236311  ENSG00000218819  \\\n",
       "ACH-000004         0.519840  ...              NaN              NaN   \n",
       "ACH-000005        -1.745288  ...              NaN              NaN   \n",
       "ACH-000007         1.769753  ...              NaN              NaN   \n",
       "\n",
       "            ENSG00000185055  ENSG00000184224  100505705  ENSG00000204403  \\\n",
       "ACH-000004              NaN              NaN        NaN              NaN   \n",
       "ACH-000005              NaN              NaN        NaN              NaN   \n",
       "ACH-000007              NaN              NaN        NaN              NaN   \n",
       "\n",
       "            ENSG00000250733  ENSG00000214447  ENSG00000283787  ENSG00000162621  \n",
       "ACH-000004              NaN              NaN              NaN              NaN  \n",
       "ACH-000005              NaN              NaN              NaN              NaN  \n",
       "ACH-000007              NaN              NaN              NaN              NaN  \n",
       "\n",
       "[3 rows x 19158 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gene_essentiality.columns = gene_essentiality.columns.astype(int)  #replace the entrez_id with ensembl_id\n",
    "gene_essentiality = gene_essentiality.rename(columns=entrez_id_to_ensembl)\n",
    "display(gene_essentiality[:3])\n",
    "\n",
    "gene_z_essentiality.columns = gene_z_essentiality.columns.astype(int)\n",
    "gene_z_essentiality = gene_z_essentiality.rename(columns=entrez_id_to_ensembl) #replace the entrez_id with ensembl_id\n",
    "display(gene_z_essentiality[:3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Paralog Pairs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "paralog_pairs_dict = {}\n",
    "list_of_paralogs = set()\n",
    "#paralog_pairs = []\n",
    "with open(dekegel_table8_path, \"r\") as f :\n",
    "    reader = csv.DictReader(f)\n",
    "    for r in reader :\n",
    "        sorted_gene_pair = r['genepair']\n",
    "        a1 = r['A1_ensembl_new']\n",
    "        a2 = r['A2_ensembl_new']\n",
    "        if a1 in gene_essentiality.columns and a2 in gene_essentiality.columns:\n",
    "            #paralog_pairs_dict[sorted_gene_pair] = (a1, a2)\n",
    "            paralog_pairs_dict[(a1, a2)] = sorted_gene_pair\n",
    "            list_of_paralogs.update((a1, a2))\n",
    "            #paralog_pairs.append(set([a1, a2]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#paralog_pairs = []\n",
    "#for key in paralog_pairs_dict.keys():\n",
    "#    if key[0] != key[1]:\n",
    "#        paralog_pairs.append([key[0], key[1]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "paralog_pairs = pd.Series(list(paralog_pairs_dict.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('ENSG00000080503', 'ENSG00000127616')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paralog_pairs[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "35656"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(paralog_pairs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pairs with internal NA values: 0\n"
     ]
    }
   ],
   "source": [
    "# Check if any tuple contains None/NA elements\n",
    "def has_na_in_tuple(pair):\n",
    "    if pd.isna(pair):\n",
    "        return True\n",
    "    try:\n",
    "        return any(pd.isna(element) for element in pair)\n",
    "    except:\n",
    "        return False\n",
    "\n",
    "# Apply the check to all pairs\n",
    "has_internal_na = paralog_pairs.apply(has_na_in_tuple)\n",
    "print(f\"Pairs with internal NA values: {has_internal_na.sum()}\")\n",
    "\n",
    "# Show pairs with internal NA (if any)\n",
    "if has_internal_na.any():\n",
    "    print(paralog_pairs[has_internal_na])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Protein-coding paralog pairs w/ their corresponding GO term annotations\n",
    "\n",
    "Ensembl Filters:\n",
    "\n",
    "- Paralogous Human Genes: Only\n",
    "- Gene type: protein_coding\n",
    "\n",
    "Attributes:\n",
    "\n",
    "- Gene Stable ID\n",
    "- GO Term Accession\n",
    "- GO Term Name\n",
    "- GO Domain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>ensembl_transcript_id</th>\n",
       "      <th>GO_accession</th>\n",
       "      <th>GO_name</th>\n",
       "      <th>GO_domain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0008137</td>\n",
       "      <td>NADH dehydrogenase (ubiquinone) activity</td>\n",
       "      <td>molecular_function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0006120</td>\n",
       "      <td>mitochondrial electron transport, NADH to ubiq...</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0016020</td>\n",
       "      <td>membrane</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0016491</td>\n",
       "      <td>oxidoreductase activity</td>\n",
       "      <td>molecular_function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0005739</td>\n",
       "      <td>mitochondrion</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ensembl_id ensembl_transcript_id GO_accession  \\\n",
       "0  ENSG00000198763       ENST00000361453   GO:0008137   \n",
       "1  ENSG00000198763       ENST00000361453   GO:0006120   \n",
       "2  ENSG00000198763       ENST00000361453   GO:0016020   \n",
       "3  ENSG00000198763       ENST00000361453   GO:0016491   \n",
       "4  ENSG00000198763       ENST00000361453   GO:0005739   \n",
       "\n",
       "                                             GO_name           GO_domain  \n",
       "0           NADH dehydrogenase (ubiquinone) activity  molecular_function  \n",
       "1  mitochondrial electron transport, NADH to ubiq...  biological_process  \n",
       "2                                           membrane  cellular_component  \n",
       "3                            oxidoreductase activity  molecular_function  \n",
       "4                                      mitochondrion  cellular_component  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read the ensemble file\n",
    "raw_data = pd.read_table(file_path_ensembl, sep=',', dtype='str')\n",
    "raw_data = raw_data.rename(columns={'Gene stable ID':'ensembl_id', 'Transcript stable ID':'ensembl_transcript_id', \n",
    "                                    'GO term accession': 'GO_accession', 'GO term name':'GO_name',\n",
    "                                    'GO domain':'GO_domain'})\n",
    "raw_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# of genes w/ their corresponding GO term annotations (BP): 16207 \n",
      " # of unique GO term accession (BP): 11911\n",
      "# of genes w/ their corresponding GO term annotations (CC): 16832 \n",
      " # of unique GO term accession (CC): 1662\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>ensembl_transcript_id</th>\n",
       "      <th>GO_accession</th>\n",
       "      <th>GO_name</th>\n",
       "      <th>GO_domain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0006120</td>\n",
       "      <td>mitochondrial electron transport, NADH to ubiq...</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0032981</td>\n",
       "      <td>mitochondrial respiratory chain complex I asse...</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0006120</td>\n",
       "      <td>mitochondrial electron transport, NADH to ubiq...</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0006120</td>\n",
       "      <td>mitochondrial electron transport, NADH to ubiq...</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0006120</td>\n",
       "      <td>mitochondrial electron transport, NADH to ubiq...</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ensembl_id ensembl_transcript_id GO_accession  \\\n",
       "0  ENSG00000198763       ENST00000361453   GO:0006120   \n",
       "1  ENSG00000198763       ENST00000361453   GO:0032981   \n",
       "2  ENSG00000198763       ENST00000361453   GO:0006120   \n",
       "3  ENSG00000198763       ENST00000361453   GO:0006120   \n",
       "4  ENSG00000198763       ENST00000361453   GO:0006120   \n",
       "\n",
       "                                             GO_name           GO_domain  \n",
       "0  mitochondrial electron transport, NADH to ubiq...  biological_process  \n",
       "1  mitochondrial respiratory chain complex I asse...  biological_process  \n",
       "2  mitochondrial electron transport, NADH to ubiq...  biological_process  \n",
       "3  mitochondrial electron transport, NADH to ubiq...  biological_process  \n",
       "4  mitochondrial electron transport, NADH to ubiq...  biological_process  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>ensembl_transcript_id</th>\n",
       "      <th>GO_accession</th>\n",
       "      <th>GO_name</th>\n",
       "      <th>GO_domain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0016020</td>\n",
       "      <td>membrane</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0005739</td>\n",
       "      <td>mitochondrion</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0005743</td>\n",
       "      <td>mitochondrial inner membrane</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0070469</td>\n",
       "      <td>respirasome</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>GO:0005747</td>\n",
       "      <td>mitochondrial respiratory chain complex I</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ensembl_id ensembl_transcript_id GO_accession  \\\n",
       "0  ENSG00000198763       ENST00000361453   GO:0016020   \n",
       "1  ENSG00000198763       ENST00000361453   GO:0005739   \n",
       "2  ENSG00000198763       ENST00000361453   GO:0005743   \n",
       "3  ENSG00000198763       ENST00000361453   GO:0070469   \n",
       "4  ENSG00000198763       ENST00000361453   GO:0005747   \n",
       "\n",
       "                                     GO_name           GO_domain  \n",
       "0                                   membrane  cellular_component  \n",
       "1                              mitochondrion  cellular_component  \n",
       "2               mitochondrial inner membrane  cellular_component  \n",
       "3                                respirasome  cellular_component  \n",
       "4  mitochondrial respiratory chain complex I  cellular_component  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "protein_coding_genes = raw_data.loc[raw_data['ensembl_id'].isin(id_map_cleaned.ensembl_id)] # filter out the protein coding genes only\n",
    "filtered_protein_coding_genes = protein_coding_genes.loc[~protein_coding_genes['GO_accession'].isna()].reset_index(drop=True) #remove the ones without GO_accession \n",
    "filtered_BP_genes = filtered_protein_coding_genes.loc[filtered_protein_coding_genes['GO_domain'] == 'biological_process',].reset_index(drop=True) #filter the biological process domain\n",
    "filtered_CC_genes = filtered_protein_coding_genes.loc[filtered_protein_coding_genes['GO_domain'] == 'cellular_component',].reset_index(drop=True) #filter the cellular component domain\n",
    "\n",
    "print(f'# of genes w/ their corresponding GO term annotations (BP): {filtered_BP_genes.ensembl_id.nunique()} \\n',\n",
    "      f'# of unique GO term accession (BP): {filtered_BP_genes.GO_accession.nunique()}')\n",
    "\n",
    "print(f'# of genes w/ their corresponding GO term annotations (CC): {filtered_CC_genes.ensembl_id.nunique()} \\n',\n",
    "      f'# of unique GO term accession (CC): {filtered_CC_genes.GO_accession.nunique()}')\n",
    "\n",
    "display(filtered_BP_genes.head())\n",
    "display(filtered_CC_genes.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# of paralogs w/ their corresponding GO term annotations (BP): 11955 \n",
      " # of unique GO term accession (BP): 11305\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>ensembl_transcript_id</th>\n",
       "      <th>GO_accession</th>\n",
       "      <th>GO_name</th>\n",
       "      <th>GO_domain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>GO:0006511</td>\n",
       "      <td>ubiquitin-dependent protein catabolic process</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>GO:0016579</td>\n",
       "      <td>protein deubiquitination</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>GO:0006508</td>\n",
       "      <td>proteolysis</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>GO:0016579</td>\n",
       "      <td>protein deubiquitination</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>GO:0016477</td>\n",
       "      <td>cell migration</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ensembl_id ensembl_transcript_id GO_accession  \\\n",
       "0  ENSG00000114374       ENST00000651177   GO:0006511   \n",
       "1  ENSG00000114374       ENST00000651177   GO:0016579   \n",
       "2  ENSG00000114374       ENST00000651177   GO:0006508   \n",
       "3  ENSG00000114374       ENST00000651177   GO:0016579   \n",
       "4  ENSG00000114374       ENST00000651177   GO:0016477   \n",
       "\n",
       "                                         GO_name           GO_domain  \n",
       "0  ubiquitin-dependent protein catabolic process  biological_process  \n",
       "1                       protein deubiquitination  biological_process  \n",
       "2                                    proteolysis  biological_process  \n",
       "3                       protein deubiquitination  biological_process  \n",
       "4                                 cell migration  biological_process  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# of paralogs w/ their corresponding GO term annotations (CC): 12423 \n",
      " # of unique GO term accession (CC): 1524\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>ensembl_transcript_id</th>\n",
       "      <th>GO_accession</th>\n",
       "      <th>GO_name</th>\n",
       "      <th>GO_domain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>GO:0005737</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>GO:0005737</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>GO:0005634</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>ENST00000651177</td>\n",
       "      <td>GO:0005829</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>ENST00000338981</td>\n",
       "      <td>GO:0005737</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ensembl_id ensembl_transcript_id GO_accession    GO_name  \\\n",
       "0  ENSG00000114374       ENST00000651177   GO:0005737  cytoplasm   \n",
       "1  ENSG00000114374       ENST00000651177   GO:0005737  cytoplasm   \n",
       "2  ENSG00000114374       ENST00000651177   GO:0005634    nucleus   \n",
       "3  ENSG00000114374       ENST00000651177   GO:0005829    cytosol   \n",
       "4  ENSG00000114374       ENST00000338981   GO:0005737  cytoplasm   \n",
       "\n",
       "            GO_domain  \n",
       "0  cellular_component  \n",
       "1  cellular_component  \n",
       "2  cellular_component  \n",
       "3  cellular_component  \n",
       "4  cellular_component  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# filter out the paralogs only\n",
    "filtered_BP_paralogs = filtered_BP_genes.loc[filtered_BP_genes['ensembl_id'].isin(list_of_paralogs)].reset_index(drop=True) \n",
    "\n",
    "print(f'# of paralogs w/ their corresponding GO term annotations (BP): {filtered_BP_paralogs.ensembl_id.nunique()} \\n',\n",
    "      f'# of unique GO term accession (BP): {filtered_BP_paralogs.GO_accession.nunique()}')\n",
    "\n",
    "display(filtered_BP_paralogs.head())\n",
    "\n",
    "filtered_CC_paralogs = filtered_CC_genes.loc[filtered_CC_genes['ensembl_id'].isin(list_of_paralogs)].reset_index(drop=True)\n",
    "\n",
    "print(f'# of paralogs w/ their corresponding GO term annotations (CC): {filtered_CC_paralogs.ensembl_id.nunique()} \\n',\n",
    "      f'# of unique GO term accession (CC): {filtered_CC_paralogs.GO_accession.nunique()}')\n",
    "\n",
    "display(filtered_CC_paralogs.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO_accession</th>\n",
       "      <th>GO_name</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000002</td>\n",
       "      <td>mitochondrial genome maintenance</td>\n",
       "      <td>{ENSG00000196365, ENSG00000114120, ENSG0000017...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000003</td>\n",
       "      <td>reproduction</td>\n",
       "      <td>{ENSG00000189409}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000012</td>\n",
       "      <td>single strand break repair</td>\n",
       "      <td>{ENSG00000096717, ENSG00000137074, ENSG0000022...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0000017</td>\n",
       "      <td>alpha-glucoside transport</td>\n",
       "      <td>{ENSG00000100170, ENSG00000140675}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0000018</td>\n",
       "      <td>regulation of DNA recombination</td>\n",
       "      <td>{ENSG00000114030, ENSG00000182481}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  GO_accession                           GO_name  \\\n",
       "0   GO:0000002  mitochondrial genome maintenance   \n",
       "1   GO:0000003                      reproduction   \n",
       "2   GO:0000012        single strand break repair   \n",
       "3   GO:0000017         alpha-glucoside transport   \n",
       "4   GO:0000018   regulation of DNA recombination   \n",
       "\n",
       "                                               genes  \n",
       "0  {ENSG00000196365, ENSG00000114120, ENSG0000017...  \n",
       "1                                  {ENSG00000189409}  \n",
       "2  {ENSG00000096717, ENSG00000137074, ENSG0000022...  \n",
       "3                 {ENSG00000100170, ENSG00000140675}  \n",
       "4                 {ENSG00000114030, ENSG00000182481}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO_accession</th>\n",
       "      <th>GO_name</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000015</td>\n",
       "      <td>phosphopyruvate hydratase complex</td>\n",
       "      <td>{ENSG00000188316, ENSG00000074800, ENSG0000010...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000118</td>\n",
       "      <td>histone deacetylase complex</td>\n",
       "      <td>{ENSG00000108840, ENSG00000094916, ENSG0000019...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000123</td>\n",
       "      <td>histone acetyltransferase complex</td>\n",
       "      <td>{ENSG00000111653, ENSG00000136504, ENSG0000010...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0000124</td>\n",
       "      <td>SAGA complex</td>\n",
       "      <td>{ENSG00000114166, ENSG00000162227, ENSG0000016...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0000137</td>\n",
       "      <td>Golgi cis cisterna</td>\n",
       "      <td>{ENSG00000174567, ENSG00000188626, ENSG0000013...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  GO_accession                            GO_name  \\\n",
       "0   GO:0000015  phosphopyruvate hydratase complex   \n",
       "1   GO:0000118        histone deacetylase complex   \n",
       "2   GO:0000123  histone acetyltransferase complex   \n",
       "3   GO:0000124                       SAGA complex   \n",
       "4   GO:0000137                 Golgi cis cisterna   \n",
       "\n",
       "                                               genes  \n",
       "0  {ENSG00000188316, ENSG00000074800, ENSG0000010...  \n",
       "1  {ENSG00000108840, ENSG00000094916, ENSG0000019...  \n",
       "2  {ENSG00000111653, ENSG00000136504, ENSG0000010...  \n",
       "3  {ENSG00000114166, ENSG00000162227, ENSG0000016...  \n",
       "4  {ENSG00000174567, ENSG00000188626, ENSG0000013...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sum_bp_paralogs = filtered_BP_paralogs.groupby(['GO_accession', 'GO_name'])['ensembl_id'].apply(set).reset_index(name=\"genes\")\n",
    "display(sum_bp_paralogs.head())\n",
    "\n",
    "sum_cc_paralogs = filtered_CC_paralogs.groupby(['GO_accession', 'GO_name'])['ensembl_id'].apply(set).reset_index(name=\"genes\")\n",
    "display(sum_cc_paralogs.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO_accession</th>\n",
       "      <th>GO_name</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000002</td>\n",
       "      <td>mitochondrial genome maintenance</td>\n",
       "      <td>{ENSG00000196365, ENSG00000114120, ENSG0000017...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000038</td>\n",
       "      <td>very long-chain fatty acid metabolic process</td>\n",
       "      <td>{ENSG00000101986, ENSG00000140284, ENSG0000011...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000045</td>\n",
       "      <td>autophagosome assembly</td>\n",
       "      <td>{ENSG00000198925, ENSG00000034713, ENSG0000015...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0000050</td>\n",
       "      <td>urea cycle</td>\n",
       "      <td>{ENSG00000118520, ENSG00000081181, ENSG0000012...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0000070</td>\n",
       "      <td>mitotic sister chromatid segregation</td>\n",
       "      <td>{ENSG00000166851, ENSG00000037042, ENSG0000023...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  GO_accession                                       GO_name  \\\n",
       "0   GO:0000002              mitochondrial genome maintenance   \n",
       "1   GO:0000038  very long-chain fatty acid metabolic process   \n",
       "2   GO:0000045                        autophagosome assembly   \n",
       "3   GO:0000050                                    urea cycle   \n",
       "4   GO:0000070          mitotic sister chromatid segregation   \n",
       "\n",
       "                                               genes  \n",
       "0  {ENSG00000196365, ENSG00000114120, ENSG0000017...  \n",
       "1  {ENSG00000101986, ENSG00000140284, ENSG0000011...  \n",
       "2  {ENSG00000198925, ENSG00000034713, ENSG0000015...  \n",
       "3  {ENSG00000118520, ENSG00000081181, ENSG0000012...  \n",
       "4  {ENSG00000166851, ENSG00000037042, ENSG0000023...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# filter biological processes - include processes with 5 to 100 expresses genes\n",
    "sum_bp_paralogs['sm_biological_process'] = sum_bp_paralogs['genes'].apply(lambda x: 100 > len(x) > 5)\n",
    "sum_bp_paralogs_df = sum_bp_paralogs[sum_bp_paralogs['sm_biological_process'] == True].reset_index(drop=True).drop(['sm_biological_process'], axis=1)\n",
    "display(sum_bp_paralogs_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO_accession</th>\n",
       "      <th>GO_name</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000118</td>\n",
       "      <td>histone deacetylase complex</td>\n",
       "      <td>{ENSG00000108840, ENSG00000094916, ENSG0000019...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000123</td>\n",
       "      <td>histone acetyltransferase complex</td>\n",
       "      <td>{ENSG00000111653, ENSG00000136504, ENSG0000010...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000124</td>\n",
       "      <td>SAGA complex</td>\n",
       "      <td>{ENSG00000114166, ENSG00000162227, ENSG0000016...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0000137</td>\n",
       "      <td>Golgi cis cisterna</td>\n",
       "      <td>{ENSG00000174567, ENSG00000188626, ENSG0000013...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0000138</td>\n",
       "      <td>Golgi trans cisterna</td>\n",
       "      <td>{ENSG00000198964, ENSG00000130733, ENSG0000008...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  GO_accession                            GO_name  \\\n",
       "0   GO:0000118        histone deacetylase complex   \n",
       "1   GO:0000123  histone acetyltransferase complex   \n",
       "2   GO:0000124                       SAGA complex   \n",
       "3   GO:0000137                 Golgi cis cisterna   \n",
       "4   GO:0000138               Golgi trans cisterna   \n",
       "\n",
       "                                               genes  \n",
       "0  {ENSG00000108840, ENSG00000094916, ENSG0000019...  \n",
       "1  {ENSG00000111653, ENSG00000136504, ENSG0000010...  \n",
       "2  {ENSG00000114166, ENSG00000162227, ENSG0000016...  \n",
       "3  {ENSG00000174567, ENSG00000188626, ENSG0000013...  \n",
       "4  {ENSG00000198964, ENSG00000130733, ENSG0000008...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# filter cellular component - include processes with 5 to 100 expresses genes\n",
    "sum_cc_paralogs['sm_cellular_component'] = sum_cc_paralogs['genes'].apply(lambda x: 100 > len(x) > 5)\n",
    "sum_cc_paralogs_df = sum_cc_paralogs[sum_cc_paralogs['sm_cellular_component'] == True].reset_index(drop=True).drop(['sm_cellular_component'], axis=1)\n",
    "display(sum_cc_paralogs_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def precompute_go_effects(go_df, gene_effect):\n",
    "    go_to_mean_effect = {}\n",
    "    valid_go_df = []\n",
    "\n",
    "    for _, row in go_df.iterrows():\n",
    "        go_id = row['GO_accession']\n",
    "        gene_set = list(row['genes'])\n",
    "        genes_in_effect = [g for g in gene_set if g in gene_effect.columns]\n",
    "        if len(genes_in_effect) >= 2:\n",
    "            go_to_mean_effect[go_id] = gene_effect[genes_in_effect].mean(axis=1)\n",
    "            valid_go_df.append(row)\n",
    "\n",
    "    return go_to_mean_effect, pd.DataFrame(valid_go_df)\n",
    "\n",
    "def build_gene_to_go_map(go_df):\n",
    "    gene_to_go = defaultdict(set)\n",
    "    for _, row in go_df.iterrows():\n",
    "        go_id = row['GO_accession']\n",
    "        for gene in row['genes']:\n",
    "            gene_to_go[gene].add(go_id)\n",
    "    return gene_to_go\n",
    "\n",
    "def process_gene_pair_optimized(pair, go_df, go_to_mean_effect, gene_to_go):\n",
    "    a1, a2 = pair\n",
    "    shared_go_ids = gene_to_go.get(a1, set()).intersection(gene_to_go.get(a2, set()))\n",
    "\n",
    "    if not shared_go_ids:\n",
    "        return pd.DataFrame(), pd.DataFrame()\n",
    "\n",
    "    go_processes_for_pair = []\n",
    "    go_name_list = []\n",
    "    len_of_go = []\n",
    "    pair_essentiality_df = pd.DataFrame()\n",
    "\n",
    "    for go_id in shared_go_ids:\n",
    "        row = go_df[go_df['GO_accession'] == go_id].iloc[0]\n",
    "        gene_set = row['genes']\n",
    "        go_processes_for_pair.append(go_id)\n",
    "        go_name_list.append(row['GO_name'])\n",
    "        len_of_go.append(len(gene_set))\n",
    "        pair_essentiality_df[go_id] = go_to_mean_effect[go_id]\n",
    "\n",
    "    pair_df = pd.DataFrame({\n",
    "        'paralog_pairs': [pair] * len(go_processes_for_pair),\n",
    "        'GO_accession': go_processes_for_pair,\n",
    "        'GO_name': go_name_list,\n",
    "        'len_of_go': len_of_go\n",
    "    }).sort_values(by='len_of_go', ascending=False).reset_index(drop=True)\n",
    "\n",
    "    return pair_df, pair_essentiality_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Precompute mean gene effect values for each GO term\n",
    "go_to_mean_effect, filtered_go_df = precompute_go_effects(sum_bp_paralogs_df, gene_essentiality)\n",
    "\n",
    "# Build a lookup map from gene → GO terms\n",
    "gene_to_go = build_gene_to_go_map(filtered_go_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def summarize_GO_scores(pair_df, pair_essentiality_df, pair, paralog_pairs_dict=None):\n",
    "    if pair_df.empty or pair_essentiality_df.empty:\n",
    "        return None\n",
    "\n",
    "    result = {}\n",
    "\n",
    "    # --- smallest_GO_accession (per cell line) ---\n",
    "    min_len = pair_df['len_of_go'].min()\n",
    "    smallest_df = pair_df[pair_df['len_of_go'] == min_len]\n",
    "    smallest_go_ids = smallest_df['GO_accession'].tolist()\n",
    "    smallest_subset = pair_essentiality_df[smallest_go_ids]\n",
    "\n",
    "    # Pick GO with lowest essentiality per cell line\n",
    "    result['smallest_GO_accession'] = smallest_subset.idxmin(axis=1)\n",
    "    result['smallest_gene_effect'] = smallest_subset.min(axis=1)\n",
    "\n",
    "    # --- min/max GO by essentiality (per cell line) ---\n",
    "    result['min_GO_accession'] = pair_essentiality_df.idxmin(axis=1)\n",
    "    result['min_gene_effect'] = pair_essentiality_df.min(axis=1)\n",
    "\n",
    "    result['max_GO_accession'] = pair_essentiality_df.idxmax(axis=1)\n",
    "    result['max_gene_effect'] = pair_essentiality_df.max(axis=1)\n",
    "\n",
    "    # --- paralog pair label ---\n",
    "    if paralog_pairs_dict is not None:\n",
    "        result['paralog_pair'] = paralog_pairs_dict.get(pair, '|'.join(pair))\n",
    "    else:\n",
    "        result['paralog_pair'] = '|'.join(pair)\n",
    "\n",
    "    return pd.DataFrame({\n",
    "        'cell_line': pair_essentiality_df.index,\n",
    "        'paralog_pair': result['paralog_pair'],\n",
    "        'smallest_GO_accession': result['smallest_GO_accession'].values,\n",
    "        'smallest_gene_effect': result['smallest_gene_effect'].values,\n",
    "        'min_GO_accession': result['min_GO_accession'].values,\n",
    "        'min_gene_effect': result['min_gene_effect'].values,\n",
    "        'max_GO_accession': result['max_GO_accession'].values,\n",
    "        'max_gene_effect': result['max_gene_effect'].values,\n",
    "    })\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_summaries = []\n",
    "\n",
    "for pair in paralog_pairs:\n",
    "    pair_df, pair_essentiality_df = process_gene_pair_optimized(\n",
    "        pair, filtered_go_df, go_to_mean_effect, gene_to_go\n",
    "    )\n",
    "    summary_df = summarize_GO_scores(pair_df, pair_essentiality_df, pair, paralog_pairs_dict)\n",
    "    \n",
    "    if summary_df is not None:\n",
    "        all_summaries.append(summary_df)\n",
    "\n",
    "final_df = pd.concat(all_summaries).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO_accession</th>\n",
       "      <th>GO_name</th>\n",
       "      <th>genes</th>\n",
       "      <th>len_of_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>GO:0001843</td>\n",
       "      <td>neural tube closure</td>\n",
       "      <td>{ENSG00000147050, ENSG00000104290, ENSG0000007...</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>GO:0003091</td>\n",
       "      <td>renal water homeostasis</td>\n",
       "      <td>{ENSG00000171885, ENSG00000169344, ENSG0000016...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>846</th>\n",
       "      <td>GO:0010737</td>\n",
       "      <td>protein kinase A signaling</td>\n",
       "      <td>{ENSG00000165059, ENSG00000136167, ENSG0000021...</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1585</th>\n",
       "      <td>GO:0034380</td>\n",
       "      <td>high-density lipoprotein particle assembly</td>\n",
       "      <td>{ENSG00000165059, ENSG00000118137, ENSG0000014...</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2936</th>\n",
       "      <td>GO:0070613</td>\n",
       "      <td>regulation of protein processing</td>\n",
       "      <td>{ENSG00000204084, ENSG00000142875, ENSG0000007...</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     GO_accession                                     GO_name  \\\n",
       "95     GO:0001843                         neural tube closure   \n",
       "226    GO:0003091                     renal water homeostasis   \n",
       "846    GO:0010737                  protein kinase A signaling   \n",
       "1585   GO:0034380  high-density lipoprotein particle assembly   \n",
       "2936   GO:0070613            regulation of protein processing   \n",
       "\n",
       "                                                  genes  len_of_genes  \n",
       "95    {ENSG00000147050, ENSG00000104290, ENSG0000007...            58  \n",
       "226   {ENSG00000171885, ENSG00000169344, ENSG0000016...            16  \n",
       "846   {ENSG00000165059, ENSG00000136167, ENSG0000021...            15  \n",
       "1585  {ENSG00000165059, ENSG00000118137, ENSG0000014...             7  \n",
       "2936  {ENSG00000204084, ENSG00000142875, ENSG0000007...             7  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>smallest_GO_accession</th>\n",
       "      <th>smallest_gene_effect</th>\n",
       "      <th>min_GO_accession</th>\n",
       "      <th>min_gene_effect</th>\n",
       "      <th>max_GO_accession</th>\n",
       "      <th>max_gene_effect</th>\n",
       "      <th>GO:0001843</th>\n",
       "      <th>GO:0003091</th>\n",
       "      <th>GO:0010737</th>\n",
       "      <th>GO:0034380</th>\n",
       "      <th>GO:0070613</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACH-000022</th>\n",
       "      <td>GO:0034380</td>\n",
       "      <td>6369.285714</td>\n",
       "      <td>GO:0034380</td>\n",
       "      <td>6369.285714</td>\n",
       "      <td>GO:0070613</td>\n",
       "      <td>10245.285714</td>\n",
       "      <td>9228.87931</td>\n",
       "      <td>9576.0</td>\n",
       "      <td>10074.5</td>\n",
       "      <td>6369.285714</td>\n",
       "      <td>10245.285714</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           smallest_GO_accession  smallest_gene_effect min_GO_accession  \\\n",
       "ACH-000022            GO:0034380           6369.285714       GO:0034380   \n",
       "\n",
       "            min_gene_effect max_GO_accession  max_gene_effect  GO:0001843  \\\n",
       "ACH-000022      6369.285714       GO:0070613     10245.285714  9228.87931   \n",
       "\n",
       "            GO:0003091  GO:0010737   GO:0034380    GO:0070613  \n",
       "ACH-000022      9576.0     10074.5  6369.285714  10245.285714  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_line</th>\n",
       "      <th>paralog_pair</th>\n",
       "      <th>smallest_GO_accession</th>\n",
       "      <th>smallest_gene_effect</th>\n",
       "      <th>min_GO_accession</th>\n",
       "      <th>min_gene_effect</th>\n",
       "      <th>max_GO_accession</th>\n",
       "      <th>max_gene_effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>164173</th>\n",
       "      <td>ACH-000022</td>\n",
       "      <td>PRKACA_PRKACB</td>\n",
       "      <td>GO:0034380</td>\n",
       "      <td>6369.285714</td>\n",
       "      <td>GO:0034380</td>\n",
       "      <td>6369.285714</td>\n",
       "      <td>GO:0070613</td>\n",
       "      <td>10245.285714</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         cell_line   paralog_pair smallest_GO_accession  smallest_gene_effect  \\\n",
       "164173  ACH-000022  PRKACA_PRKACB            GO:0034380           6369.285714   \n",
       "\n",
       "       min_GO_accession  min_gene_effect max_GO_accession  max_gene_effect  \n",
       "164173       GO:0034380      6369.285714       GO:0070613     10245.285714  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "###### Test ######\n",
    "\n",
    "#PRKACA_PRKACB\n",
    "pair = {'ENSG00000072062', 'ENSG00000142875'}\n",
    "\n",
    "filtered_df = sum_bp_paralogs_df[sum_bp_paralogs_df['genes'].apply(lambda x: pair.issubset(x))]\n",
    "filtered_df.insert(filtered_df.shape[1], 'len_of_genes', filtered_df['genes'].apply(len))\n",
    "display(filtered_df)\n",
    "\n",
    "go_effect_subset = {go: go_to_mean_effect[go] for go in filtered_df['GO_accession']}\n",
    "go_effect_df = pd.DataFrame(go_effect_subset)\n",
    "\n",
    "min_len = filtered_df['len_of_genes'].min()\n",
    "smallest_df = filtered_df[filtered_df['len_of_genes'] == min_len]\n",
    "smallest_go_ids = smallest_df['GO_accession'].tolist()\n",
    "smallest_subset = go_effect_df[smallest_go_ids]\n",
    "\n",
    "#smallest GOs: GO:0070613 and GO:0034380\n",
    "\n",
    "result_df = pd.DataFrame(index=go_effect_df.index)\n",
    "result_df['smallest_GO_accession'] = smallest_subset.idxmin(axis=1)\n",
    "result_df['smallest_gene_effect'] = smallest_subset.min(axis=1)\n",
    "result_df['min_GO_accession'] = go_effect_df.idxmin(axis=1)\n",
    "result_df['min_gene_effect'] = go_effect_df.min(axis=1)\n",
    "result_df['max_GO_accession'] = go_effect_df.idxmax(axis=1)\n",
    "result_df['max_gene_effect'] = go_effect_df.max(axis=1)\n",
    "\n",
    "# include all GO expression values for comparison\n",
    "result_df = pd.concat([result_df, go_effect_df], axis=1)\n",
    "display(result_df.loc[result_df.index == 'ACH-000022'])\n",
    "\n",
    "display(final_df.loc[(final_df.paralog_pair == 'PRKACA_PRKACB') & (final_df.cell_line == 'ACH-000022'), :])\n",
    "\n",
    "###### Test ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_to_mean_effect_bp_norm, filtered_go_df_bp_norm = precompute_go_effects(sum_bp_paralogs_df, gene_z_essentiality)\n",
    "\n",
    "gene_to_go_map_bp_norm = build_gene_to_go_map(filtered_go_df_bp_norm)\n",
    "\n",
    "bp_norm_summaries = []\n",
    "\n",
    "for pair in paralog_pairs:\n",
    "    pair_df, pair_essentiality_df = process_gene_pair_optimized(\n",
    "        pair, filtered_go_df_bp_norm, go_to_mean_effect_bp_norm, gene_to_go_map_bp_norm\n",
    "    )\n",
    "\n",
    "    summary_df = summarize_GO_scores(pair_df, pair_essentiality_df, pair, paralog_pairs_dict)\n",
    "\n",
    "    if summary_df is not None:\n",
    "        bp_norm_summaries.append(summary_df)\n",
    "\n",
    "# Step 4: Concatenate into final normalized BP summary\n",
    "final_bp_norm_df = pd.concat(bp_norm_summaries).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15795, 35656)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_bp_norm_df['paralog_pair'].nunique(), paralog_pairs.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df = final_df.rename(columns={\n",
    "    'smallest_GO_accession': 'smallest_BP_GO_accession',\n",
    "    'smallest_gene_effect': 'smallest_BP_GO_essentiality',\n",
    "    'min_GO_accession': 'min_BP_GO_accession', \n",
    "    'min_gene_effect': 'min_BP_GO_essentiality',\n",
    "    'max_GO_accession': 'max_BP_GO_accession', \n",
    "    'max_gene_effect': 'max_BP_GO_essentiality',\n",
    "})\n",
    "\n",
    "final_bp_norm_df = final_bp_norm_df.rename(columns={\n",
    "    'smallest_GO_accession': 'zsmallest_BP_GO_accession',\n",
    "    'smallest_gene_effect': 'zsmallest_BP_GO_essentiality',\n",
    "    'min_GO_accession': 'zmin_BP_GO_accession', \n",
    "    'min_gene_effect': 'zmin_BP_GO_essentiality',\n",
    "    'max_GO_accession': 'zmax_BP_GO_accession', \n",
    "    'max_gene_effect': 'zmax_BP_GO_essentiality',\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "BP_GO_merged_df = pd.merge(\n",
    "    final_df, final_bp_norm_df,\n",
    "    on=['paralog_pair', 'cell_line'],\n",
    "    how='outer',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_line</th>\n",
       "      <th>paralog_pair</th>\n",
       "      <th>smallest_BP_GO_accession</th>\n",
       "      <th>smallest_BP_GO_essentiality</th>\n",
       "      <th>min_BP_GO_accession</th>\n",
       "      <th>min_BP_GO_essentiality</th>\n",
       "      <th>max_BP_GO_accession</th>\n",
       "      <th>max_BP_GO_essentiality</th>\n",
       "      <th>zsmallest_BP_GO_accession</th>\n",
       "      <th>zsmallest_BP_GO_essentiality</th>\n",
       "      <th>zmin_BP_GO_accession</th>\n",
       "      <th>zmin_BP_GO_essentiality</th>\n",
       "      <th>zmax_BP_GO_accession</th>\n",
       "      <th>zmax_BP_GO_essentiality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>A1CF_RBM47</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>8093.583333</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>8093.583333</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>8093.583333</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>-0.693127</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>-0.693127</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>-0.693127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000004</td>\n",
       "      <td>A1CF_RBM47</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>10267.583333</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>10267.583333</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>10267.583333</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>-0.030233</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>-0.030233</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>-0.030233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000005</td>\n",
       "      <td>A1CF_RBM47</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>10641.000000</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>10641.000000</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>10641.000000</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>0.100157</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>0.100157</td>\n",
       "      <td>GO:0016554</td>\n",
       "      <td>0.100157</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cell_line paralog_pair smallest_BP_GO_accession  \\\n",
       "0  ACH-000001   A1CF_RBM47               GO:0016554   \n",
       "1  ACH-000004   A1CF_RBM47               GO:0016554   \n",
       "2  ACH-000005   A1CF_RBM47               GO:0016554   \n",
       "\n",
       "   smallest_BP_GO_essentiality min_BP_GO_accession  min_BP_GO_essentiality  \\\n",
       "0                  8093.583333          GO:0016554             8093.583333   \n",
       "1                 10267.583333          GO:0016554            10267.583333   \n",
       "2                 10641.000000          GO:0016554            10641.000000   \n",
       "\n",
       "  max_BP_GO_accession  max_BP_GO_essentiality zsmallest_BP_GO_accession  \\\n",
       "0          GO:0016554             8093.583333                GO:0016554   \n",
       "1          GO:0016554            10267.583333                GO:0016554   \n",
       "2          GO:0016554            10641.000000                GO:0016554   \n",
       "\n",
       "   zsmallest_BP_GO_essentiality zmin_BP_GO_accession  zmin_BP_GO_essentiality  \\\n",
       "0                     -0.693127           GO:0016554                -0.693127   \n",
       "1                     -0.030233           GO:0016554                -0.030233   \n",
       "2                      0.100157           GO:0016554                 0.100157   \n",
       "\n",
       "  zmax_BP_GO_accession  zmax_BP_GO_essentiality  \n",
       "0           GO:0016554                -0.693127  \n",
       "1           GO:0016554                -0.030233  \n",
       "2           GO:0016554                 0.100157  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BP_GO_merged_df[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def summarize_GO_scores(pair_df, pair_essentiality_df, pair, paralog_pairs_dict=None):\n",
    "    # --- Initial validation ---\n",
    "    if pair_df.empty or pair_essentiality_df.empty:\n",
    "        return None\n",
    "\n",
    "    # --- Drop invalid rows and columns ---\n",
    "    pair_essentiality_df = pair_essentiality_df.dropna(how='all')\n",
    "    if pair_essentiality_df.empty:\n",
    "        return None\n",
    "\n",
    "    pair_essentiality_df = pair_essentiality_df.dropna(axis=1, how='all')\n",
    "    if pair_essentiality_df.empty:\n",
    "        return None\n",
    "\n",
    "    # --- Set base index (cell lines) ---\n",
    "    cell_lines = pair_essentiality_df.index\n",
    "\n",
    "    result = {}\n",
    "\n",
    "    # --- Smallest GO logic ---\n",
    "    min_len = pair_df['len_of_go'].min()\n",
    "    smallest_df = pair_df[pair_df['len_of_go'] == min_len]\n",
    "    smallest_go_ids = [go for go in smallest_df['GO_accession'] if go in pair_essentiality_df.columns]\n",
    "\n",
    "    if not smallest_go_ids:\n",
    "        return None\n",
    "\n",
    "    smallest_subset = pair_essentiality_df[smallest_go_ids].dropna(how='all')\n",
    "    if smallest_subset.empty:\n",
    "        return None\n",
    "\n",
    "    result['smallest_GO_accession'] = smallest_subset.idxmin(axis=1).reindex(cell_lines)\n",
    "    result['smallest_gene_effect'] = smallest_subset.min(axis=1).reindex(cell_lines)\n",
    "\n",
    "    # --- Min/Max across all GO terms (row-wise) ---\n",
    "    result['min_GO_accession'] = pair_essentiality_df.idxmin(axis=1).reindex(cell_lines)\n",
    "    result['min_gene_effect'] = pair_essentiality_df.min(axis=1).reindex(cell_lines)\n",
    "\n",
    "    result['max_GO_accession'] = pair_essentiality_df.idxmax(axis=1).reindex(cell_lines)\n",
    "    result['max_gene_effect'] = pair_essentiality_df.max(axis=1).reindex(cell_lines)\n",
    "\n",
    "    # --- Paralog pair label ---\n",
    "    if paralog_pairs_dict is not None:\n",
    "        pair_name = paralog_pairs_dict.get(pair, '|'.join(pair))\n",
    "    else:\n",
    "        pair_name = '|'.join(pair)\n",
    "\n",
    "    result['paralog_pair'] = [pair_name] * len(cell_lines)\n",
    "\n",
    "    # --- Final DataFrame ---\n",
    "    return pd.DataFrame({\n",
    "        'cell_line': cell_lines,\n",
    "        'paralog_pair': result['paralog_pair'],\n",
    "        'smallest_GO_accession': result['smallest_GO_accession'].values,\n",
    "        'smallest_gene_effect': result['smallest_gene_effect'].values,\n",
    "        'min_GO_accession': result['min_GO_accession'].values,\n",
    "        'min_gene_effect': result['min_gene_effect'].values,\n",
    "        'max_GO_accession': result['max_GO_accession'].values,\n",
    "        'max_gene_effect': result['max_gene_effect'].values,\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Precompute for CC\n",
    "go_to_mean_effect_cc, filtered_go_df_cc = precompute_go_effects(sum_cc_paralogs_df, gene_essentiality)\n",
    "\n",
    "# Build gene → GO map for CC\n",
    "gene_to_go_cc = build_gene_to_go_map(filtered_go_df_cc)\n",
    "\n",
    "all_summaries_cc = []\n",
    "\n",
    "for pair in paralog_pairs:\n",
    "    pair_df, pair_essentiality_df = process_gene_pair_optimized(\n",
    "        pair, filtered_go_df_cc, go_to_mean_effect_cc, gene_to_go_cc\n",
    "    )\n",
    "    \n",
    "    summary_df = summarize_GO_scores(pair_df, pair_essentiality_df, pair, paralog_pairs_dict)\n",
    "    \n",
    "    if summary_df is not None:\n",
    "        all_summaries_cc.append(summary_df)\n",
    "\n",
    "# Concatenate the results\n",
    "final_df_cc = pd.concat(all_summaries_cc).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6296, 35656)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df_cc['paralog_pair'].nunique(), paralog_pairs.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_line</th>\n",
       "      <th>paralog_pair</th>\n",
       "      <th>smallest_GO_accession</th>\n",
       "      <th>smallest_gene_effect</th>\n",
       "      <th>min_GO_accession</th>\n",
       "      <th>min_gene_effect</th>\n",
       "      <th>max_GO_accession</th>\n",
       "      <th>max_gene_effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>118813</th>\n",
       "      <td>ACH-000022</td>\n",
       "      <td>PRKACA_PRKACB</td>\n",
       "      <td>GO:0005952</td>\n",
       "      <td>8908.375</td>\n",
       "      <td>GO:0005952</td>\n",
       "      <td>8908.375</td>\n",
       "      <td>GO:0097546</td>\n",
       "      <td>10236.24</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         cell_line   paralog_pair smallest_GO_accession  smallest_gene_effect  \\\n",
       "118813  ACH-000022  PRKACA_PRKACB            GO:0005952              8908.375   \n",
       "\n",
       "       min_GO_accession  min_gene_effect max_GO_accession  max_gene_effect  \n",
       "118813       GO:0005952         8908.375       GO:0097546         10236.24  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(final_df_cc.loc[(final_df_cc.paralog_pair == 'PRKACA_PRKACB') & (final_df_cc.cell_line == 'ACH-000022'), :])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Precompute for CC using normalized data\n",
    "go_to_mean_effect_cc_norm, filtered_go_df_cc_norm = precompute_go_effects(sum_cc_paralogs_df, gene_z_essentiality)\n",
    "\n",
    "# Build gene → GO map\n",
    "gene_to_go_cc_norm = build_gene_to_go_map(filtered_go_df_cc_norm)\n",
    "\n",
    "# Process all paralog pairs\n",
    "all_summaries_cc_norm = []\n",
    "\n",
    "for pair in paralog_pairs:\n",
    "    pair_df, pair_essentiality_df = process_gene_pair_optimized(\n",
    "        pair, filtered_go_df_cc_norm, go_to_mean_effect_cc_norm, gene_to_go_cc_norm\n",
    "    )\n",
    "\n",
    "    summary_df = summarize_GO_scores(pair_df, pair_essentiality_df, pair, paralog_pairs_dict)\n",
    "\n",
    "    if summary_df is not None:\n",
    "        all_summaries_cc_norm.append(summary_df)\n",
    "\n",
    "# Concatenate results\n",
    "final_df_cc_norm = pd.concat(all_summaries_cc_norm).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6296, 35656)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df_cc_norm['paralog_pair'].nunique(), paralog_pairs.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df_cc = final_df_cc.rename(columns={\n",
    "    'smallest_GO_accession': 'smallest_CC_GO_accession',\n",
    "    'smallest_gene_effect': 'smallest_CC_GO_essentiality',\n",
    "    'min_GO_accession': 'min_CC_GO_accession', \n",
    "    'min_gene_effect': 'min_CC_GO_essentiality',\n",
    "    'max_GO_accession': 'max_CC_GO_accession', \n",
    "    'max_gene_effect': 'max_CC_GO_essentiality',\n",
    "})\n",
    "\n",
    "final_df_cc_norm = final_df_cc_norm.rename(columns={\n",
    "    'smallest_GO_accession': 'zsmallest_CC_GO_accession',\n",
    "    'smallest_gene_effect': 'zsmallest_CC_GO_essentiality',\n",
    "    'min_GO_accession': 'zmin_CC_GO_accession', \n",
    "    'min_gene_effect': 'zmin_CC_GO_essentiality',\n",
    "    'max_GO_accession': 'zmax_CC_GO_accession', \n",
    "    'max_gene_effect': 'zmax_CC_GO_essentiality',\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "CC_GO_merged_df = pd.merge(\n",
    "    final_df_cc, final_df_cc_norm,\n",
    "    on=['paralog_pair', 'cell_line'],\n",
    "    how='outer',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_line</th>\n",
       "      <th>paralog_pair</th>\n",
       "      <th>smallest_CC_GO_accession</th>\n",
       "      <th>smallest_CC_GO_essentiality</th>\n",
       "      <th>min_CC_GO_accession</th>\n",
       "      <th>min_CC_GO_essentiality</th>\n",
       "      <th>max_CC_GO_accession</th>\n",
       "      <th>max_CC_GO_essentiality</th>\n",
       "      <th>zsmallest_CC_GO_accession</th>\n",
       "      <th>zsmallest_CC_GO_essentiality</th>\n",
       "      <th>zmin_CC_GO_accession</th>\n",
       "      <th>zmin_CC_GO_essentiality</th>\n",
       "      <th>zmax_CC_GO_accession</th>\n",
       "      <th>zmax_CC_GO_essentiality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>A2M_C3</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>10072.863636</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>10072.863636</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>10072.863636</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>0.131185</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>0.131185</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>0.131185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000004</td>\n",
       "      <td>A2M_C3</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>9107.590909</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>9107.590909</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>9107.590909</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>-0.191389</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>-0.191389</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>-0.191389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000005</td>\n",
       "      <td>A2M_C3</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>9475.579545</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>9475.579545</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>9475.579545</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>-0.090731</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>-0.090731</td>\n",
       "      <td>GO:0072562</td>\n",
       "      <td>-0.090731</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cell_line paralog_pair smallest_CC_GO_accession  \\\n",
       "0  ACH-000001       A2M_C3               GO:0072562   \n",
       "1  ACH-000004       A2M_C3               GO:0072562   \n",
       "2  ACH-000005       A2M_C3               GO:0072562   \n",
       "\n",
       "   smallest_CC_GO_essentiality min_CC_GO_accession  min_CC_GO_essentiality  \\\n",
       "0                 10072.863636          GO:0072562            10072.863636   \n",
       "1                  9107.590909          GO:0072562             9107.590909   \n",
       "2                  9475.579545          GO:0072562             9475.579545   \n",
       "\n",
       "  max_CC_GO_accession  max_CC_GO_essentiality zsmallest_CC_GO_accession  \\\n",
       "0          GO:0072562            10072.863636                GO:0072562   \n",
       "1          GO:0072562             9107.590909                GO:0072562   \n",
       "2          GO:0072562             9475.579545                GO:0072562   \n",
       "\n",
       "   zsmallest_CC_GO_essentiality zmin_CC_GO_accession  zmin_CC_GO_essentiality  \\\n",
       "0                      0.131185           GO:0072562                 0.131185   \n",
       "1                     -0.191389           GO:0072562                -0.191389   \n",
       "2                     -0.090731           GO:0072562                -0.090731   \n",
       "\n",
       "  zmax_CC_GO_accession  zmax_CC_GO_essentiality  \n",
       "0           GO:0072562                 0.131185  \n",
       "1           GO:0072562                -0.191389  \n",
       "2           GO:0072562                -0.090731  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CC_GO_merged_df[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the files\n",
    "output_path = get_data_path(['data', 'input', 'GO'], '')\n",
    "\n",
    "BP_GO_merged_df.to_parquet(os.path.join(output_path, 'go_BP_ranked_essentiality.parquet'), engine='pyarrow', index=True)\n",
    "CC_GO_merged_df.to_parquet(os.path.join(output_path, 'go_CC_ranked_essentiality.parquet'), engine='pyarrow', index=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "predictor",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
